{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "350a519c",
   "metadata": {},
   "source": [
    "# Create the 'padding'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "27ef8b5d",
   "metadata": {},
   "source": [
    "We extend the seq_df with three nucleotides before and three after each 2003 long sequence in the df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "58768286",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/gnomad_db/database.py:9: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "from gnomad_db.database import gnomAD_DB\n",
    "from pathlib import Path\n",
    "from Bio import SeqIO\n",
    "import numpy as np\n",
    "import pyranges as pr\n",
    "import math "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "becc3164",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63564</td>\n",
       "      <td>65567</td>\n",
       "      <td>+</td>\n",
       "      <td>TATCGATGGGCACCTTCTTTTTCTTAATTGTATCATACATTTTTAT...</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>922431</td>\n",
       "      <td>924434</td>\n",
       "      <td>+</td>\n",
       "      <td>AGAAGACACAGACTTCAGGAGAGGAAGGCACAGGAACTCACTGGCA...</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923941</td>\n",
       "      <td>925944</td>\n",
       "      <td>+</td>\n",
       "      <td>TCCCCGCCGGGCGGGCGCGCGCCAGTGGACGCGGGTGCACGACTGA...</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>958693</td>\n",
       "      <td>960696</td>\n",
       "      <td>+</td>\n",
       "      <td>TCGGGAAGAGATTTTTGCACAACTCACCAACATACGCTCCCTGCCT...</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>964531</td>\n",
       "      <td>966534</td>\n",
       "      <td>+</td>\n",
       "      <td>TCCGCAGTGGGGCTGCGGGGAGGGGGGCGCGGGTCCGCAGTGGGGC...</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome   Start     End Strand  \\\n",
       "0       chr1   63564   65567      +   \n",
       "1       chr1  922431  924434      +   \n",
       "2       chr1  923941  925944      +   \n",
       "3       chr1  958693  960696      +   \n",
       "4       chr1  964531  966534      +   \n",
       "\n",
       "                                                 seq  seq_len  \n",
       "0  TATCGATGGGCACCTTCTTTTTCTTAATTGTATCATACATTTTTAT...     2003  \n",
       "1  AGAAGACACAGACTTCAGGAGAGGAAGGCACAGGAACTCACTGGCA...     2003  \n",
       "2  TCCCCGCCGGGCGGGCGCGCGCCAGTGGACGCGGGTGCACGACTGA...     2003  \n",
       "3  TCGGGAAGAGATTTTTGCACAACTCACCAACATACGCTCCCTGCCT...     2003  \n",
       "4  TCCGCAGTGGGGCTGCGGGGAGGGGGGCGCGGGTCCGCAGTGGGGC...     2003  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seqs_df = pd.read_csv(\"/s/project/ml4rg_students/2025/project07/data/gtf_start_extended_ints_df_2003_seq.csv\")\n",
    "seqs_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f9163dab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63563</td>\n",
       "      <td>65568</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome  Start    End\n",
       "0       chr1  63563  65568"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "row = row[['Chromosome', 'Start', 'End']]\n",
    "row.at[0, 'Start'] = 63563\n",
    "row.at[0, 'End'] = 65568\n",
    "row\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3c66e67e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63561</td>\n",
       "      <td>65570</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>922428</td>\n",
       "      <td>924437</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923938</td>\n",
       "      <td>925947</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>958690</td>\n",
       "      <td>960699</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>964528</td>\n",
       "      <td>966537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30055</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50577909</td>\n",
       "      <td>50579918</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30056</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50582775</td>\n",
       "      <td>50584784</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30057</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627773</td>\n",
       "      <td>50629782</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30058</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627366</td>\n",
       "      <td>50629375</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30059</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50782288</td>\n",
       "      <td>50784297</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30060 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End\n",
       "0           chr1     63561     65570\n",
       "1           chr1    922428    924437\n",
       "2           chr1    923938    925947\n",
       "3           chr1    958690    960699\n",
       "4           chr1    964528    966537\n",
       "...          ...       ...       ...\n",
       "30055      chr22  50577909  50579918\n",
       "30056      chr22  50582775  50584784\n",
       "30057      chr22  50627773  50629782\n",
       "30058      chr22  50627366  50629375\n",
       "30059      chr22  50782288  50784297\n",
       "\n",
       "[30060 rows x 3 columns]"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "input_df = seqs_df[['Chromosome', 'Start', 'End']].copy()\n",
    "input_df['Start'] -= 3\n",
    "input_df['End'] += 3\n",
    "input_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "e6bea759",
   "metadata": {},
   "outputs": [],
   "source": [
    "fasta_path = Path(\"/s/project/ml4rg_students/2025/project07/data/GRCh38.primary_assembly.genome.fa\")\n",
    "\n",
    "def fetch_sequence_from_region_df(regions_df, model_window_size, fasta_path=fasta_path):\n",
    "    '''\n",
    "    Fetches DNA sequences from genomic regions using a reference FASTA file.\n",
    "    Input:\n",
    "        - regions_df (pd.DataFrame): A DataFrame containing genomic regions with at least\n",
    "                                     'Chromosome', 'Start', and 'End' columns.\n",
    "        - model_window_size (int): The expected length of each DNA sequence to extract.\n",
    "        - fasta_path (str): Path to the reference genome FASTA file from which sequences are extracted.\n",
    "\n",
    "    Output:\n",
    "        - seqs_df (pd.DataFrame): A new DataFrame including the original region information\n",
    "                                  plus the following additional columns:\n",
    "                                  * 'seq': the uppercase DNA sequence\n",
    "                                  * 'seq_len': the length of the sequence (should equal model_window_size)\n",
    "    '''\n",
    "    \n",
    "    #Added comment: Make a copy of the input DataFrame to avoid modifying the original\n",
    "    seqs_df = regions_df.copy()\n",
    "\n",
    "    seqs_df[\"index\"] = np.arange(\n",
    "            len(seqs_df)\n",
    "        )  # Pyranges changes the ordering of the rows when converting to PyRanges. we first create a new column with the original order\n",
    "\n",
    "    \n",
    "    #Added comment: Convert the pandas DataFrame to a PyRanges object, which is optimized for genomic interval operations\n",
    "    \n",
    "    # create a pyranges df from the pandas df\n",
    "    seqs_pr = pr.PyRanges(seqs_df)\n",
    "\n",
    "    #Added comment: Fetch DNA sequences for each region using the FASTA file at fasta_path\n",
    "    # The sequences are stored in the `.seq` attribute of the PyRanges object\n",
    "    seqs_pr.seq = pr.get_sequence(\n",
    "        seqs_pr, path=fasta_path\n",
    "    )  # retrieve sequenc based on prepared start and end columns\n",
    "\n",
    "    \n",
    "    seqs_df = seqs_pr.df.copy()  # convert to pandas df\n",
    "\n",
    "    seqs_df = (\n",
    "        seqs_df.sort_values(\"index\").reset_index(drop=True).copy()\n",
    "    )  # sort the dataframe back to the original order\n",
    "\n",
    "    #Added comment: Drop the temporary index column as it is no longer needed\n",
    "    seqs_df = seqs_df.drop(columns=[\"index\"])\n",
    "\n",
    "    #Added comment: Convert all sequences to uppercase (standard for DNA: A, T, C, G)\n",
    "    seqs_df[\"seq\"] = seqs_df[\"seq\"].apply(lambda x: x.upper())  # make seq uppercase\n",
    "    \n",
    "    seqs_df[\"seq_len\"] = seqs_df[\"seq\"].apply(lambda x: len(x))  # add a sequence length\n",
    "\n",
    "    # assert  that the sequence length is the same as the model window size\n",
    "    assert (\n",
    "        seqs_df[\"seq_len\"] == model_window_size\n",
    "    ).all(), f\"{seqs_df['seq_len']} is not {model_window_size}\"\n",
    "\n",
    "\n",
    "    return seqs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "45cfcde7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr1</td>\n",
       "      <td>63561</td>\n",
       "      <td>65570</td>\n",
       "      <td>TCATATCGATGGGCACCTTCTTTTTCTTAATTGTATCATACATTTT...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr1</td>\n",
       "      <td>922428</td>\n",
       "      <td>924437</td>\n",
       "      <td>GGGAGAAGACACAGACTTCAGGAGAGGAAGGCACAGGAACTCACTG...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr1</td>\n",
       "      <td>923938</td>\n",
       "      <td>925947</td>\n",
       "      <td>AAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGCGGGTGCACGAC...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chr1</td>\n",
       "      <td>958690</td>\n",
       "      <td>960699</td>\n",
       "      <td>AGATCGGGAAGAGATTTTTGCACAACTCACCAACATACGCTCCCTG...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chr1</td>\n",
       "      <td>964528</td>\n",
       "      <td>966537</td>\n",
       "      <td>AGGTCCGCAGTGGGGCTGCGGGGAGGGGGGCGCGGGTCCGCAGTGG...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30055</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50577909</td>\n",
       "      <td>50579918</td>\n",
       "      <td>CGCCATCCTGGGGGTTGGTCGGCACCTAGGACGGGGGCAGATGGGT...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30056</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50582775</td>\n",
       "      <td>50584784</td>\n",
       "      <td>CGCCATGGCGCGGGCTCGACCGGGCCCCAGGCCAGGCTGCGCTCCG...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30057</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627773</td>\n",
       "      <td>50629782</td>\n",
       "      <td>GGACATGGGACCGAGGGGTCTGTCCCAAGAGAGGGAGGGCTACTTG...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30058</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50627366</td>\n",
       "      <td>50629375</td>\n",
       "      <td>GCCCATCCGAACCGGGAGCCGGCCGGTCAGGAGGGCGGCCCTGCGG...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30059</th>\n",
       "      <td>chr22</td>\n",
       "      <td>50782288</td>\n",
       "      <td>50784297</td>\n",
       "      <td>TGCCATTGGCTCCTAGCTGTAGAGAGGGGTCCAGCCCAAGGGAGGG...</td>\n",
       "      <td>2009</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30060 rows × 5 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      Chromosome     Start       End  \\\n",
       "0           chr1     63561     65570   \n",
       "1           chr1    922428    924437   \n",
       "2           chr1    923938    925947   \n",
       "3           chr1    958690    960699   \n",
       "4           chr1    964528    966537   \n",
       "...          ...       ...       ...   \n",
       "30055      chr22  50577909  50579918   \n",
       "30056      chr22  50582775  50584784   \n",
       "30057      chr22  50627773  50629782   \n",
       "30058      chr22  50627366  50629375   \n",
       "30059      chr22  50782288  50784297   \n",
       "\n",
       "                                                     seq  seq_len  \n",
       "0      TCATATCGATGGGCACCTTCTTTTTCTTAATTGTATCATACATTTT...     2009  \n",
       "1      GGGAGAAGACACAGACTTCAGGAGAGGAAGGCACAGGAACTCACTG...     2009  \n",
       "2      AAGTCCCCGCCGGGCGGGCGCGCGCCAGTGGACGCGGGTGCACGAC...     2009  \n",
       "3      AGATCGGGAAGAGATTTTTGCACAACTCACCAACATACGCTCCCTG...     2009  \n",
       "4      AGGTCCGCAGTGGGGCTGCGGGGAGGGGGGCGCGGGTCCGCAGTGG...     2009  \n",
       "...                                                  ...      ...  \n",
       "30055  CGCCATCCTGGGGGTTGGTCGGCACCTAGGACGGGGGCAGATGGGT...     2009  \n",
       "30056  CGCCATGGCGCGGGCTCGACCGGGCCCCAGGCCAGGCTGCGCTCCG...     2009  \n",
       "30057  GGACATGGGACCGAGGGGTCTGTCCCAAGAGAGGGAGGGCTACTTG...     2009  \n",
       "30058  GCCCATCCGAACCGGGAGCCGGCCGGTCAGGAGGGCGGCCCTGCGG...     2009  \n",
       "30059  TGCCATTGGCTCCTAGCTGTAGAGAGGGGTCCAGCCCAAGGGAGGG...     2009  \n",
       "\n",
       "[30060 rows x 5 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_seqs_df = fetch_sequence_from_region_df(input_df, 2009)\n",
    "new_seqs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "f2f409da",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "'TATCGATGGGCACCTTCTTTTTCTTAATTGTATCATACATTTT'== 'TATCGATGGGCACCTTCTTTTTCTTAATTGTATCATACATTTT'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "c97cae41",
   "metadata": {},
   "outputs": [],
   "source": [
    "new_seqs_df.to_csv(\"/s/project/ml4rg_students/2025/project07/group_2/data/seqs_2009.csv\", index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c04998bd",
   "metadata": {},
   "source": [
    "# Trying for the neutral model notebook"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "630a20d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_location = '/s/project/benchmark-lm/ssd-cache'\n",
    "fasta = Path(\"/s/project/ml4rg_students/2025/project07/data/GRCh38.primary_assembly.genome.fa\")\n",
    "seqs = pd.read_csv(\"/s/project/ml4rg_students/2025/project07/group_2/data/seqs_2009.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "c4800d32",
   "metadata": {},
   "outputs": [],
   "source": [
    "db = gnomAD_DB(data_location, gnomad_version=\"v4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "86b8bb33",
   "metadata": {},
   "outputs": [],
   "source": [
    "def prep(chrom, start, end):\n",
    "    db_region = db.get_info_for_interval(chrom=chrom, interval_start=start, interval_end = end, query=\"*\")\n",
    "\n",
    "    db_region = db_region[db_region['filter'] == 'PASS'] \n",
    "\n",
    "    interval_length = end - start\n",
    "    db_region['len_ref'] = db_region['ref'].apply(len)\n",
    "    db_region['len_alt'] = db_region['alt'].apply(len)  \n",
    "    db_region['is_snv'] = (db_region['len_ref'] == 1) & (db_region['len_alt'] == 1)\n",
    "\n",
    "    db_region = db_region[db_region['is_snv']].reset_index(drop=True)\n",
    "\n",
    "    return db_region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "e3123ffd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>...</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "      <th>len_ref</th>\n",
       "      <th>len_alt</th>\n",
       "      <th>is_snv</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>10111</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>44330.0</td>\n",
       "      <td>0.000023</td>\n",
       "      <td>35.0416</td>\n",
       "      <td>1.38889</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000487</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>10131</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>105956.0</td>\n",
       "      <td>0.000009</td>\n",
       "      <td>37.0788</td>\n",
       "      <td>2.20376</td>\n",
       "      <td>...</td>\n",
       "      <td>46046.0</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000022</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>10139</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>58784.0</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>36.9422</td>\n",
       "      <td>4.42722</td>\n",
       "      <td>...</td>\n",
       "      <td>14494.0</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000069</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>10140</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>2.0</td>\n",
       "      <td>72760.0</td>\n",
       "      <td>0.000027</td>\n",
       "      <td>36.4297</td>\n",
       "      <td>11.88040</td>\n",
       "      <td>...</td>\n",
       "      <td>35206.0</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000057</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>10141</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>50842.0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>36.3017</td>\n",
       "      <td>4.64246</td>\n",
       "      <td>...</td>\n",
       "      <td>26636.0</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000038</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>661</th>\n",
       "      <td>1</td>\n",
       "      <td>19591</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>9.0</td>\n",
       "      <td>70134.0</td>\n",
       "      <td>0.000128</td>\n",
       "      <td>23.8685</td>\n",
       "      <td>4.30368</td>\n",
       "      <td>...</td>\n",
       "      <td>3876.0</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.001548</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000116</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>662</th>\n",
       "      <td>1</td>\n",
       "      <td>19600</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>66084.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>23.7667</td>\n",
       "      <td>4.18158</td>\n",
       "      <td>...</td>\n",
       "      <td>24606.0</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000041</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>663</th>\n",
       "      <td>1</td>\n",
       "      <td>19664</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67516.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>24.6777</td>\n",
       "      <td>7.29310</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>664</th>\n",
       "      <td>1</td>\n",
       "      <td>19911</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>10.0</td>\n",
       "      <td>50892.0</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>25.626</td>\n",
       "      <td>5.01126</td>\n",
       "      <td>...</td>\n",
       "      <td>3586.0</td>\n",
       "      <td>0.001115</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000196</td>\n",
       "      <td>0.001289</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>665</th>\n",
       "      <td>1</td>\n",
       "      <td>19941</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>52146.0</td>\n",
       "      <td>0.000019</td>\n",
       "      <td>24.4368</td>\n",
       "      <td>4.23050</td>\n",
       "      <td>...</td>\n",
       "      <td>19140.0</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000052</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>666 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom    pos ref alt filter    AC        AN        AF       MQ        QD  \\\n",
       "0       1  10111   C   A   PASS   1.0   44330.0  0.000023  35.0416   1.38889   \n",
       "1       1  10131   C   A   PASS   1.0  105956.0  0.000009  37.0788   2.20376   \n",
       "2       1  10139   A   T   PASS   1.0   58784.0  0.000017  36.9422   4.42722   \n",
       "3       1  10140   A   C   PASS   2.0   72760.0  0.000027  36.4297  11.88040   \n",
       "4       1  10141   C   G   PASS   1.0   50842.0  0.000020  36.3017   4.64246   \n",
       "..    ...    ...  ..  ..    ...   ...       ...       ...      ...       ...   \n",
       "661     1  19591   G   A   PASS   9.0   70134.0  0.000128  23.8685   4.30368   \n",
       "662     1  19600   A   T   PASS   1.0   66084.0  0.000015  23.7667   4.18158   \n",
       "663     1  19664   G   T   PASS   1.0   67516.0  0.000015  24.6777   7.29310   \n",
       "664     1  19911   G   A   PASS  10.0   50892.0  0.000196   25.626   5.01126   \n",
       "665     1  19941   C   T   PASS   1.0   52146.0  0.000019  24.4368   4.23050   \n",
       "\n",
       "     ...  AN_grpmax  AF_grpmax    AF_eas    AF_nfe    AF_fin    AF_afr  \\\n",
       "0    ...        NaN        NaN  0.000000  0.000000  0.000487  0.000000   \n",
       "1    ...    46046.0   0.000022  0.000000  0.000022  0.000000  0.000000   \n",
       "2    ...    14494.0   0.000069  0.000000  0.000000  0.000000  0.000069   \n",
       "3    ...    35206.0   0.000057  0.000000  0.000057  0.000000  0.000000   \n",
       "4    ...    26636.0   0.000038  0.000000  0.000038  0.000000  0.000000   \n",
       "..   ...        ...        ...       ...       ...       ...       ...   \n",
       "661  ...     3876.0   0.001548  0.001548  0.000000  0.000000  0.000116   \n",
       "662  ...    24606.0   0.000041  0.000000  0.000000  0.000000  0.000041   \n",
       "663  ...        NaN        NaN  0.000000  0.000000  0.000000  0.000000   \n",
       "664  ...     3586.0   0.001115  0.000000  0.000052  0.000000  0.000196   \n",
       "665  ...    19140.0   0.000052  0.000000  0.000052  0.000000  0.000000   \n",
       "\n",
       "       AF_asj  len_ref  len_alt  is_snv  \n",
       "0    0.000000        1        1    True  \n",
       "1    0.000000        1        1    True  \n",
       "2    0.000000        1        1    True  \n",
       "3    0.000000        1        1    True  \n",
       "4    0.000000        1        1    True  \n",
       "..        ...      ...      ...     ...  \n",
       "661  0.000000        1        1    True  \n",
       "662  0.000000        1        1    True  \n",
       "663  0.000000        1        1    True  \n",
       "664  0.001289        1        1    True  \n",
       "665  0.000000        1        1    True  \n",
       "\n",
       "[666 rows x 24 columns]"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "region = prep(1, 10000, 20000)\n",
    "region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "947f653e",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml4rg_p7_2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
