{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Kickoff Notebook - Project 07"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To run this notebook you can use the flash2_pedro conda environment. Please do not install any package here. The best is if you can clone this environment ot a new one, or install the required packages in a new environment."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Quickly fetch the allele counts/frequency and other parameters from a specific window in the genome"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/gnomad_db/database.py:9: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.\n",
      "  import pkg_resources\n"
     ]
    }
   ],
   "source": [
    "from collections import Counter, defaultdict\n",
    "import pandas as pd\n",
    "from concurrent.futures import ThreadPoolExecutor\n",
    "import tqdm\n",
    "\n",
    "def extract_kmers(seq, k=7):\n",
    "    return [seq[i:i+k] for i in range(len(seq) - k + 1)]\n",
    "\n",
    "def process_row(row):\n",
    "    region = fetcher.fetch_pass_snvs(chrom=row.Chromosome, start=row.Start, end=row.End)\n",
    "    region['zero_pos'] = region['pos'] - row.Start - 1\n",
    "    \n",
    "    all_kmers = extract_kmers(row.seq, k=7)\n",
    "    total_kmer_counts = Counter(all_kmers)\n",
    "\n",
    "    variation_kmer_counts = Counter()\n",
    "    valid = region[(region['zero_pos'] >= 3) & (region['zero_pos'] <= len(row.seq) - 4)]\n",
    "    \n",
    "    for zero_pos in valid['zero_pos']:\n",
    "        kmer = row.seq[zero_pos - 3 : zero_pos + 4]\n",
    "        variation_kmer_counts[kmer] += 1\n",
    "\n",
    "    return total_kmer_counts, variation_kmer_counts\n",
    "\n",
    "# Parallel execution\n",
    "with ThreadPoolExecutor(max_workers=8) as executor:\n",
    "    results = list(tqdm.tqdm(executor.map(process_row, seqs.itertuples(index=False)), total=len(seqs)))\n",
    "\n",
    "# Aggregate counts\n",
    "total_kmer_counts = Counter()\n",
    "variation_kmer_counts = Counter()\n",
    "\n",
    "for total, var in results:\n",
    "    total_kmer_counts.update(total)\n",
    "    variation_kmer_counts.update(var)\n",
    "\n",
    "# Compute variation frequency\n",
    "kmer_data = []\n",
    "for kmer in total_kmer_counts:\n",
    "    total_count = total_kmer_counts[kmer]\n",
    "    var_count = variation_kmer_counts.get(kmer, 0)\n",
    "    freq = var_count / total_count\n",
    "    kmer_data.append({\n",
    "        'kmer': kmer,\n",
    "        'total_count': total_count,\n",
    "        'var_count': var_count,\n",
    "        'var_freq': freq\n",
    "    })\n",
    "\n",
    "kmer_df = pd.DataFrame(kmer_data)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'gnomAD_DB' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[1], line 7\u001b[0m\n\u001b[1;32m      1\u001b[0m database_location \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/s/project/benchmark-lm/ssd-cache\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m#database_location = '/data/nasif12/home_if12/s_gjamn/gnomad_cache'\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \n\u001b[1;32m      4\u001b[0m \u001b[38;5;66;03m#here I just used the data which is in our project, I hope it is the right one\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \u001b[38;5;66;03m#Ok it is not the right one. Here I am not working with the right data\u001b[39;00m\n\u001b[1;32m      6\u001b[0m \u001b[38;5;66;03m#database_location = '/s/project/ml4rg_students/2025/project07/data'\u001b[39;00m\n\u001b[0;32m----> 7\u001b[0m db \u001b[38;5;241m=\u001b[39m \u001b[43mgnomAD_DB\u001b[49m(database_location, gnomad_version\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mv4\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'gnomAD_DB' is not defined"
     ]
    }
   ],
   "source": [
    "database_location = '/s/project/benchmark-lm/ssd-cache'\n",
    "#database_location = '/data/nasif12/home_if12/s_gjamn/gnomad_cache'\n",
    "\n",
    "#here I just used the data which is in our project, I hope it is the right one\n",
    "#Ok it is not the right one. Here I am not working with the right data\n",
    "#database_location = '/s/project/ml4rg_students/2025/project07/data'\n",
    "db = gnomAD_DB(database_location, gnomad_version=\"v4\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/s/project/benchmark-lm/ssd-cache/gnomad_db.sqlite3'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "db.db_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This I think it is just some start and finish to load the data \n",
    "\n",
    "interval_start = 1000000\n",
    "interval_end = 2000000\n",
    "\n",
    "#this I think it is for the data to be loaded \n",
    "df_region = db.get_info_for_interval(chrom=1, interval_start=interval_start, interval_end=interval_end, query=\"*\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You should only consider variants that pass their QC filters (e.g. `PASS` in the `filter` column).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>...</th>\n",
       "      <th>VarDP</th>\n",
       "      <th>AS_VQSLOD</th>\n",
       "      <th>AC_grpmax</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>1000001</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151566.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.9457</td>\n",
       "      <td>17.66670</td>\n",
       "      <td>...</td>\n",
       "      <td>105.0</td>\n",
       "      <td>-2.2368</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5168.0</td>\n",
       "      <td>0.000193</td>\n",
       "      <td>0.000193</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>1000002</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151266.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.9748</td>\n",
       "      <td>19.13270</td>\n",
       "      <td>...</td>\n",
       "      <td>226.0</td>\n",
       "      <td>-1.0867</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5132.0</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.000195</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>1000002</td>\n",
       "      <td>T</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151156.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.9748</td>\n",
       "      <td>19.13270</td>\n",
       "      <td>...</td>\n",
       "      <td>226.0</td>\n",
       "      <td>3.7004</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67744.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>1000006</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>5.0</td>\n",
       "      <td>151614.0</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>60.0</td>\n",
       "      <td>12.45360</td>\n",
       "      <td>...</td>\n",
       "      <td>549.0</td>\n",
       "      <td>5.1104</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4832.0</td>\n",
       "      <td>0.001035</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>1000006</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151506.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>12.45360</td>\n",
       "      <td>...</td>\n",
       "      <td>549.0</td>\n",
       "      <td>5.0193</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4836.0</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377158</th>\n",
       "      <td>1</td>\n",
       "      <td>1999992</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152006.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>55.0771</td>\n",
       "      <td>5.35882</td>\n",
       "      <td>...</td>\n",
       "      <td>170.0</td>\n",
       "      <td>-2.4849</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377159</th>\n",
       "      <td>1</td>\n",
       "      <td>1999993</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>294.0</td>\n",
       "      <td>152136.0</td>\n",
       "      <td>0.001932</td>\n",
       "      <td>59.0521</td>\n",
       "      <td>14.05260</td>\n",
       "      <td>...</td>\n",
       "      <td>16079.0</td>\n",
       "      <td>0.0339</td>\n",
       "      <td>279.0</td>\n",
       "      <td>41510.0</td>\n",
       "      <td>0.006721</td>\n",
       "      <td>0.000581</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.006721</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377160</th>\n",
       "      <td>1</td>\n",
       "      <td>1999995</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152054.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.5027</td>\n",
       "      <td>15.54300</td>\n",
       "      <td>...</td>\n",
       "      <td>256.0</td>\n",
       "      <td>0.6208</td>\n",
       "      <td>1.0</td>\n",
       "      <td>68020.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377161</th>\n",
       "      <td>1</td>\n",
       "      <td>1999997</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152060.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>58.7809</td>\n",
       "      <td>12.35000</td>\n",
       "      <td>...</td>\n",
       "      <td>20.0</td>\n",
       "      <td>-2.1178</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41396.0</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>377162</th>\n",
       "      <td>1</td>\n",
       "      <td>2000000</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152050.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.7382</td>\n",
       "      <td>11.88000</td>\n",
       "      <td>...</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.4127</td>\n",
       "      <td>1.0</td>\n",
       "      <td>68018.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>291938 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       chrom      pos ref alt filter     AC        AN        AF       MQ  \\\n",
       "0          1  1000001   G   T   PASS    1.0  151566.0  0.000007  59.9457   \n",
       "1          1  1000002   T   C   PASS    1.0  151266.0  0.000007  59.9748   \n",
       "2          1  1000002   T   G   PASS    1.0  151156.0  0.000007  59.9748   \n",
       "3          1  1000006   G   A   PASS    5.0  151614.0  0.000033     60.0   \n",
       "4          1  1000006   G   C   PASS    1.0  151506.0  0.000007     60.0   \n",
       "...      ...      ...  ..  ..    ...    ...       ...       ...      ...   \n",
       "377158     1  1999992   C   T   PASS    1.0  152006.0  0.000007  55.0771   \n",
       "377159     1  1999993   G   A   PASS  294.0  152136.0  0.001932  59.0521   \n",
       "377160     1  1999995   C   T   PASS    1.0  152054.0  0.000007  59.5027   \n",
       "377161     1  1999997   C   T   PASS    1.0  152060.0  0.000007  58.7809   \n",
       "377162     1  2000000   C   T   PASS    1.0  152050.0  0.000007  59.7382   \n",
       "\n",
       "              QD  ...    VarDP  AS_VQSLOD  AC_grpmax  AN_grpmax  AF_grpmax  \\\n",
       "0       17.66670  ...    105.0    -2.2368        1.0     5168.0   0.000193   \n",
       "1       19.13270  ...    226.0    -1.0867        1.0     5132.0   0.000195   \n",
       "2       19.13270  ...    226.0     3.7004        1.0    67744.0   0.000015   \n",
       "3       12.45360  ...    549.0     5.1104        5.0     4832.0   0.001035   \n",
       "4       12.45360  ...    549.0     5.0193        1.0     4836.0   0.000207   \n",
       "...          ...  ...      ...        ...        ...        ...        ...   \n",
       "377158   5.35882  ...    170.0    -2.4849        NaN        NaN        NaN   \n",
       "377159  14.05260  ...  16079.0     0.0339      279.0    41510.0   0.006721   \n",
       "377160  15.54300  ...    256.0     0.6208        1.0    68020.0   0.000015   \n",
       "377161  12.35000  ...     20.0    -2.1178        1.0    41396.0   0.000024   \n",
       "377162  11.88000  ...     25.0     0.4127        1.0    68018.0   0.000015   \n",
       "\n",
       "          AF_eas    AF_nfe  AF_fin    AF_afr  AF_asj  \n",
       "0       0.000193  0.000000     0.0  0.000000     0.0  \n",
       "1       0.000195  0.000000     0.0  0.000000     0.0  \n",
       "2       0.000000  0.000015     0.0  0.000000     0.0  \n",
       "3       0.000000  0.000000     0.0  0.000000     0.0  \n",
       "4       0.000000  0.000000     0.0  0.000000     0.0  \n",
       "...          ...       ...     ...       ...     ...  \n",
       "377158  0.000000  0.000000     0.0  0.000000     0.0  \n",
       "377159  0.000581  0.000015     0.0  0.006721     0.0  \n",
       "377160  0.000000  0.000015     0.0  0.000000     0.0  \n",
       "377161  0.000000  0.000000     0.0  0.000024     0.0  \n",
       "377162  0.000000  0.000015     0.0  0.000000     0.0  \n",
       "\n",
       "[291938 rows x 21 columns]"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#this is just to visualize how it should be the matrix representing the data\n",
    "\n",
    "df_region[df_region['filter'] == 'PASS'] "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "- chrom-Chromosome number, \n",
    "- pos-Genomic position, \n",
    "- ref-Reference nucleotide(s) at that position, \n",
    "- alt-Alternate nucleotide(s) observed in the population, \n",
    "- filter-Quality filter status, \n",
    "- AC-Allele Count – how many times the alternate allele was seen in the data, \n",
    "- AN-Allele Number – total number of alleles observed at this site,\n",
    "- AF-Allele Frequency – proportion of AC / AN"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(0.249397, 0.230222)"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interval_length = interval_end - interval_start\n",
    "df_region['len_ref'] = df_region['ref'].apply(len)\n",
    "df_region['len_alt'] = df_region['alt'].apply(len)\n",
    "df_region['is_snv'] = (df_region['len_ref'] == 1) & (df_region['len_alt'] == 1)\n",
    "\n",
    "#calculates variant densities—how frequently variants appear—within a given genomic interval\n",
    "df_region[df_region['filter'] == 'PASS']['pos'].unique().size/interval_length, df_region[(df_region['filter'] == 'PASS') & (df_region['is_snv'] == True)]['pos'].unique().size/interval_length\n",
    "\n",
    "#obv here everything is 0 because I am not really working with any data"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "On average we see a position with at least one variant in the population (regardless of its frequency) every 1/.23 bases ~ 4-5 bases"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "LDLR promoter region (region is upstream of the LDLR gene, which is involved in cholesterol metabolism)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>ReadPosRankSum</th>\n",
       "      <th>VarDP</th>\n",
       "      <th>AS_VQSLOD</th>\n",
       "      <th>AC_grpmax</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>11087549</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>150800.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.8265</td>\n",
       "      <td>17.19190</td>\n",
       "      <td>0.857</td>\n",
       "      <td>99.0</td>\n",
       "      <td>-0.6348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>11087550</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151718.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>17.17390</td>\n",
       "      <td>0.424</td>\n",
       "      <td>46.0</td>\n",
       "      <td>4.8335</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67936.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>11087554</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>149664.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.6771</td>\n",
       "      <td>12.86670</td>\n",
       "      <td>0.166</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.1907</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67386.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>11087561</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>AC0;AS_VQSR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>143510.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>59.9323</td>\n",
       "      <td>1.07377</td>\n",
       "      <td>0.756</td>\n",
       "      <td>122.0</td>\n",
       "      <td>-8.8454</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>11087564</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>148386.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.5248</td>\n",
       "      <td>8.64179</td>\n",
       "      <td>-0.351</td>\n",
       "      <td>67.0</td>\n",
       "      <td>-0.5145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40194.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>19</td>\n",
       "      <td>11089526</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3.0</td>\n",
       "      <td>152152.0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>60.0</td>\n",
       "      <td>13.09220</td>\n",
       "      <td>1.226</td>\n",
       "      <td>206.0</td>\n",
       "      <td>5.1780</td>\n",
       "      <td>3.0</td>\n",
       "      <td>68030.0</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>19</td>\n",
       "      <td>11089535</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152156.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.958</td>\n",
       "      <td>13.80260</td>\n",
       "      <td>0.693</td>\n",
       "      <td>152.0</td>\n",
       "      <td>7.6434</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41448.0</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>19</td>\n",
       "      <td>11089535</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3.0</td>\n",
       "      <td>152156.0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>59.958</td>\n",
       "      <td>13.80260</td>\n",
       "      <td>0.693</td>\n",
       "      <td>152.0</td>\n",
       "      <td>1.2823</td>\n",
       "      <td>3.0</td>\n",
       "      <td>41448.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>19</td>\n",
       "      <td>11089536</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>2.0</td>\n",
       "      <td>152166.0</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>60.0</td>\n",
       "      <td>15.21010</td>\n",
       "      <td>1.462</td>\n",
       "      <td>138.0</td>\n",
       "      <td>5.9164</td>\n",
       "      <td>2.0</td>\n",
       "      <td>68026.0</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>19</td>\n",
       "      <td>11089537</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152314.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>13.82500</td>\n",
       "      <td>-0.433</td>\n",
       "      <td>40.0</td>\n",
       "      <td>4.4322</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4824.0</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom       pos ref alt       filter   AC        AN        AF       MQ  \\\n",
       "0      19  11087549   C   T         PASS  1.0  150800.0  0.000007  59.8265   \n",
       "1      19  11087550   T   C         PASS  1.0  151718.0  0.000007     60.0   \n",
       "2      19  11087554   C   G         PASS  1.0  149664.0  0.000007  59.6771   \n",
       "3      19  11087561   A   T  AC0;AS_VQSR  0.0  143510.0  0.000000  59.9323   \n",
       "4      19  11087564   G   A         PASS  1.0  148386.0  0.000007  59.5248   \n",
       "..    ...       ...  ..  ..          ...  ...       ...       ...      ...   \n",
       "477    19  11089526   A   C         PASS  3.0  152152.0  0.000020     60.0   \n",
       "478    19  11089535   C   A         PASS  1.0  152156.0  0.000007   59.958   \n",
       "479    19  11089535   C  CA         PASS  3.0  152156.0  0.000020   59.958   \n",
       "480    19  11089536   A   G         PASS  2.0  152166.0  0.000013     60.0   \n",
       "481    19  11089537   G   A         PASS  1.0  152314.0  0.000007     60.0   \n",
       "\n",
       "           QD  ReadPosRankSum  VarDP  AS_VQSLOD  AC_grpmax  AN_grpmax  \\\n",
       "0    17.19190           0.857   99.0    -0.6348        NaN        NaN   \n",
       "1    17.17390           0.424   46.0     4.8335        1.0    67936.0   \n",
       "2    12.86670           0.166   30.0     1.1907        1.0    67386.0   \n",
       "3     1.07377           0.756  122.0    -8.8454        NaN        NaN   \n",
       "4     8.64179          -0.351   67.0    -0.5145        1.0    40194.0   \n",
       "..        ...             ...    ...        ...        ...        ...   \n",
       "477  13.09220           1.226  206.0     5.1780        3.0    68030.0   \n",
       "478  13.80260           0.693  152.0     7.6434        1.0    41448.0   \n",
       "479  13.80260           0.693  152.0     1.2823        3.0    41448.0   \n",
       "480  15.21010           1.462  138.0     5.9164        2.0    68026.0   \n",
       "481  13.82500          -0.433   40.0     4.4322        1.0     4824.0   \n",
       "\n",
       "     AF_grpmax  AF_eas    AF_nfe  AF_fin    AF_afr  AF_asj  \n",
       "0          NaN     0.0  0.000000     0.0  0.000000     0.0  \n",
       "1     0.000015     0.0  0.000015     0.0  0.000000     0.0  \n",
       "2     0.000015     0.0  0.000015     0.0  0.000000     0.0  \n",
       "3          NaN     0.0  0.000000     0.0  0.000000     0.0  \n",
       "4     0.000025     0.0  0.000000     0.0  0.000025     0.0  \n",
       "..         ...     ...       ...     ...       ...     ...  \n",
       "477   0.000044     0.0  0.000044     0.0  0.000000     0.0  \n",
       "478   0.000024     0.0  0.000000     0.0  0.000024     0.0  \n",
       "479   0.000072     0.0  0.000000     0.0  0.000072     0.0  \n",
       "480   0.000029     0.0  0.000029     0.0  0.000000     0.0  \n",
       "481   0.000207     0.0  0.000000     0.0  0.000000     0.0  \n",
       "\n",
       "[482 rows x 21 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# pandas max displayed columns to 100\n",
    "pd.set_option('display.max_columns', 100)\n",
    "ldlr_upstream_region_interval = (11087549, 11089551)\n",
    "ldlr_chr = 19\n",
    "\n",
    "df_region = db.get_info_for_interval(chrom=ldlr_chr, interval_start=ldlr_upstream_region_interval[0], interval_end=ldlr_upstream_region_interval[1], query=\"*\")\n",
    "df_region"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AF -> allele frequency = AC/AN. You should only consider SNPs (substitution variants)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Fetch the sequence of a specific region"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This is only needed for the speciesLM model because GPN MSA already fetches the alignment and sequences from its dataloader."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087548</td>\n",
       "      <td>11089551</td>\n",
       "      <td>+</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start       End Strand\n",
       "0      chr19  11087548  11089551      +"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import pandas as pd\n",
    "\n",
    "regions_df = pd.DataFrame({'Chromosome': [f'chr{ldlr_chr}'], 'Start': [ldlr_upstream_region_interval[0]-1], 'End': [ldlr_upstream_region_interval[1]], 'Strand':['+']}) #ldlr_upstream_region_interval[0]-1 because in pyranges the coordinates are 0-based!\n",
    "regions_df\n",
    "#This regions_df is passed to a function that uses PyRanges to fetch the actual DNA sequence from a FASTA file for that region"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This codes takes a data frame with varinat information and fetches the sequence around the variant from a fasta file. It fetches a window with the same length as the context length of the model.\n",
    "#In some rare cases that window will surpass the boundaries of a crhomosome (when the variant is close to the end of a chromosome). In that case, the variant is discarded (for now).\n",
    "import pandas as pd\n",
    "import pyranges as pr\n",
    "import math \n",
    "from pathlib import Path\n",
    "import numpy as np\n",
    "from Bio import SeqIO\n",
    "\n",
    "#fasta_path = Path(\"/s/project/ml4rg_students/2025/project07/data/GRCh38.primary_assembly.genome.fa\")\n",
    "\n",
    "#path to have access to the data (when using Species LM)\n",
    "\n",
    "fasta_path = Path(\"/s/project/ml4rg_students/2025/project07/data/GRCh38.primary_assembly.genome.fa\")\n",
    "\n",
    "#Added comment-example of use: You’re working on a model that needs to see 2003 base pairs of DNA around a gene's promoter. \n",
    "# You already know where that promoter is on the genome (e.g., chromosome 19, position 11087549–11089551)\n",
    "def fetch_sequence_from_region_df(regions_df, model_window_size, fasta_path=fasta_path):\n",
    "    '''\n",
    "    Fetches DNA sequences from genomic regions using a reference FASTA file.\n",
    "    Input:\n",
    "        - regions_df (pd.DataFrame): A DataFrame containing genomic regions with at least\n",
    "                                     'Chromosome', 'Start', and 'End' columns.\n",
    "        - model_window_size (int): The expected length of each DNA sequence to extract.\n",
    "        - fasta_path (str): Path to the reference genome FASTA file from which sequences are extracted.\n",
    "\n",
    "    Output:\n",
    "        - seqs_df (pd.DataFrame): A new DataFrame including the original region information\n",
    "                                  plus the following additional columns:\n",
    "                                  * 'seq': the uppercase DNA sequence\n",
    "                                  * 'seq_len': the length of the sequence (should equal model_window_size)\n",
    "    '''\n",
    "    \n",
    "    #Added comment: Make a copy of the input DataFrame to avoid modifying the original\n",
    "    seqs_df = regions_df.copy()\n",
    "\n",
    "    seqs_df[\"index\"] = np.arange(\n",
    "            len(seqs_df)\n",
    "        )  # Pyranges changes the ordering of the rows when converting to PyRanges. we first create a new column with the original order\n",
    "\n",
    "    \n",
    "    #Added comment: Convert the pandas DataFrame to a PyRanges object, which is optimized for genomic interval operations\n",
    "    \n",
    "    # create a pyranges df from the pandas df\n",
    "    seqs_pr = pr.PyRanges(seqs_df)\n",
    "\n",
    "    #Added comment: Fetch DNA sequences for each region using the FASTA file at fasta_path\n",
    "    # The sequences are stored in the `.seq` attribute of the PyRanges object\n",
    "    seqs_pr.seq = pr.get_sequence(\n",
    "        seqs_pr, path=fasta_path\n",
    "    )  # retrieve sequenc based on prepared start and end columns\n",
    "\n",
    "    \n",
    "    seqs_df = seqs_pr.df.copy()  # convert to pandas df\n",
    "\n",
    "    seqs_df = (\n",
    "        seqs_df.sort_values(\"index\").reset_index(drop=True).copy()\n",
    "    )  # sort the dataframe back to the original order\n",
    "\n",
    "    #Added comment: Drop the temporary index column as it is no longer needed\n",
    "    seqs_df = seqs_df.drop(columns=[\"index\"])\n",
    "\n",
    "    #Added comment: Convert all sequences to uppercase (standard for DNA: A, T, C, G)\n",
    "    seqs_df[\"seq\"] = seqs_df[\"seq\"].apply(lambda x: x.upper())  # make seq uppercase\n",
    "    \n",
    "    seqs_df[\"seq_len\"] = seqs_df[\"seq\"].apply(lambda x: len(x))  # add a sequence length\n",
    "\n",
    "    # assert  that the sequence length is the same as the model window size\n",
    "    assert (\n",
    "        seqs_df[\"seq_len\"] == model_window_size\n",
    "    ).all(), f\"{seqs_df['seq_len']} is not {model_window_size}\"\n",
    "\n",
    "\n",
    "    return seqs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Strand</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087548</td>\n",
       "      <td>11089551</td>\n",
       "      <td>+</td>\n",
       "      <td>CTTCTCTAGGTGATTGCTAATTACTTCTTTTTTTTTTTTTTTTTTT...</td>\n",
       "      <td>2003</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome     Start       End Strand  \\\n",
       "0      chr19  11087548  11089551      +   \n",
       "\n",
       "                                                 seq  seq_len  \n",
       "0  CTTCTCTAGGTGATTGCTAATTACTTCTTTTTTTTTTTTTTTTTTT...     2003  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#For the speciesLM model the context length is 2003\n",
    "seqs_df = fetch_sequence_from_region_df(regions_df, model_window_size=2003)\n",
    "seqs_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# it is important to assert that the reference sequence you obtain matches the reference nucleotides from the gnomAD database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>11087549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>11087550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>11087551</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>11087552</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>11087553</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>G</td>\n",
       "      <td>11089547</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>C</td>\n",
       "      <td>11089548</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2000</th>\n",
       "      <td>A</td>\n",
       "      <td>11089549</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2001</th>\n",
       "      <td>T</td>\n",
       "      <td>11089550</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2002</th>\n",
       "      <td>G</td>\n",
       "      <td>11089551</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2003 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq       pos\n",
       "0      C  11087549\n",
       "1      T  11087550\n",
       "2      T  11087551\n",
       "3      C  11087552\n",
       "4      T  11087553\n",
       "...   ..       ...\n",
       "1998   G  11089547\n",
       "1999   C  11089548\n",
       "2000   A  11089549\n",
       "2001   T  11089550\n",
       "2002   G  11089551\n",
       "\n",
       "[2003 rows x 2 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seq_df = pd.DataFrame({'seq': list(seqs_df['seq'][0]), 'pos': ldlr_upstream_region_interval[0] + np.arange(2003)})\n",
    "ref_seq_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chrom</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>ReadPosRankSum</th>\n",
       "      <th>VarDP</th>\n",
       "      <th>AS_VQSLOD</th>\n",
       "      <th>AC_grpmax</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>19</td>\n",
       "      <td>11087549</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>150800.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.8265</td>\n",
       "      <td>17.19190</td>\n",
       "      <td>0.857</td>\n",
       "      <td>99.0</td>\n",
       "      <td>-0.6348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>19</td>\n",
       "      <td>11087550</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151718.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>17.17390</td>\n",
       "      <td>0.424</td>\n",
       "      <td>46.0</td>\n",
       "      <td>4.8335</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67936.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>19</td>\n",
       "      <td>11087554</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>149664.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.6771</td>\n",
       "      <td>12.86670</td>\n",
       "      <td>0.166</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.1907</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67386.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>19</td>\n",
       "      <td>11087561</td>\n",
       "      <td>A</td>\n",
       "      <td>T</td>\n",
       "      <td>AC0;AS_VQSR</td>\n",
       "      <td>0.0</td>\n",
       "      <td>143510.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>59.9323</td>\n",
       "      <td>1.07377</td>\n",
       "      <td>0.756</td>\n",
       "      <td>122.0</td>\n",
       "      <td>-8.8454</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>19</td>\n",
       "      <td>11087564</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>148386.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.5248</td>\n",
       "      <td>8.64179</td>\n",
       "      <td>-0.351</td>\n",
       "      <td>67.0</td>\n",
       "      <td>-0.5145</td>\n",
       "      <td>1.0</td>\n",
       "      <td>40194.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000025</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>19</td>\n",
       "      <td>11089526</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3.0</td>\n",
       "      <td>152152.0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>60.0</td>\n",
       "      <td>13.09220</td>\n",
       "      <td>1.226</td>\n",
       "      <td>206.0</td>\n",
       "      <td>5.1780</td>\n",
       "      <td>3.0</td>\n",
       "      <td>68030.0</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000044</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>19</td>\n",
       "      <td>11089535</td>\n",
       "      <td>C</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152156.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.958</td>\n",
       "      <td>13.80260</td>\n",
       "      <td>0.693</td>\n",
       "      <td>152.0</td>\n",
       "      <td>7.6434</td>\n",
       "      <td>1.0</td>\n",
       "      <td>41448.0</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000024</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>19</td>\n",
       "      <td>11089535</td>\n",
       "      <td>C</td>\n",
       "      <td>CA</td>\n",
       "      <td>PASS</td>\n",
       "      <td>3.0</td>\n",
       "      <td>152156.0</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>59.958</td>\n",
       "      <td>13.80260</td>\n",
       "      <td>0.693</td>\n",
       "      <td>152.0</td>\n",
       "      <td>1.2823</td>\n",
       "      <td>3.0</td>\n",
       "      <td>41448.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000072</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>19</td>\n",
       "      <td>11089536</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>2.0</td>\n",
       "      <td>152166.0</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>60.0</td>\n",
       "      <td>15.21010</td>\n",
       "      <td>1.462</td>\n",
       "      <td>138.0</td>\n",
       "      <td>5.9164</td>\n",
       "      <td>2.0</td>\n",
       "      <td>68026.0</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000029</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>481</th>\n",
       "      <td>19</td>\n",
       "      <td>11089537</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152314.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>13.82500</td>\n",
       "      <td>-0.433</td>\n",
       "      <td>40.0</td>\n",
       "      <td>4.4322</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4824.0</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>482 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    chrom       pos ref alt       filter   AC        AN        AF       MQ  \\\n",
       "0      19  11087549   C   T         PASS  1.0  150800.0  0.000007  59.8265   \n",
       "1      19  11087550   T   C         PASS  1.0  151718.0  0.000007     60.0   \n",
       "2      19  11087554   C   G         PASS  1.0  149664.0  0.000007  59.6771   \n",
       "3      19  11087561   A   T  AC0;AS_VQSR  0.0  143510.0  0.000000  59.9323   \n",
       "4      19  11087564   G   A         PASS  1.0  148386.0  0.000007  59.5248   \n",
       "..    ...       ...  ..  ..          ...  ...       ...       ...      ...   \n",
       "477    19  11089526   A   C         PASS  3.0  152152.0  0.000020     60.0   \n",
       "478    19  11089535   C   A         PASS  1.0  152156.0  0.000007   59.958   \n",
       "479    19  11089535   C  CA         PASS  3.0  152156.0  0.000020   59.958   \n",
       "480    19  11089536   A   G         PASS  2.0  152166.0  0.000013     60.0   \n",
       "481    19  11089537   G   A         PASS  1.0  152314.0  0.000007     60.0   \n",
       "\n",
       "           QD  ReadPosRankSum  VarDP  AS_VQSLOD  AC_grpmax  AN_grpmax  \\\n",
       "0    17.19190           0.857   99.0    -0.6348        NaN        NaN   \n",
       "1    17.17390           0.424   46.0     4.8335        1.0    67936.0   \n",
       "2    12.86670           0.166   30.0     1.1907        1.0    67386.0   \n",
       "3     1.07377           0.756  122.0    -8.8454        NaN        NaN   \n",
       "4     8.64179          -0.351   67.0    -0.5145        1.0    40194.0   \n",
       "..        ...             ...    ...        ...        ...        ...   \n",
       "477  13.09220           1.226  206.0     5.1780        3.0    68030.0   \n",
       "478  13.80260           0.693  152.0     7.6434        1.0    41448.0   \n",
       "479  13.80260           0.693  152.0     1.2823        3.0    41448.0   \n",
       "480  15.21010           1.462  138.0     5.9164        2.0    68026.0   \n",
       "481  13.82500          -0.433   40.0     4.4322        1.0     4824.0   \n",
       "\n",
       "     AF_grpmax  AF_eas    AF_nfe  AF_fin    AF_afr  AF_asj  \n",
       "0          NaN     0.0  0.000000     0.0  0.000000     0.0  \n",
       "1     0.000015     0.0  0.000015     0.0  0.000000     0.0  \n",
       "2     0.000015     0.0  0.000015     0.0  0.000000     0.0  \n",
       "3          NaN     0.0  0.000000     0.0  0.000000     0.0  \n",
       "4     0.000025     0.0  0.000000     0.0  0.000025     0.0  \n",
       "..         ...     ...       ...     ...       ...     ...  \n",
       "477   0.000044     0.0  0.000044     0.0  0.000000     0.0  \n",
       "478   0.000024     0.0  0.000000     0.0  0.000024     0.0  \n",
       "479   0.000072     0.0  0.000000     0.0  0.000072     0.0  \n",
       "480   0.000029     0.0  0.000029     0.0  0.000000     0.0  \n",
       "481   0.000207     0.0  0.000000     0.0  0.000000     0.0  \n",
       "\n",
       "[482 rows x 21 columns]"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_region"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is important to assert that the reference sequence you fetch is the same as the one used in the gnomad database. You might find some faster code to do this... but this works"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "assert len(df_region.merge(ref_seq_df, on='pos', how='inner')) == len(df_region), \"Some or all fetched reference sequence nucleotides not match the gnomAD database nucleotides\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Sometimes you might be interested in fetching the sequence around a variant position. The following code fetches the sequence around specific variants stored in a dataframe."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "#make chromosome Start End dict from the fasta file\n",
    "def make_chrom_start_end_dict(fasta_file):\n",
    "    '''\n",
    "    make chromosome Start End dict from the fasta file\n",
    "    Input:\n",
    "        - fasta_file: path to the file that contains the data \n",
    "\n",
    "    Output:\n",
    "        - chrom_sizes_dict: dictionary which contains the beginning and the end of every chromosome\n",
    "    '''\n",
    "\n",
    "    #Added comment: create the dictionary\n",
    "    chrom_sizes_dict = {}\n",
    "\n",
    "    #Added comment: for cicle over the data in the column fasta\n",
    "    for record in SeqIO.parse(fasta_file, \"fasta\"):\n",
    "\n",
    "        #Added comment: the key in the dictionary is the chromosome ID \n",
    "        chrom = record.id\n",
    "\n",
    "        #Added comment: retrieve the size of the chromosome\n",
    "        chrom_size = len(record.seq)\n",
    "\n",
    "        #Added comment: add to the dictionary the chromosome size\n",
    "        chrom_sizes_dict[chrom] = chrom_size\n",
    "\n",
    "    #if os.path.exists(temp_unzipped_fasta_path):\n",
    "    #    os.remove(temp_unzipped_fasta_path)\n",
    "\n",
    "    return chrom_sizes_dict\n",
    "\n",
    "chrom_sizes_dict = make_chrom_start_end_dict(fasta_path)\n",
    "\n",
    "#Added comment: converts the dictionary in a data frame \n",
    "chrom_sizes_df = pd.DataFrame(chrom_sizes_dict.items(), columns=['Chromosome', 'End'])\n",
    "\n",
    "#Added comment: it adds the start column and convert to 0-based the end\n",
    "chrom_sizes_df['Start'] = 0\n",
    "chrom_sizes_df['End'] = chrom_sizes_df['End'] - 1 # convert to 0-based\n",
    "\n",
    "#Added comment: it filters the chromosomes by the standard ones keeping only the one which are \"significative\"\n",
    "standard_chromosomes = ['chr1', 'chr2', 'chr3', 'chr4', 'chr5', 'chr6', 'chr7', 'chr8',\n",
    "       'chr9', 'chr10', 'chr11', 'chr12', 'chr13', 'chr14', 'chr15',\n",
    "       'chr16', 'chr17', 'chr18', 'chr19', 'chr20', 'chr21', 'chr22',\n",
    "       'chrX', 'chrY', 'chrM']\n",
    "\n",
    "chrom_sizes_df = chrom_sizes_df[chrom_sizes_df['Chromosome'].isin(standard_chromosomes)].copy()\n",
    " \n",
    "\n",
    "def get_sequences_from_var_df(vars_df, fasta_path, model_window_size, assert_all_refs_in_fasta=False, pos_column=\"pos\"):\n",
    "    '''\n",
    "    From a data frame of genomics variantes the function extracts a DNA sequence centered on the variant from a \n",
    "    fasta file and gives back a new data frame containing the reference sequence, the muted sequence with the alternative \n",
    "    alleles, the validity controls on the basis and the corrects genomics coordinates\n",
    "    \n",
    "    Input:\n",
    "        -vars_df: data frame containing the variants infos (must have 'Chromosome', 'pos', 'ref', 'alt')\n",
    "        -fasta_path: path to the reference genome fasta file\n",
    "        -model_window_size: size of the sequence window to extract\n",
    "        -assert_all_refs_in_fasta: (optional) if True enforce that fasta and input reference alleles match\n",
    "        -pos_column: name of the column with variant positions\n",
    "        \n",
    "    Output:\n",
    "        -seqs_df: a new DataFrame with:\n",
    "            *all original variant info\n",
    "            *Start, End for window extraction\n",
    "            *seq: reference sequence\n",
    "            *seq_ALT: mutated sequence\n",
    "            *ref_nuc_fasta: base in FASTA at the variant position\n",
    "            *alt_nuc_seq: base in mutated sequence at the center (should match alt)\n",
    "    '''\n",
    "    \n",
    "    print(\"Fetching sequences..\")\n",
    "\n",
    "    #Added comment: create the new data frame from the input\n",
    "    seqs_df = vars_df.copy()\n",
    "\n",
    "    #Added comment: add a 'Strand' column (which is required by PyRanges) with default value +\n",
    "    if 'Strand' not in seqs_df.columns:\n",
    "        seqs_df['Strand'] = '+'\n",
    "\n",
    "    #Added comment: define start and end position for sequence extraction\n",
    "    seqs_df[\"index\"] = np.arange(\n",
    "        len(seqs_df)\n",
    "    )  # Pyranges changes the ordering of the rows when converting to PyRanges. we first create a new column with the original order\n",
    "    \n",
    "    #Added comment: since the new dataframe contains for every variant position a sequence centered in his index of lenght model_window_size\n",
    "    # then the start will be index - model_window_size / 2 and the end  will be index + model_window_size / 2\n",
    "    seqs_df[\"Start\"] = seqs_df[pos_column] - model_window_size // 2\n",
    "    seqs_df[\"End\"] = seqs_df[pos_column] + math.ceil(model_window_size / 2)\n",
    "\n",
    "    #Added comment: merge in chromosome sizes to know the bounds of each chromosome and\n",
    "    # keep only those sequences that stay within chromosome boundaries\n",
    "    seqs_w_chr_df = seqs_df.merge(\n",
    "    chrom_sizes_df, on='Chromosome', how='inner', suffixes=('', '_chrom')\n",
    "    ).copy()\n",
    "\n",
    "    \n",
    "    seqs_df = seqs_w_chr_df[(seqs_w_chr_df['Start'] >= 0) & (seqs_w_chr_df['End'] <= seqs_w_chr_df['End_chrom'])].copy()\n",
    "    print('Removed {} windows that were out of chromosome bounds.'.format(len(seqs_w_chr_df) - len(seqs_df)))\n",
    "    # create a pyranges df from the pandas df\n",
    "    seqs_pr = pr.PyRanges(seqs_df)\n",
    "    seqs_pr.seq = pr.get_sequence(\n",
    "        seqs_pr, path=fasta_path\n",
    "    )  # retrieve sequenc based on prepared start and end columns\n",
    "    seqs_df = seqs_pr.df.copy()  # convert to pandas df\n",
    "    seqs_df = (\n",
    "        seqs_df.sort_values(\"index\").reset_index(drop=True).copy()\n",
    "    )  # sort the dataframe back to the original order\n",
    "    seqs_df = seqs_df.drop(columns=[\"index\"])\n",
    "\n",
    "    seqs_df[\"seq\"] = seqs_df[\"seq\"].apply(lambda x: x.upper())  # make seq uppercase\n",
    "    seqs_df[\"seq_len\"] = seqs_df[\"seq\"].apply(lambda x: len(x))  # add a sequence length\n",
    "\n",
    "    # assert  that the sequence length is the same as the model window size\n",
    "    assert (\n",
    "        seqs_df[\"seq_len\"] == model_window_size\n",
    "    ).all(), f\"{seqs_df['seq_len']} is not {model_window_size}\"\n",
    "\n",
    "    seqs_df['ref_nuc_fasta'] = seqs_df['seq'].apply(lambda x: x[model_window_size//2])      # extract center nucleotide\n",
    "    #assert that the ref nucleotide in the data frame is the same as the one in the fetched sequence at the position of the variant\n",
    "\n",
    "    ref_match = seqs_df['ref_nuc_fasta'] == seqs_df['ref']\n",
    "    match_ratio = ref_match.sum() / len(ref_match)\n",
    "    print('Ratio of ref nucs in data table that are also in the fasta:', match_ratio)\n",
    "\n",
    "    if assert_all_refs_in_fasta:\n",
    "        if not ref_match.all():\n",
    "            mismatches = seqs_df.loc[~ref_match, ['Chromosome', pos_column, 'ref', 'ref_nuc_fasta']]\n",
    "            print(\"\\n[Warning] Mismatched reference nucleotides at these rows:\")\n",
    "            print(mismatches)\n",
    "            print(f\"Excluding {len(mismatches)} mismatched variants and proceeding...\\n\")\n",
    "            \n",
    "            # Filter out mismatches\n",
    "            seqs_df = seqs_df[ref_match].reset_index(drop=True)\n",
    "\n",
    "            #raise AssertionError(\"Some ref nucleotides in the variant table do not match those in the FASTA.\")\n",
    "\n",
    "    # introduce mutation and save as alternative sequence\n",
    "    seqs_df[\"seq_ALT\"] = seqs_df.apply(\n",
    "        lambda row: row[\"seq\"][: model_window_size // 2]\n",
    "        + row[\"alt\"]\n",
    "        + row[\"seq\"][(model_window_size // 2 + 1) :],\n",
    "        axis=1,\n",
    "    )\n",
    "    seqs_df[\"seq_len_ALT\"] = seqs_df[\"seq_ALT\"].apply(lambda x: len(x))\n",
    "    assert (\n",
    "        seqs_df[\"seq_len_ALT\"] == model_window_size\n",
    "    ).all(), \"(seqs_df['seq_len_ALT'] == model_window_size).all()\"\n",
    "\n",
    "    # assert that the alt nucleotide is the same as the one in the sequence at the position of the variant\n",
    "    seqs_df[\"alt_nuc_seq\"] = seqs_df[\"seq_ALT\"].apply(\n",
    "        lambda x: x[model_window_size // 2]\n",
    "    )\n",
    "    assert (\n",
    "        seqs_df[\"alt_nuc_seq\"] == seqs_df[\"alt\"]\n",
    "    ).all(), \"(seqs_df['alt_nuc_seq'] == seqs_df['alt']).all()\"\n",
    "\n",
    "    return seqs_df   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>ReadPosRankSum</th>\n",
       "      <th>VarDP</th>\n",
       "      <th>AS_VQSLOD</th>\n",
       "      <th>AC_grpmax</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087548</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>150800.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.8265</td>\n",
       "      <td>17.1919</td>\n",
       "      <td>0.857</td>\n",
       "      <td>99.0</td>\n",
       "      <td>-0.6348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087549</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151718.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>17.1739</td>\n",
       "      <td>0.424</td>\n",
       "      <td>46.0</td>\n",
       "      <td>4.8335</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67936.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087553</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>149664.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.6771</td>\n",
       "      <td>12.8667</td>\n",
       "      <td>0.166</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.1907</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67386.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome       pos ref alt filter   AC        AN        AF       MQ  \\\n",
       "0      chr19  11087548   C   T   PASS  1.0  150800.0  0.000007  59.8265   \n",
       "1      chr19  11087549   T   C   PASS  1.0  151718.0  0.000007     60.0   \n",
       "2      chr19  11087553   C   G   PASS  1.0  149664.0  0.000007  59.6771   \n",
       "\n",
       "        QD  ReadPosRankSum  VarDP  AS_VQSLOD  AC_grpmax  AN_grpmax  AF_grpmax  \\\n",
       "0  17.1919           0.857   99.0    -0.6348        NaN        NaN        NaN   \n",
       "1  17.1739           0.424   46.0     4.8335        1.0    67936.0   0.000015   \n",
       "2  12.8667           0.166   30.0     1.1907        1.0    67386.0   0.000015   \n",
       "\n",
       "   AF_eas    AF_nfe  AF_fin  AF_afr  AF_asj  \n",
       "0     0.0  0.000000     0.0     0.0     0.0  \n",
       "1     0.0  0.000015     0.0     0.0     0.0  \n",
       "2     0.0  0.000015     0.0     0.0     0.0  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "toy_df = df_region.iloc[:3].copy()\n",
    "toy_df =toy_df.rename(columns={'chrom': 'Chromosome'})\n",
    "toy_df['pos'] = toy_df['pos'] - 1 # convert to 0-based coordinates\n",
    "toy_df['Chromosome'] = toy_df['Chromosome'].apply(lambda x: f'chr{x}' if not x.startswith('chr') else x)\n",
    "toy_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Fetching sequences..\n",
      "Removed 0 windows that were out of chromosome bounds.\n",
      "Ratio of ref nucs in data table that are also in the fasta: 1.0\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>pos</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>ReadPosRankSum</th>\n",
       "      <th>VarDP</th>\n",
       "      <th>AS_VQSLOD</th>\n",
       "      <th>AC_grpmax</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "      <th>Strand</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>End_chrom</th>\n",
       "      <th>Start_chrom</th>\n",
       "      <th>seq</th>\n",
       "      <th>seq_len</th>\n",
       "      <th>ref_nuc_fasta</th>\n",
       "      <th>seq_ALT</th>\n",
       "      <th>seq_len_ALT</th>\n",
       "      <th>alt_nuc_seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087548</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>150800.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.8265</td>\n",
       "      <td>17.1919</td>\n",
       "      <td>0.857</td>\n",
       "      <td>99.0</td>\n",
       "      <td>-0.6348</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>11086547</td>\n",
       "      <td>11088550</td>\n",
       "      <td>58617615</td>\n",
       "      <td>0</td>\n",
       "      <td>TCCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCAC...</td>\n",
       "      <td>2003</td>\n",
       "      <td>C</td>\n",
       "      <td>TCCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCAC...</td>\n",
       "      <td>2003</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087549</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151718.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>17.1739</td>\n",
       "      <td>0.424</td>\n",
       "      <td>46.0</td>\n",
       "      <td>4.8335</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67936.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>11086548</td>\n",
       "      <td>11088551</td>\n",
       "      <td>58617615</td>\n",
       "      <td>0</td>\n",
       "      <td>CCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACC...</td>\n",
       "      <td>2003</td>\n",
       "      <td>T</td>\n",
       "      <td>CCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACC...</td>\n",
       "      <td>2003</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chr19</td>\n",
       "      <td>11087553</td>\n",
       "      <td>C</td>\n",
       "      <td>G</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>149664.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.6771</td>\n",
       "      <td>12.8667</td>\n",
       "      <td>0.166</td>\n",
       "      <td>30.0</td>\n",
       "      <td>1.1907</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67386.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>+</td>\n",
       "      <td>11086552</td>\n",
       "      <td>11088555</td>\n",
       "      <td>58617615</td>\n",
       "      <td>0</td>\n",
       "      <td>CCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACCACGC...</td>\n",
       "      <td>2003</td>\n",
       "      <td>C</td>\n",
       "      <td>CCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACCACGC...</td>\n",
       "      <td>2003</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Chromosome       pos ref alt filter   AC        AN        AF       MQ  \\\n",
       "0      chr19  11087548   C   T   PASS  1.0  150800.0  0.000007  59.8265   \n",
       "1      chr19  11087549   T   C   PASS  1.0  151718.0  0.000007     60.0   \n",
       "2      chr19  11087553   C   G   PASS  1.0  149664.0  0.000007  59.6771   \n",
       "\n",
       "        QD  ReadPosRankSum  VarDP  AS_VQSLOD  AC_grpmax  AN_grpmax  AF_grpmax  \\\n",
       "0  17.1919           0.857   99.0    -0.6348        NaN        NaN        NaN   \n",
       "1  17.1739           0.424   46.0     4.8335        1.0    67936.0   0.000015   \n",
       "2  12.8667           0.166   30.0     1.1907        1.0    67386.0   0.000015   \n",
       "\n",
       "   AF_eas    AF_nfe  AF_fin  AF_afr  AF_asj Strand     Start       End  \\\n",
       "0     0.0  0.000000     0.0     0.0     0.0      +  11086547  11088550   \n",
       "1     0.0  0.000015     0.0     0.0     0.0      +  11086548  11088551   \n",
       "2     0.0  0.000015     0.0     0.0     0.0      +  11086552  11088555   \n",
       "\n",
       "   End_chrom  Start_chrom                                                seq  \\\n",
       "0   58617615            0  TCCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCAC...   \n",
       "1   58617615            0  CCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACC...   \n",
       "2   58617615            0  CCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACCACGC...   \n",
       "\n",
       "   seq_len ref_nuc_fasta                                            seq_ALT  \\\n",
       "0     2003             C  TCCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCAC...   \n",
       "1     2003             T  CCTGCCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACC...   \n",
       "2     2003             C  CCTCTGCCTCATGAGTAACTGAGACTACAGGCGCCCACCACCACGC...   \n",
       "\n",
       "   seq_len_ALT alt_nuc_seq  \n",
       "0         2003           T  \n",
       "1         2003           C  \n",
       "2         2003           G  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "toy_df = get_sequences_from_var_df(toy_df, model_window_size=2003, fasta_path=fasta_path, assert_all_refs_in_fasta=True)\n",
    "toy_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Species LM - get Reconstruction probabilities for a specific region - LDLR promoter example"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "SpeciesLM is a BERT model which inputs sequences tokenized into overlapping 6-mers, together with a beginning of sentence, species and end of sentence token. \n",
    "The model outputs a reconstruction probability for each 6-mer at a position but we can aggregate the probabilities for a specific nucleotide. Here you have all the code to run the model and get the reconstruction probabilities per nucleotide and position for a specific region."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "You are using a model of type rotarybert to instantiate a model of type bert. This is not supported for all configurations of models and can yield errors.\n"
     ]
    }
   ],
   "source": [
    "from transformers import AutoModelForMaskedLM, AutoTokenizer\n",
    "import torch\n",
    "import torch.nn as nn\n",
    "from torch.amp import autocast\n",
    "import tqdm\n",
    "\n",
    "# Load the model\n",
    "model_name = \"johahi/specieslm-metazoa-upstream-k6\"\n",
    "model = AutoModelForMaskedLM.from_pretrained(model_name, trust_remote_code=True)\n",
    "\n",
    "# Load the corresponding tokenizer\n",
    "tokenizer = AutoTokenizer.from_pretrained(model_name, trust_remote_code=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RotaryBertForMaskedLM(\n",
       "  (bert): RotaryBertModel(\n",
       "    (embeddings): BertEmbeddings(\n",
       "      (word_embeddings): Embedding(4608, 768, padding_idx=0)\n",
       "      (position_embeddings): Embedding(8192, 768)\n",
       "      (token_type_embeddings): Embedding(2, 768)\n",
       "      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "      (dropout): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (encoder): RotaryBertEncoder(\n",
       "      (layer): ModuleList(\n",
       "        (0-11): 12 x RotaryBertLayer(\n",
       "          (attention): RotaryBertAttention(\n",
       "            (self): RotaryBertSdpaSelfAttention(\n",
       "              (query): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (key): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (value): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (dropout): Dropout(p=0.1, inplace=False)\n",
       "            )\n",
       "            (output): BertSelfOutput(\n",
       "              (dense): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "              (dropout): Dropout(p=0.1, inplace=False)\n",
       "            )\n",
       "          )\n",
       "          (intermediate): BertIntermediate(\n",
       "            (dense): Linear(in_features=768, out_features=3072, bias=True)\n",
       "            (intermediate_act_fn): NewGELUActivation()\n",
       "          )\n",
       "          (output): BertOutput(\n",
       "            (dense): Linear(in_features=3072, out_features=768, bias=True)\n",
       "            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "            (dropout): Dropout(p=0.1, inplace=False)\n",
       "          )\n",
       "        )\n",
       "      )\n",
       "    )\n",
       "    (pooler): BertPooler(\n",
       "      (dense): Linear(in_features=768, out_features=768, bias=True)\n",
       "      (activation): Tanh()\n",
       "    )\n",
       "  )\n",
       "  (cls): BertOnlyMLMHead(\n",
       "    (predictions): BertLMPredictionHead(\n",
       "      (transform): BertPredictionHeadTransform(\n",
       "        (dense): Linear(in_features=768, out_features=768, bias=True)\n",
       "        (transform_act_fn): NewGELUActivation()\n",
       "        (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "      )\n",
       "      (decoder): Linear(in_features=768, out_features=4608, bias=True)\n",
       "    )\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Added comment: Sets the computing device to CPU.\n",
    "device = \"cpu\"\n",
    "#Added comment: Moves the entire model (weights, layers, etc.) to the chosen device.\n",
    "model.to(device)\n",
    "#Added comment: Puts the model into evaluation mode.\n",
    "model.eval()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "proxy_species = 'homo_sapiens'\n",
    "assert proxy_species in tokenizer.get_vocab()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'datasets'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[23], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mdatasets\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m Dataset\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m#Added comment: from hugging face (🤗), it converts a data frame in a dataframe easily usable\u001b[39;00m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mtransformers\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m DefaultDataCollator\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'datasets'"
     ]
    }
   ],
   "source": [
    "from datasets import Dataset\n",
    "#Added comment: from hugging face (🤗), it converts a data frame in a dataframe easily usable\n",
    "from transformers import DefaultDataCollator\n",
    "#Added comment: it tokenize data in batches with automatic padding \n",
    "\n",
    "def kmers(seq, k=6): #for codons, k = 6\n",
    "    # splits a sequence into non-overlapping k-mers\n",
    "    return [seq[i:i + k] for i in range(0, len(seq), k) if i + k <= len(seq)]\n",
    "\n",
    "def kmers_stride1(seq, k=6):\n",
    "    # splits a sequence into overlapping k-mers\n",
    "    return [seq[i:i + k] for i in range(0, len(seq)-k+1)]  \n",
    "\n",
    "def tok_func_species(x, proxy_species, seq_col):\n",
    "    #Added comment: Tokenization function -> prepend species and tokenize overlapping k-mers\n",
    "    # Create text string: \"species kmer1 kmer2 kmer3...\"\n",
    "    # Then tokenize it using the shared tokenizer\n",
    "    res = tokenizer(proxy_species + \" \" +  \" \".join(kmers_stride1(x[seq_col])))\n",
    "    return res\n",
    "\n",
    "def tok_func_species_rolling_masking(x, proxy_species):\n",
    "    #Added comment: tokenization for sequences that are already masked\n",
    "    # Prepend species and tokenize the pre-masked sequence column\n",
    "    res = tokenizer(proxy_species + x['seq_masked'])\n",
    " \n",
    "    return res\n",
    "\n",
    "def get_dataloader(dataset, seq_col, shuffle = True, bs = 4, \n",
    "                   has_label=True, provide_covariates=True, rolling_masking=False):\n",
    "    '''\n",
    "    Prepares a PyTorch DataLoader from a pandas DataFrame containing biological sequences.\n",
    "    It tokenizes sequences using overlapping k-mers, adds optional covariates and labels, \n",
    "    and wraps the data into batches ready for training or evaluation with HuggingFace Transformers.\n",
    "\n",
    "    Input:\n",
    "        - dataset (pandas.DataFrame): A DataFrame containing sequence data and optional features/labels.\n",
    "        - seq_col (str): The name of the column containing sequences.\n",
    "        - shuffle (bool): Whether to shuffle the data in the DataLoader (default: True).\n",
    "        - bs (int): Batch size for the DataLoader (default: 4).\n",
    "        - has_label (bool): Whether the dataset includes a target column 'log_hl' to use as label (default: True).\n",
    "        - provide_covariates (bool): Whether to include additional covariate features (default: True).\n",
    "        - rolling_masking (bool): Whether to use a pre-masked sequence column ('seq_masked') instead of generating k-mers (default: False).\n",
    "\n",
    "    Output:\n",
    "        - data_loader (torch.utils.data.DataLoader): A DataLoader containing tokenized, padded batches\n",
    "        (optionally with covariates and labels), ready for input into a HuggingFace model.    \n",
    "    '''\n",
    "    #tok_func = lambda x: tok_func_species(x, proxy_species, seq_col) \n",
    "    \n",
    "    # Set initial list of columns to keep\n",
    "    keys = [seq_col]\n",
    "    \n",
    "    # Optionally compute and include covariate features\n",
    "    if provide_covariates:\n",
    "        dataset[\"covariates\"] = [x.astype('float') for x in np.array(dataset[codon_vars[:-1] + signif_vars[:-4]])]\n",
    "        keys.append('covariates')\n",
    "\n",
    "    # If labels are provided, rename and include them\n",
    "    if has_label:\n",
    "        dataset = dataset.rename(columns = {'log_hl' : 'label'})\n",
    "        keys.append('label')\n",
    "        \n",
    "    # Convert the selected columns into a HuggingFace Dataset\n",
    "    ds = Dataset.from_pandas(dataset[keys]) \n",
    "\n",
    "    #tok_ds = ds.map(tok_func, batched=False,  num_proc=12)\n",
    "\n",
    "    # Apply tokenization to each example in the dataset\n",
    "    if not rolling_masking:\n",
    "        tok_ds = ds.map(\n",
    "            lambda x: tok_func_species(x, proxy_species=proxy_species, seq_col=seq_col),\n",
    "            batched=False,\n",
    "            num_proc=12\n",
    "            )\n",
    "\n",
    "    else:\n",
    "        ds.map(\n",
    "            lambda x: tok_func_species_rolling_masking(x, proxy_species=proxy_species),\n",
    "            batched=False,\n",
    "            num_proc=12\n",
    "            )\n",
    "    \n",
    "    # Remove the original raw sequence column after tokenization\n",
    "    rem_tok_ds = tok_ds.remove_columns(seq_col)\n",
    "    \n",
    "    # Use DefaultDataCollator to handle padding and batching on the fly\n",
    "    data_collator = DefaultDataCollator()\n",
    "\n",
    "    # Create a PyTorch DataLoader for iterating batches during training/evaluation\n",
    "    data_loader = torch.utils.data.DataLoader(\n",
    "        rem_tok_ds, \n",
    "        batch_size=bs, \n",
    "        num_workers = 4, \n",
    "        shuffle = shuffle, \n",
    "        collate_fn = data_collator\n",
    "        )\n",
    "    \n",
    "    return data_loader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "class ReconstructionModel(nn.Module):\n",
    "    \n",
    "    def __init__(self, lm, tokenizer, device=device, \n",
    "                 kmer_size = 6, \n",
    "                 left_special_tokens = 2,\n",
    "                 right_special_tokens = 1,\n",
    "                 only_predict_masked = False,\n",
    "                 require_grad=False):\n",
    "        super().__init__()\n",
    "\n",
    "        #store arguments\n",
    "        self.lm = lm                                            # language model (e.g., SpeciesLM)\n",
    "        self.tokenizer = tokenizer                              # tokenizer for k-mers\n",
    "        self.require_grad = require_grad                        # whether to track gradients\n",
    "        self.only_predict_masked = only_predict_masked          # predict only masked nucleotides\n",
    "        self.kmer_size = kmer_size                              # k-mer length (e.g., 6)\n",
    "        self.device = device                                    # computing device (CPU or GPU)\n",
    "        self.left_special_tokens = left_special_tokens          # tokens before sequence (e.g., species)\n",
    "        self.right_special_tokens = right_special_tokens        # tokens after sequence (e.g., [CLS])\n",
    "\n",
    "        # Create the filter mapping kmers to their final nucleotide\n",
    "        self.create_prb_filter()\n",
    "        \n",
    "        self.word_embeddings = None                             # to store word embeddings if needed\n",
    "        self.hook_dict = {}                                     # dictionary to store hooks (for gradient tracking)\n",
    "        \n",
    "    def set_grad_computation(self, switch):\n",
    "        # Enable or disable gradient computation\n",
    "        self.require_grad = switch\n",
    "        \n",
    "    def set_word_embedding_hook(self):\n",
    "        # Register a hook to capture the embeddings during forward pass\n",
    "        def getHook():\n",
    "            def hook(model, input, output):\n",
    "                output.retain_grad()\n",
    "                self.word_embeddings = output\n",
    "            return hook\n",
    "        self.hook_dict[\"words\"] = self.lm.bert.embeddings.word_embeddings.register_forward_hook(getHook())\n",
    "        \n",
    "    def create_prb_filter(self):\n",
    "        \"\"\"\n",
    "        Make a convolutional filter for each nt\n",
    "        The way this works:\n",
    "        Take the kmer ACGTGC which maps to token 739, its last nt is C\n",
    "        This would be the prediction for the masked nucleotide from this kmer, if the kmer is the first in a masked span\n",
    "        So the first row of column 739 searches for C, in other words filter_xyz = 1 for x = 0, y = 739, z = 2\n",
    "        Equally, the second row of column 739 searches for G etc...\n",
    "        \n",
    "        Builds a 3D filter (shape: [k, 4**k, 4]) that maps each k-mer token to the nucleotide it ends with.\n",
    "        This is used to convert k-mer probabilities to per-nucleotide predictions.\n",
    "        \"\"\"\n",
    "\n",
    "        # Get the vocabulary from the tokenizer (maps k-mer strings to token IDs)\n",
    "        vocab = tokenizer.get_vocab()\n",
    "\n",
    "        # Generate all possible k-mers using combinations of \"A\", \"C\", \"G\", \"T\"\n",
    "        # For k=6 → 4096 k-mers (4^6 combinations)\n",
    "        kmer_list = [\"\".join(x) for x in itertools.product(\"ACGT\",repeat=6)]\n",
    "\n",
    "        # Mapping from nucleotide character to index (A=0, C=1, G=2, T=3)\n",
    "        nt_mapping = {\"A\":0,\"C\":1,\"G\":2,\"T\":3}\n",
    "\n",
    "        # Initialize the filter: shape [k, vocab_size, 4]\n",
    "        #   - k: position in the k-mer (0 to k-1)\n",
    "        #   - vocab_size: number of k-mers (4^k)\n",
    "        #   - 4: number of nucleotides (A, C, G, T)\n",
    "        prb_filter = np.zeros((self.kmer_size, 4**self.kmer_size, 4))\n",
    "\n",
    "        # For each k-mer string, determine its token ID and its nucleotides\n",
    "        for kmer in kmer_list:\n",
    "            # Get the token index for the k-mer\n",
    "            # Assumes the tokenizer vocabulary has 5 special tokens at the beginning, so we subtract 5\n",
    "            token = vocab[kmer] - 5 # there are 5 special tokens\n",
    "\n",
    "            # For each position in the k-mer, update the filter\n",
    "            for idx, nt in enumerate(kmer):\n",
    "                # Map the nucleotide character to its index (0–3)\n",
    "                nt_idx = nt_mapping[nt]\n",
    "\n",
    "                # Fill the filter at:\n",
    "                # [position in k-mer (reversed), token index, nucleotide index] = 1\n",
    "                #\n",
    "                # NOTE: self.kmer_size - 1 - idx reverses the k-mer so that the\n",
    "                # last character is at position 0 (important for masking logic).\n",
    "                prb_filter[5-idx, token, nt_idx] = 1\n",
    "\n",
    "        # Convert to torch tensor and move to the correct device\n",
    "        prb_filter = torch.from_numpy(prb_filter)\n",
    "\n",
    "        # Convert the numpy filter array to a PyTorch tensor,\n",
    "        # cast it to float32 (required for neural network operations),\n",
    "        # and move it to the appropriate device (CPU or GPU)\n",
    "        self.prb_filter = prb_filter.to(self.device).float() # k, 4**k, 4\n",
    "    \n",
    "    @autocast(device)   # Enable mixed precision on the specified device (saves memory & speeds up on GPUs)\n",
    "    def forward(self, tokens):\n",
    "        # Enable or disable gradient computation depending on the model setting\n",
    "        grad_context = contextlib.nullcontext() if self.require_grad else torch.no_grad()\n",
    "\n",
    "        with grad_context:\n",
    "            # Run the forward pass of the language model (e.g., RotaryBERT)\n",
    "            # Output: predictions is a dict with \"logits\" of shape (B, L_in, vocab_size)\n",
    "            predictions = self.lm(tokens)[\"logits\"] # Dim: B, L_in, vocab_size\n",
    "\n",
    "            # Remove logits for the first 5 special tokens → only keep k-mer logits\n",
    "            logits = predictions[:,:,5:(5+self.prb_filter.shape[1])] # remove any non k-mer dims (there are 5 special tokens)\n",
    "\n",
    "            # Apply softmax to get probability distribution over k-mers\n",
    "            kmer_preds = torch.softmax(logits,dim=2)\n",
    "\n",
    "            # remove special tokens:\n",
    "            # Remove the left/right special tokens (e.g., species prefix, [CLS]/[SEP] etc.)\n",
    "            kmer_preds = kmer_preds[:,(self.left_special_tokens):(kmer_preds.shape[1] - self.right_special_tokens),:]\n",
    "\n",
    "            # pad to predict first k-1 and last k-1 nt\n",
    "            # Pad with zeros on both sides so we can extract valid k-mer windows at the ends\n",
    "            kmer_pad = torch.zeros((kmer_preds.shape[0], (self.kmer_size-1), kmer_preds.shape[2]),device=self.device)\n",
    "\n",
    "            # Concatenate padding before and after the sequence\n",
    "            kmer_preds = torch.concat([kmer_pad,kmer_preds,kmer_pad],axis=1)\n",
    "\n",
    "            # reshape so that each span (representing one nucleotide) is its own entry\n",
    "            # Create sliding windows over the sequence to simulate overlapping k-mers\n",
    "            # Resulting shape: (B, L_seq, k, vocab_k)\n",
    "            kmer_preds = kmer_preds.unfold(dimension=1,size=self.kmer_size,step=1).swapaxes(2,3) # B, L_seq, k, 4**k\n",
    "\n",
    "            # convert kmer predictions to nucleotide predictions\n",
    "            # Add a dimension for nucleotides and broadcast to match prb_filter shape\n",
    "            # Shape becomes (B, L_seq, vocab_k, k, 4)\n",
    "            nt_preds = kmer_preds.unsqueeze(-1).expand((kmer_preds.shape[0],kmer_preds.shape[1],kmer_preds.shape[2],kmer_preds.shape[3],4)) # B, L_seq, k, 4**k, 4\n",
    "\n",
    "            # Multiply with prebuilt filter → sum over vocab & k-mer positions to get nucleotide predictions\n",
    "            # Final shape: (B, L_seq, 4), where 4 = [A, C, G, T]\n",
    "            nt_preds = (nt_preds * self.prb_filter).sum(axis=(2,3)) # B, L_seq, 4\n",
    "\n",
    "            # renormalize so that it sums to one\n",
    "            # Normalize by number of overlapping k-mers (i.e., k)\n",
    "            nt_prbs = nt_preds/self.kmer_size\n",
    "\n",
    "            # If we want to predict only masked positions (used for masked LM evaluation)\n",
    "            if self.only_predict_masked: # only record predictions for properly masked spans\n",
    "                # find the properly masked spans\n",
    "                # remove special tokens\n",
    "                tokens = tokens[:,self.left_special_tokens:(tokens.shape[1] - self.right_special_tokens)]\n",
    "                # pad with mask token (to ensure edge cases become spans)\n",
    "                token_pad = torch.zeros((tokens.shape[0],self.kmer_size-1), dtype=torch.int64, device=self.device) + 4\n",
    "                tokens = torch.concat([token_pad,tokens,token_pad],axis=1)\n",
    "                # unfold to B, L_seq, k\n",
    "                tokens = tokens.unfold(dimension=1,size=self.kmer_size,step=1)\n",
    "                # find masked spans\n",
    "                masked_positions = ((tokens == 4).sum(axis=2) == self.kmer_size)\n",
    "                nt_prbs = nt_prbs * masked_positions.unsqueeze(-1) # mask\n",
    "        return nt_prbs\n",
    "    \n",
    "    def mask_tokens(self, tokens):\n",
    "        # create diagonal identity matrix of the same shape as basis for the masking\n",
    "        diag_matrix = torch.eye(tokens.shape[1]).numpy()\n",
    "        # propagate the ones to kmer-sized spans\n",
    "\n",
    "        # For each row, apply a 1D convolution with a window of size `kmer_size`\n",
    "        # This simulates a mask that spans `kmer_size` consecutive positions\n",
    "        # Result: a binary matrix where each row has 1s in a k-mer span\n",
    "        masked_indices = np.apply_along_axis(\n",
    "            lambda m : np.convolve(m, [1] * self.kmer_size, mode = 'same' ),\n",
    "            axis = 1, \n",
    "            arr = diag_matrix\n",
    "            ).astype(bool)\n",
    "\n",
    "        # Convert the result back to a PyTorch tensor (boolean mask)\n",
    "        masked_indices = torch.from_numpy(masked_indices)\n",
    "\n",
    "        # do not mask special tokens and do not repeat at edges (this math works for k = 6, unclear if it generalizes)\n",
    "        # Remove positions corresponding to special tokens and boundary positions\n",
    "        # The slicing avoids applying masks too close to the sequence edges\n",
    "        masked_indices = masked_indices[\n",
    "            (self.kmer_size // 2 - 1) + self.left_special_tokens : \n",
    "            -((self.kmer_size // 2 ) + self.right_special_tokens)\n",
    "            ]\n",
    "\n",
    "        # Expand the token input so we can create one masked copy per row in the mask\n",
    "        # Shape becomes (batch_size, num_masks, sequence_length)\n",
    "        masked_tokens = tokens.unsqueeze(1).expand(\n",
    "            tokens.shape[0],\n",
    "            masked_indices.shape[0],\n",
    "            tokens.shape[1]\n",
    "            ).clone() # reshape\n",
    "\n",
    "        # Apply the mask: replace masked positions with token ID 4 (usually [MASK])\n",
    "        masked_tokens[:,masked_indices] = 4 # mask\n",
    "\n",
    "        # Return the masked versions of the tokens\n",
    "        return masked_tokens    # Shape: (B, num_masked_positions, L)\n",
    "    \n",
    "    def predict_all_from_dataloader(self, data_loader):\n",
    "\n",
    "        # Ensure the model is set to predict all positions, not just masked ones\n",
    "        assert not self.only_predict_masked\n",
    "\n",
    "        output_arrays = []\n",
    "\n",
    "        # Iterate through the data loader with a progress bar\n",
    "        for i, batch in tqdm.tqdm(enumerate(data_loader)):\n",
    "            # get some tokenized sequences (B, L_in)\n",
    "            # Get the input token IDs from the current batch\n",
    "            # Shape: (batch_size, sequence_length)\n",
    "            tokens = batch['input_ids']\n",
    "\n",
    "            # predict\n",
    "            # Pass tokens through the model to get nucleotide probability predictions\n",
    "            # Shape: (batch_size, sequence_length, 4) → for A, C, G, T\n",
    "            outputs = self(tokens.to(self.device)) # B, L_seq, 4\n",
    "\n",
    "            # Move the result from GPU to CPU to save memory on the device\n",
    "            output_arrays.append(outputs.cpu()) # send to cpu to conserve memory\n",
    "        \n",
    "        # rebuild to B, L_seq, 4\n",
    "        # Concatenate all batch results into one tensor\n",
    "        # Final shape: (total_sequences, sequence_length, 4)\n",
    "        predictions = torch.concat(output_arrays, axis=0)\n",
    "        return predictions\n",
    "    \n",
    "    def predict_all_from_dataloader_masked(self, \n",
    "                                          data_loader,\n",
    "                                          batch_size = 128):\n",
    "        all_preds = []  # To collect predictions from all groups in the DataLoader\n",
    "\n",
    "        for i, group in enumerate(data_loader):\n",
    "            output_arrays = []\n",
    "            # get some tokenized sequences (B, L_in)\n",
    "            tokens = group['input_ids']\n",
    "            # mask them\n",
    "            masked_tokens = self.mask_tokens(tokens)\n",
    "            # remember the number of sequences (shape[0] = B) and number of maskings (shape[1] = n_masked)\n",
    "            group_shape = masked_tokens.shape\n",
    "            # reshape to B*n_masked, L_in\n",
    "            masked_tokens = masked_tokens.reshape(masked_tokens.shape[0]*masked_tokens.shape[1], masked_tokens.shape[2])\n",
    "            token_loader = torch.utils.data.DataLoader(masked_tokens, batch_size=batch_size, \n",
    "                                                       num_workers = 4, shuffle = False, collate_fn = None)\n",
    "            # predict\n",
    "            for j, batch in enumerate(token_loader):\n",
    "                outputs = self(batch.to(self.device)) # b, L_seq, 4\n",
    "                output_arrays.append(outputs.cpu()) # send to cpu to conserve memory\n",
    "            # rebuild to B*n_masked, L_seq, 4\n",
    "            predictions = torch.concat(output_arrays, axis=0)\n",
    "            # reshape to B, n_masked, L_seq, 4\n",
    "            predictions = predictions.reshape((group_shape[0],group_shape[1],predictions.shape[1],predictions.shape[2]))\n",
    "            if self.only_predict_masked:\n",
    "                predictions = predictions.sum(axis=1)\n",
    "            all_preds.append(predictions)\n",
    "        predictions = torch.concat(all_preds, axis=0)\n",
    "        return predictions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Human LDLR 2kb 5' of the start codon"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'CTTCTCTAGGTGATTGCTAATTACTTCTTTTTTTTTTTTTTTTTTTGAGACGGAGTCTCACTCTGTCGCCCAGGCTGGAGTACAGTGGCGCAATCTCGTCTCACTGCAAACTCCACCTCCCGGGTTCACGCCATTCTCCTGCCTCAGCCTCCCGAGTAGCTGGGACTACAGGCACCCGCCACCACGCCCGGCTAATTTTTTGTATTTTTTTTTAGTAGAGGTGGGGTTTCACTGTGTTATCCAGGATGGTCTTGATCTCCTGACCTCGTGATCCACCTGCCTCAGCCTCCCAAAGTGCTGGGATTACAGGCGTGAGCCACCATGCCCAGCCCGCTAATTATTTCAATTTGACCTTGACACTGAGCCTGCCAAGTAGGTTCAAGCATTTTGATGGCCCCTTTACAGGTTGGGAAAGCTAATTTATCTGTCCAAGGCCGAATTCTGAAACTGAGTCTTAACTGCCAAAAATTCTTATCATCAATTTCTTCTTCTGGGTTGGGCACAGTGGCTCATGCCTGTAAAGCCAGCAATTTGAGAGGCATCATGATGCAAGAGGAAGAGGATTGAGTGAAGCTAGGAGTTTGGGACCAGCCTGGGCAACATAGTGAGACCCCATCTATAAAAAAAAATTAAAAATTAGTTGGGCATGGTGGTGCACTCCTGTGGTCCTAGCTATTCAGGAGGCTGAGGTGGGAGGATTCCTTGAGCCCAGGGTTGACGCTGCAGAGAGCTGTGATCACGCCACTGCAGTCCAGCCTGAGTGACAGCTGGAAATAATGATAAATAAATAATAAATAATTATTTAAAAAATTATAATAAAAATAATTAAAAAATTATTTTCCCTGATTAATCTTTTTTTTTGTCCTTCTGAGAGTTCAATTTGTCCCTTTTCTGCCTGGTCTCCTAGGTTTCCCTAAAATCCTGCTGAGAGGTTAGCACTGCCTGCCAAAGTCAGTTTGCAAAATCCCAGAGAAATCCAGCTTATTCCTGGGGGAACCGCCAAGACTGCCCAGCCCTGTGTGGGGTTCAGGCAAGTTTCTCACATGTGCCTTTTTGGCAAGAGGCCTCTGGCAACCCCATGAGTCCCCAAAGAGACTCAATTCTAAAAGTTGGTCTCCACCAGCTCTCTGTGGCTTAGGGGTTCAAGTTCAACTGTGAAAGCCCTGTTTTGTTTTGATTTTGCTTTGAGGGAGAGGAAACCGCCCTTCTGTTTGTTCAACTCCTTCTCCTAAGGGGAGAAATCAATATTTACGTCCAGACTCCAGGTATCCGTACAATTGATTTTTCAGATGTTTATACTCAGCCAAAGGCGGGATCCCACAAAACAAAAAATATTTTTTTGGCTGTACTTTTGTGAAGATTTTATTTAAATTCCTGATTGATCAGTGTCTATTAGGTGATTTGGAATAACAATGTAAAAACAATATACAACGAAAGGAAGCTAAAAATCTATACACAATTCCTAGAAAGGAAAAGGCAAATATAGAAAGTGGCGGAAGTTCCCAACATTTTTAGTGTTTTCCTTTTGAGGCAGAGAGGACAATGGCATTAGGCTATTGGAGGATCTTGAAAGGCTGTTGTTATCCTTCTGTGGACAACAACAGCAAAATGTTAACAGTTAAACATCGAGAAATTTCAGGAGGATCTTTCAGAAGATGCGTTTCCAATTTTGAGGGGGCGTCAGCTCTTCACCGGAGACCCAAATACAACAAATCAAGTCGCCTGCCCTGGCGACACTTTCGAAGGACTGGAGTGGGAATCAGAGCTTCACGGGTTAAAAAGCCGATGTCACATCGGCCGTTCGAAACTCCTCCTCTTGCAGTGAGGTGAAGACATTTGAAAATCACCCCACTGCAAACTCCTCCCCCTGCTAGAAACCTCACATTGAAATGCTGTAAATGACGTGGGCCCCGAGTGCAATCGCGGGAAGCCAGGGTTTCCAGCTAGGACACAGCAGGTCGTGATCCGGGTCGGGACACTGCCTGGCAGAGGCTGCGAGCATG'"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ldlr_five_prime_seq = seqs_df['seq'].values[0]\n",
    "ldlr_five_prime_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'get_dataloader' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[29], line 6\u001b[0m\n\u001b[1;32m      4\u001b[0m batch_size \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m10\u001b[39m\n\u001b[1;32m      5\u001b[0m dataset \u001b[38;5;241m=\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame({\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mseq\u001b[39m\u001b[38;5;124m'\u001b[39m: [ldlr_five_prime_seq]})\n\u001b[0;32m----> 6\u001b[0m dl \u001b[38;5;241m=\u001b[39m \u001b[43mget_dataloader\u001b[49m(dataset, shuffle \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mFalse\u001b[39;00m, bs \u001b[38;5;241m=\u001b[39m batch_size, seq_col \u001b[38;5;241m=\u001b[39m seq_col, has_label\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, provide_covariates\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m) \n\u001b[1;32m      8\u001b[0m reconstructor \u001b[38;5;241m=\u001b[39m ReconstructionModel(model, tokenizer, require_grad\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m, only_predict_masked\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[1;32m      9\u001b[0m snp_reconstruct \u001b[38;5;241m=\u001b[39m reconstructor\u001b[38;5;241m.\u001b[39mpredict_all_from_dataloader_masked(dl, batch_size\u001b[38;5;241m=\u001b[39mbatch_size)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'get_dataloader' is not defined"
     ]
    }
   ],
   "source": [
    "#import itertools  #We needed to import itertools to create kmer_list\n",
    "\n",
    "seq_col = 'seq'\n",
    "batch_size = 10\n",
    "dataset = pd.DataFrame({'seq': [ldlr_five_prime_seq]})\n",
    "dl = get_dataloader(dataset, shuffle = False, bs = batch_size, seq_col = seq_col, has_label=False, provide_covariates=False) \n",
    "\n",
    "reconstructor = ReconstructionModel(model, tokenizer, require_grad=False, only_predict_masked=True)\n",
    "snp_reconstruct = reconstructor.predict_all_from_dataloader_masked(dl, batch_size=batch_size)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snp_reconstruct' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[30], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43msnp_reconstruct\u001b[49m\u001b[38;5;241m.\u001b[39mshape\n",
      "\u001b[0;31mNameError\u001b[0m: name 'snp_reconstruct' is not defined"
     ]
    }
   ],
   "source": [
    "snp_reconstruct.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following contains functions to plot the reconstruction predicted by the LM (not relevant to understand this code)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "def ic_scale(pwm,background):\n",
    "    odds_ratio = ((pwm+0.001)/(1.004))/(background[None,:])\n",
    "    ic = ((np.log((pwm+0.001)/(1.004))/np.log(2))*pwm -\\\n",
    "            (np.log(background)*background/np.log(2))[None,:])\n",
    "    return pwm*(np.sum(ic,axis=1)[:,None])\n",
    "\n",
    "\n",
    "def plot_a(ax, base, left_edge, height, color):\n",
    "    a_polygon_coords = [\n",
    "        np.array([\n",
    "           [0.0, 0.0],\n",
    "           [0.5, 1.0],\n",
    "           [0.5, 0.8],\n",
    "           [0.2, 0.0],\n",
    "        ]),\n",
    "        np.array([\n",
    "           [1.0, 0.0],\n",
    "           [0.5, 1.0],\n",
    "           [0.5, 0.8],\n",
    "           [0.8, 0.0],\n",
    "        ]),\n",
    "        np.array([\n",
    "           [0.225, 0.45],\n",
    "           [0.775, 0.45],\n",
    "           [0.85, 0.3],\n",
    "           [0.15, 0.3],\n",
    "        ])\n",
    "    ]\n",
    "    for polygon_coords in a_polygon_coords:\n",
    "        ax.add_patch(\n",
    "            matplotlib.patches.Polygon(\n",
    "                (np.array([1,height])[None,:]*polygon_coords\n",
    "                 + np.array([left_edge,base])[None,:]),\n",
    "                facecolor=color, edgecolor=color))\n",
    "\n",
    "\n",
    "def plot_c(ax, base, left_edge, height, color):\n",
    "    ax.add_patch(matplotlib.patches.Ellipse(\n",
    "        xy=[left_edge+0.65, base+0.5*height], width=1.3, height=height,\n",
    "        facecolor=color, edgecolor=color))\n",
    "    ax.add_patch(matplotlib.patches.Ellipse(\n",
    "        xy=[left_edge+0.65, base+0.5*height], width=0.7*1.3, height=0.7*height,\n",
    "        facecolor='white', edgecolor='white'))\n",
    "    ax.add_patch(matplotlib.patches.Rectangle(\n",
    "        xy=[left_edge+1, base], width=1.0, height=height,\n",
    "        facecolor='white', edgecolor='white', fill=True))\n",
    "\n",
    "\n",
    "def plot_g(ax, base, left_edge, height, color):\n",
    "    ax.add_patch(matplotlib.patches.Ellipse(\n",
    "        xy=[left_edge+0.65, base+0.5*height], width=1.3, height=height,\n",
    "        facecolor=color, edgecolor=color))\n",
    "    ax.add_patch(matplotlib.patches.Ellipse(\n",
    "        xy=[left_edge+0.65, base+0.5*height], width=0.7*1.3, height=0.7*height,\n",
    "        facecolor='white', edgecolor='white'))\n",
    "    ax.add_patch(matplotlib.patches.Rectangle(\n",
    "        xy=[left_edge+1, base], width=1.0, height=height,\n",
    "        facecolor='white', edgecolor='white', fill=True))\n",
    "    ax.add_patch(matplotlib.patches.Rectangle(\n",
    "        xy=[left_edge+0.825, base+0.085*height],\n",
    "        width=0.174, height=0.415*height,\n",
    "        facecolor=color, edgecolor=color, fill=True))\n",
    "    ax.add_patch(matplotlib.patches.Rectangle(\n",
    "        xy=[left_edge+0.625, base+0.35*height],\n",
    "        width=0.374, height=0.15*height,\n",
    "        facecolor=color, edgecolor=color, fill=True))\n",
    "\n",
    "\n",
    "def plot_t(ax, base, left_edge, height, color):\n",
    "    ax.add_patch(\n",
    "        matplotlib.patches.Rectangle(xy=[left_edge+0.4, base],\n",
    "        width=0.2, height=height, facecolor=color, edgecolor=color, fill=True))\n",
    "    ax.add_patch(\n",
    "        matplotlib.patches.Rectangle(xy=[left_edge, base+0.8*height],\n",
    "        width=1.0, height=0.2*height, facecolor=color,\n",
    "        edgecolor=color, fill=True))\n",
    "\n",
    "    \n",
    "def plot_u(ax, base, left_edge, height, color):\n",
    "    ax.add_patch(matplotlib.patches.Ellipse(\n",
    "        xy=[left_edge+0.5, base+0.4*height], width=0.95, height=0.8*height,\n",
    "        facecolor=color, edgecolor=color))\n",
    "    ax.add_patch(matplotlib.patches.Rectangle(\n",
    "        xy=[left_edge+0.025, base+0.4*height], width=0.95, height=0.6*height,\n",
    "        facecolor=color, edgecolor=color, fill=True))\n",
    "    ax.add_patch(matplotlib.patches.Ellipse(\n",
    "        xy=[left_edge+0.5, base+0.4*height], width=0.6175, height=0.52*height,\n",
    "        facecolor='white', edgecolor='white'))\n",
    "    ax.add_patch(matplotlib.patches.Rectangle(\n",
    "        xy=[left_edge+0.19125, base+0.4*height], width=0.6175, height=0.6*height,\n",
    "        facecolor='white', edgecolor='white', fill=True))\n",
    "    \"\"\"ax.add_patch(matplotlib.patches.Ellipse(\n",
    "        xy=[left_edge+0.65, base+0.5*height], width=0.7*1.3, height=0.7*height,\n",
    "        facecolor='white', edgecolor='white'))\n",
    "    ax.add_patch(matplotlib.patches.Rectangle(\n",
    "        xy=[left_edge+1, base], width=1.0, height=height,\n",
    "        facecolor='white', edgecolor='white', fill=True))\"\"\"\n",
    "    \n",
    "\n",
    "default_colors = {0:'green', 1:'blue', 2:'orange', 3:'red'}\n",
    "dna_plot_funcs = {0:plot_a, 1:plot_c, 2:plot_g, 3:plot_t}\n",
    "rna_plot_funcs = {0:plot_a, 1:plot_c, 2:plot_g, 3:plot_u}\n",
    "\n",
    "\n",
    "def plot_weights_given_ax(ax, array,          \n",
    "                 height_padding_factor=0.2,\n",
    "                 length_padding=1.0,\n",
    "                 subticks_frequency=\"auto\",\n",
    "                 colors=default_colors,\n",
    "                 plot_funcs=dna_plot_funcs,\n",
    "                 highlight={},\n",
    "                 ylabel=\"\"):\n",
    "    if len(array.shape)==3:\n",
    "        array = np.squeeze(array)\n",
    "    assert len(array.shape)==2, array.shape\n",
    "    if (array.shape[0]==4 and array.shape[1] != 4):\n",
    "        array = array.transpose(1,0)\n",
    "    assert array.shape[1]==4\n",
    "    if (subticks_frequency==\"auto\"):\n",
    "        subticks_frequency = 1.0 if len(array) <= 40 else int(len(array)/40)\n",
    "    max_pos_height = 0.0\n",
    "    min_neg_height = 0.0\n",
    "    heights_at_positions = []\n",
    "    depths_at_positions = []\n",
    "    for i in range(array.shape[0]):\n",
    "        #sort from smallest to highest magnitude\n",
    "        acgt_vals = sorted(enumerate(array[i,:]), key=lambda x: abs(x[1]))\n",
    "        positive_height_so_far = 0.0\n",
    "        negative_height_so_far = 0.0\n",
    "        for letter in acgt_vals:\n",
    "            plot_func = plot_funcs[letter[0]]\n",
    "            color=colors[letter[0]]\n",
    "            if (letter[1] > 0):\n",
    "                height_so_far = positive_height_so_far\n",
    "                positive_height_so_far += letter[1]                \n",
    "            else:\n",
    "                height_so_far = negative_height_so_far\n",
    "                negative_height_so_far += letter[1]\n",
    "            plot_func(ax=ax, base=height_so_far,\n",
    "                      left_edge=i+0.5, height=letter[1], color=color)\n",
    "        max_pos_height = max(max_pos_height, positive_height_so_far)\n",
    "        min_neg_height = min(min_neg_height, negative_height_so_far)\n",
    "        heights_at_positions.append(positive_height_so_far)\n",
    "        depths_at_positions.append(negative_height_so_far)\n",
    "\n",
    "    #now highlight any desired positions; the key of\n",
    "    #the highlight dict should be the color\n",
    "    for color in highlight:\n",
    "        for start_pos, end_pos in highlight[color]:\n",
    "            assert start_pos >= 0.0 and end_pos <= array.shape[0]\n",
    "            min_depth = np.min(depths_at_positions[start_pos:end_pos])\n",
    "            max_height = np.max(heights_at_positions[start_pos:end_pos])\n",
    "            ax.add_patch(\n",
    "                matplotlib.patches.Rectangle(xy=[start_pos,min_depth],\n",
    "                    width=end_pos-start_pos,\n",
    "                    height=max_height-min_depth,\n",
    "                    edgecolor=color, fill=False))\n",
    "            \n",
    "    ax.set_xlim(0.5-length_padding, 0.5+array.shape[0]+length_padding)\n",
    "    ax.xaxis.set_ticks(np.arange(1.0, array.shape[0]+1, subticks_frequency))\n",
    "    height_padding = max(abs(min_neg_height)*(height_padding_factor),\n",
    "                         abs(max_pos_height)*(height_padding_factor))\n",
    "    ax.set_ylim(min_neg_height-height_padding, max_pos_height+height_padding)\n",
    "    ax.set_ylabel(ylabel)\n",
    "    ax.yaxis.label.set_fontsize(15)\n",
    "\n",
    "\n",
    "def plot_weights(array,\n",
    "                 figsize=(20,2),\n",
    "                 ax_transform_func=lambda x: x,\n",
    "                 **kwargs):\n",
    "    fig = plt.figure(figsize=figsize)\n",
    "    ax = fig.add_subplot(111) \n",
    "    ax_transform_func(ax)\n",
    "    plot_weights_given_ax(ax=ax,\n",
    "        array=array,\n",
    "        **kwargs)\n",
    "    plt.show()\n",
    "    \n",
    "    \n",
    "def dna_plot_weights(array, **kwargs):\n",
    "  plot_weights(array=array, plot_funcs=dna_plot_funcs, **kwargs)\n",
    "\n",
    "\n",
    "def rna_plot_weights(array, **kwargs):\n",
    "  plot_weights(array=array, plot_funcs=rna_plot_funcs, **kwargs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "#computes per position information content to scale the probabilities for better visualization\n",
    "\n",
    "def compute_per_position_ic(ppm, \n",
    "                            background=torch.tensor([0.295, 0.205, 0.205, 0.295]), #[0.295, 0.205, 0.205, 0.295]background nucleotide frequency in Human\n",
    "                            pseudocount=0):\n",
    "    ppm=ppm.unsqueeze(0)\n",
    "    alphabet_len = len(background)\n",
    "    bg = torch.log2(background)*background\n",
    "    pseudocounted_ppm = (ppm+pseudocount)/(1 + pseudocount*alphabet_len)\n",
    "    ic = torch.log2(pseudocounted_ppm)*ppm - bg.unsqueeze(0)\n",
    "    return torch.sum(ic,axis=2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snp_reconstruct' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[33], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m i\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m\n\u001b[0;32m----> 2\u001b[0m plot_weights((\u001b[43msnp_reconstruct\u001b[49m[i]\u001b[38;5;241m*\u001b[39mcompute_per_position_ic(snp_reconstruct[i])\u001b[38;5;241m.\u001b[39mswapaxes(\u001b[38;5;241m0\u001b[39m,\u001b[38;5;241m1\u001b[39m))\u001b[38;5;241m.\u001b[39mnumpy(), subticks_frequency\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m50\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'snp_reconstruct' is not defined"
     ]
    }
   ],
   "source": [
    "i=0\n",
    "plot_weights((snp_reconstruct[i]*compute_per_position_ic(snp_reconstruct[i]).swapaxes(0,1)).numpy(), subticks_frequency=50)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snp_reconstruct' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[34], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m plot_weights((\u001b[43msnp_reconstruct\u001b[49m[i]\u001b[38;5;241m*\u001b[39mcompute_per_position_ic(snp_reconstruct[i])\u001b[38;5;241m.\u001b[39mswapaxes(\u001b[38;5;241m0\u001b[39m,\u001b[38;5;241m1\u001b[39m))\u001b[38;5;241m.\u001b[39mnumpy()[\u001b[38;5;241m1764\u001b[39m:\u001b[38;5;241m1874\u001b[39m], subticks_frequency\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m50\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'snp_reconstruct' is not defined"
     ]
    }
   ],
   "source": [
    "plot_weights((snp_reconstruct[i]*compute_per_position_ic(snp_reconstruct[i]).swapaxes(0,1)).numpy()[1764:1874], subticks_frequency=50)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This reconstruction highlights 3 TF motifs in the promoter of the LDLR gene"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do the allele frequencies look like at this region?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>relative_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>11087549</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>11087550</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>11087551</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>11087552</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>11087553</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>G</td>\n",
       "      <td>11089547</td>\n",
       "      <td>1998</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>C</td>\n",
       "      <td>11089548</td>\n",
       "      <td>1999</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2000</th>\n",
       "      <td>A</td>\n",
       "      <td>11089549</td>\n",
       "      <td>2000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2001</th>\n",
       "      <td>T</td>\n",
       "      <td>11089550</td>\n",
       "      <td>2001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2002</th>\n",
       "      <td>G</td>\n",
       "      <td>11089551</td>\n",
       "      <td>2002</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2003 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq       pos  relative_pos\n",
       "0      C  11087549             0\n",
       "1      T  11087550             1\n",
       "2      T  11087551             2\n",
       "3      C  11087552             3\n",
       "4      T  11087553             4\n",
       "...   ..       ...           ...\n",
       "1998   G  11089547          1998\n",
       "1999   C  11089548          1999\n",
       "2000   A  11089549          2000\n",
       "2001   T  11089550          2001\n",
       "2002   G  11089551          2002\n",
       "\n",
       "[2003 rows x 3 columns]"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seq_df = pd.DataFrame({'seq': list(ldlr_five_prime_seq), 'pos': ldlr_upstream_region_interval[0] + np.arange(2003), 'relative_pos': np.arange(2003)})\n",
    "ref_seq_df\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_region['is_polymorfic_pos'] = True"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/scratch/tmp/s_aariz/ipykernel_2627180/309387599.py:2: FutureWarning: Downcasting object dtype arrays on .fillna, .ffill, .bfill is deprecated and will change in a future version. Call result.infer_objects(copy=False) instead. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`\n",
      "  ref_seq_w_afrequency_df.fillna(False, inplace=True)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>relative_pos</th>\n",
       "      <th>is_polymorfic_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>11087549</td>\n",
       "      <td>0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>11087550</td>\n",
       "      <td>1</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>11087551</td>\n",
       "      <td>2</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>11087552</td>\n",
       "      <td>3</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>11087553</td>\n",
       "      <td>4</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>G</td>\n",
       "      <td>11089547</td>\n",
       "      <td>1998</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>C</td>\n",
       "      <td>11089548</td>\n",
       "      <td>1999</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2000</th>\n",
       "      <td>A</td>\n",
       "      <td>11089549</td>\n",
       "      <td>2000</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2001</th>\n",
       "      <td>T</td>\n",
       "      <td>11089550</td>\n",
       "      <td>2001</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2002</th>\n",
       "      <td>G</td>\n",
       "      <td>11089551</td>\n",
       "      <td>2002</td>\n",
       "      <td>False</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2003 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq       pos  relative_pos  is_polymorfic_pos\n",
       "0      C  11087549             0               True\n",
       "1      T  11087550             1               True\n",
       "2      T  11087551             2              False\n",
       "3      C  11087552             3              False\n",
       "4      T  11087553             4              False\n",
       "...   ..       ...           ...                ...\n",
       "1998   G  11089547          1998              False\n",
       "1999   C  11089548          1999              False\n",
       "2000   A  11089549          2000              False\n",
       "2001   T  11089550          2001              False\n",
       "2002   G  11089551          2002              False\n",
       "\n",
       "[2003 rows x 4 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seq_w_afrequency_df = ref_seq_df.merge(df_region[['pos','is_polymorfic_pos']].drop_duplicates(), how='left')\n",
    "ref_seq_w_afrequency_df.fillna(False, inplace=True)\n",
    "ref_seq_w_afrequency_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The following plots whether a position has a variant in at least one indivdual out of the 76k"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f0e1c3e2710>"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABkoAAADFCAYAAAAbg64/AAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAOgJJREFUeJzt3Xl4lNXdxvF7EkhC0IQ1gUDCLriwyJaCVm2JSShl0VZxBVGxUGmhaRVoBXxLWxY3LFBBCghqAW0VtNQgRsFXG0F2tYissgYQTAJBE5r83j/yZmSyzxBmyXw/1zVX4Mk5M/d5njNzMvObecZhZiYAAAAAAAAAAIAgFOLrAAAAAAAAAAAAAL5CoQQAAAAAAAAAAAQtCiUAAAAAAAAAACBoUSgBAAAAAAAAAABBi0IJAAAAAAAAAAAIWhRKAAAAAAAAAABA0KJQAgAAAAAAAAAAglYdXweoCUVFRTp69Kguv/xyORwOX8cBAAAAAAAAAAA+ZGY6c+aM4uLiFBJS+WdGakWh5OjRo4qPj/d1DAAAAAAAAAAA4EcOHTqkli1bVtqmVhRKLr/8cknFA46KivJxGgAAAAAAAAAA4Eu5ubmKj4931g8qUysKJSWn24qKiqJQAgAAAAAAAAAAJKlaX9fBl7kDAAAAAAAAAICg5Xah5P3339fAgQMVFxcnh8OhlStXVtln3bp16t69u8LDw9W+fXu98MILZdrMnTtXrVu3VkREhBITE7Vx40Z3o3nV4cPS9u3FP/Edb+8XjgO8hbkGb/F0rjFH/QfHwn/U9mNR28eH4MXcrlig7JtAyemp2j4+VIxjD7gKhvtEoIwxUHL6M7dPvZWXl6euXbvq/vvv16233lpl+/3792vAgAEaNWqUXn75ZWVkZOjBBx9U8+bNlZKSIklasWKF0tLSNG/ePCUmJmrWrFlKSUnRrl27FBMT4/6oLrG1a6UVK6ScHCk6Who6VLr5Zl+n8j1v7xeOA7yFuQZv8XSuMUf9B8fCf9T2Y1Hbx4fgxdyuWKDsm0DJ6anaPj5UjGMPuAqG+0SgjDFQcvo7tz9R0r9/f/3hD3/QLbfcUq328+bNU5s2bfTUU0/pyiuv1JgxY/TTn/5UzzzzjLPN008/rZEjR2rEiBG66qqrNG/ePEVGRmrRokXlXmd+fr5yc3NdLt5y+HDxxCsqkjp1Kv65YgXVOm/vF44DvIW5Bm/xdK4xR/0Hx8J/1PZjUdvHh+DF3K5YoOybQMnpqdo+PlSMYw+4Cob7RKCMMVByBoJL/h0lmZmZSkpKctmWkpKizMxMSVJBQYE2b97s0iYkJERJSUnONqVNmzZN0dHRzkt8fPylG0App04VV+datJBCQ4t/5uQUbw9m3t4vHAd4C3MN3uLpXGOO+g+Ohf+o7ceito8PwYu5XbFA2TeBktNTtX18qBjHHnAVDPeJQBljoOQMBJe8UJKVlaXY2FiXbbGxscrNzdU333yjr776SoWFheW2ycrKKvc6J06cqJycHOfl0KFDlyx/aY0bF3+E6cgRqbCw+Gd0dPH2YObt/cJxgLcw1+Atns415qj/4Fj4j9p+LGr7+BC8mNsVC5R9Eyg5PVXbx4eKcewBV8FwnwiUMQZKzkBwyQsll0J4eLiioqJcLt7SsmXxed5CQqTPPy/+OXRo8fZg5u39wnGAtzDX4C2ezjXmqP/gWPiP2n4savv4ELyY2xULlH0TKDk9VdvHh4px7AFXwXCfCJQxBkrOQOAwM/O4s8Oh119/XUOGDKmwzQ033KDu3btr1qxZzm2LFy/WuHHjlJOTo4KCAkVGRurvf/+7y/UMHz5c2dnZWrVqVZU5cnNzFR0drZycHK8VTQ4fLv4IU+PGTLwLeXu/cBzgLcw1eIunc4056j84Fv6jth+L2j4+BC/mdsUCZd8ESk5P1fbxoWIce8BVMNwnAmWMgZLT29ypG1zyQsn48eP1r3/9S5988olz21133aXTp08rPT1dkpSYmKjevXtr9uzZkqSioiIlJCRozJgxmjBhQpU5fFEoAQAAAAAAAAAA/smduoHbp946e/astm3bpm3btkmS9u/fr23btungwYOSir8/ZNiwYc72o0aN0r59+/Too4/q888/11/+8he98sor+tWvfuVsk5aWpgULFmjJkiXauXOnRo8erby8PI0YMcLdeAAAAAAAAAAAANVWx90OmzZt0g9+8APn/9PS0iQVnyrrhRde0LFjx5xFE0lq06aNVq9erV/96ld69tln1bJlS/31r39VSkqKs83QoUN18uRJTZ48WVlZWerWrZvS09PLfME7AAAAAAAAAABATbqoU2/5C069BQAAAAAAAAAASlzSU28BAAAAAAAAAADUFhRKAAAAAAAAAABA0KJQAgAAAAAAAAAAghaFEgAAAAAAAAAAELQolAAAAAAAAAAAgKBFoQQAAAAAAAAAAAQtCiUAAAAAAAAAACBoUSgBAAAAAAAAAABBi0IJAAAAAAAAAAAIWhRKAAAAAAAAAABA0KJQAgAAAAAAAAAAghaFEgAAAAAAAAAAELQolAAAAAAAAAAAgKBFoQQAAAAAAAAAAAQtCiUAAAAAAAAAACBoUSgBAAAAAAAAAABBi0IJAAAAAAAAAAAIWhRKAAAAAAAAAABA0KJQAgAAAAAAAAAAghaFEgAAAAAAAAAAELQolAAAAAAAAAAAgKBFoQQAAAAAAAAAAAQtCiUAAAAAAAAAACBoUSgBAAAAAAAAAABBy6NCydy5c9W6dWtFREQoMTFRGzdurLDtTTfdJIfDUeYyYMAAZ5v77ruvzO9TU1M9iQYAAAAAAAAAAFBtddztsGLFCqWlpWnevHlKTEzUrFmzlJKSol27dikmJqZM+9dee00FBQXO/586dUpdu3bVbbfd5tIuNTVVixcvdv4/PDzc3WgAAAAAAAAAAABucbtQ8vTTT2vkyJEaMWKEJGnevHlavXq1Fi1apAkTJpRp36hRI5f/L1++XJGRkWUKJeHh4WrWrFm1MuTn5ys/P9/5/9zcXHeHAQAAAAAAAAAA4N6ptwoKCrR582YlJSV9dwUhIUpKSlJmZma1rmPhwoW64447VL9+fZft69atU0xMjDp27KjRo0fr1KlTFV7HtGnTFB0d7bzEx8e7MwwAAAAAAAAAAABJbhZKvvrqKxUWFio2NtZle2xsrLKysqrsv3HjRn366ad68MEHXbanpqZq6dKlysjI0IwZM7R+/Xr1799fhYWF5V7PxIkTlZOT47wcOnTInWEAAAAAAAAAAABI8uDUWxdj4cKF6ty5s3r37u2y/Y477nD+u3PnzurSpYvatWundevWqV+/fmWuJzw8nO8wAQAAAAAAAAAAF82tT5Q0adJEoaGhOn78uMv248ePV/n9Inl5eVq+fLkeeOCBKm+nbdu2atKkifbs2eNOPAAAAAAAAAAAALe4VSgJCwtTjx49lJGR4dxWVFSkjIwM9enTp9K+r776qvLz83XPPfdUeTuHDx/WqVOn1Lx5c3fiAQAAAAAAAAAAuMWtQokkpaWlacGCBVqyZIl27typ0aNHKy8vTyNGjJAkDRs2TBMnTizTb+HChRoyZIgaN27ssv3s2bN65JFH9NFHH+nAgQPKyMjQ4MGD1b59e6WkpHg4LAAAAAAAAAAAgKq5/R0lQ4cO1cmTJzV58mRlZWWpW7duSk9Pd37B+8GDBxUS4lp/2bVrlz744AO9/fbbZa4vNDRUO3bs0JIlS5Sdna24uDglJydr6tSpfA8JAAAAAAAAAAC4pBxmZr4OcbFyc3MVHR2tnJwcRUVF+ToOAAAAAAAAAADwIXfqBm6fegsAAAAAAAAAAKC2oFACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaHhVK5s6dq9atWysiIkKJiYnauHFjhW1feOEFORwOl0tERIRLGzPT5MmT1bx5c9WrV09JSUnavXu3J9EAAAAAAAAAAACqze1CyYoVK5SWlqYpU6Zoy5Yt6tq1q1JSUnTixIkK+0RFRenYsWPOy5dffuny+5kzZ+rPf/6z5s2bpw0bNqh+/fpKSUnRt99+6/6IAAAAAAAAAAAAqsntQsnTTz+tkSNHasSIEbrqqqs0b948RUZGatGiRRX2cTgcatasmfMSGxvr/J2ZadasWXrsscc0ePBgdenSRUuXLtXRo0e1cuXKcq8vPz9fubm5LhcAAAAAAAAAAAB3uVUoKSgo0ObNm5WUlPTdFYSEKCkpSZmZmRX2O3v2rFq1aqX4+HgNHjxYn332mfN3+/fvV1ZWlst1RkdHKzExscLrnDZtmqKjo52X+Ph4d4YBAAAAAAAAAAAgyc1CyVdffaXCwkKXT4RIUmxsrLKyssrt07FjRy1atEirVq3SSy+9pKKiIvXt21eHDx+WJGc/d65z4sSJysnJcV4OHTrkzjAAAAAAAAAAAAAkSXUu9Q306dNHffr0cf6/b9++uvLKKzV//nxNnTrVo+sMDw9XeHh4TUUEAAAAAAAAAABByq1PlDRp0kShoaE6fvy4y/bjx4+rWbNm1bqOunXr6tprr9WePXskydnvYq4TAAAAAAAAAADAE24VSsLCwtSjRw9lZGQ4txUVFSkjI8PlUyOVKSws1CeffKLmzZtLktq0aaNmzZq5XGdubq42bNhQ7esEAAAAAAAAAADwhNun3kpLS9Pw4cPVs2dP9e7dW7NmzVJeXp5GjBghSRo2bJhatGihadOmSZJ+//vf63vf+57at2+v7OxsPfHEE/ryyy/14IMPSpIcDofGjRunP/zhD+rQoYPatGmjSZMmKS4uTkOGDKm5kQIAAAAAAAAAAJTidqFk6NChOnnypCZPnqysrCx169ZN6enpzi9jP3jwoEJCvvugytdff62RI0cqKytLDRs2VI8ePfTvf/9bV111lbPNo48+qry8PD300EPKzs7W9ddfr/T0dEVERNTAEAEAAAAAAAAAAMrnMDPzdYiLlZubq+joaOXk5CgqKsrXcQAAAAAAAAAAgA+5Uzdw6ztKAAAAAAAAAAAAahMKJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoOVRoWTu3Llq3bq1IiIilJiYqI0bN1bYdsGCBfr+97+vhg0bqmHDhkpKSirT/r777pPD4XC5pKamehINAAAAAAAAAACg2twulKxYsUJpaWmaMmWKtmzZoq5duyolJUUnTpwot/26det055136r333lNmZqbi4+OVnJysI0eOuLRLTU3VsWPHnJdly5Z5NiIAAAAAAAAAAIBqcpiZudMhMTFRvXr10pw5cyRJRUVFio+P1y9+8QtNmDChyv6FhYVq2LCh5syZo2HDhkkq/kRJdna2Vq5cWa0M+fn5ys/Pd/4/NzdX8fHxysnJUVRUlDvDAQAAAAAAAAAAtUxubq6io6OrVTdw6xMlBQUF2rx5s5KSkr67gpAQJSUlKTMzs1rXce7cOZ0/f16NGjVy2b5u3TrFxMSoY8eOGj16tE6dOlXhdUybNk3R0dHOS3x8vDvDAAAAAAAAAAAAkORmoeSrr75SYWGhYmNjXbbHxsYqKyurWtcxfvx4xcXFuRRbUlNTtXTpUmVkZGjGjBlav369+vfvr8LCwnKvY+LEicrJyXFeDh065M4wAAAAAAAAAAAAJEl1vHlj06dP1/Lly7Vu3TpFREQ4t99xxx3Of3fu3FldunRRu3bttG7dOvXr16/M9YSHhys8PNwrmQEAAAAAAAAAQO3l1idKmjRpotDQUB0/ftxl+/Hjx9WsWbNK+z755JOaPn263n77bXXp0qXStm3btlWTJk20Z88ed+IBAAAAAAAAAAC4xa1CSVhYmHr06KGMjAzntqKiImVkZKhPnz4V9ps5c6amTp2q9PR09ezZs8rbOXz4sE6dOqXmzZu7Ew8AAAAAAAAAAMAtbhVKJCktLU0LFizQkiVLtHPnTo0ePVp5eXkaMWKEJGnYsGGaOHGis/2MGTM0adIkLVq0SK1bt1ZWVpaysrJ09uxZSdLZs2f1yCOP6KOPPtKBAweUkZGhwYMHq3379kpJSamhYQIAAAAAAAAAAJTl9neUDB06VCdPntTkyZOVlZWlbt26KT093fkF7wcPHlRIyHf1l+eee04FBQX66U9/6nI9U6ZM0eOPP67Q0FDt2LFDS5YsUXZ2tuLi4pScnKypU6fyPSQAAAAAAAAAAOCScpiZ+TrExcrNzVV0dLRycnIUFRXl6zgAAAAAAAAAAMCH3KkbuH3qLQAAAAAAAAAAgNqCQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGhRKAEAAAAAAAAAAEGLQgkAAAAAAAAAAAhaFEoAAAAAAAAAAEDQolACAAAAAAAAAACCFoUSAAAAAAAAAAAQtCiUAAAAAAAAAACAoEWhBAAAAAAAAAAABC0KJQAAAAAAAAAAIGh5VCiZO3euWrdurYiICCUmJmrjxo2Vtn/11VfVqVMnRUREqHPnzvrXv/7l8nsz0+TJk9W8eXPVq1dPSUlJ2r17tyfR3Hb4sLR9e/FPf769QOnnqUC5vYvJGSjHkH41i7lGP2/NNU8x12q+n6cCZXyB9LjmqUDaN54IlPEF0lyjX83281Qgja+27xtPBdIx9EQgzTX61axAGR9zjX7+fp+4mL6BMsbantPb/Y4cqX7bOu5dtbRixQqlpaVp3rx5SkxM1KxZs5SSkqJdu3YpJiamTPt///vfuvPOOzVt2jT9+Mc/1t/+9jcNGTJEW7Zs0TXXXCNJmjlzpv785z9ryZIlatOmjSZNmqSUlBT95z//UUREhLsRq23tWmnFCiknR4qOloYOlW6++ZLdnMe3Fyj9PBUot3cxOQPlGNKvZjHX6OetueYp5lrN9/NUoIwvkB7XPBVI+8YTgTK+QJpr9KvZfp4KpPHV9n0TKDlr+/h8cZu1vZ+nAmV8zDX6+ft94mL6BsoYa3tOX/R78cWq25Vw+xMlTz/9tEaOHKkRI0boqquu0rx58xQZGalFixaV2/7ZZ59VamqqHnnkEV155ZWaOnWqunfvrjlz5kgq/jTJrFmz9Nhjj2nw4MHq0qWLli5dqqNHj2rlypXlXmd+fr5yc3NdLu46fLh4BxcVSZ06Ff9cseLSVes8vb1A6eepQLm9i8kZKMeQfjWLuUY/b801TzHXar6fpwJlfIH0uOapQNo3ngiU8QXSXKNfzfbzVCCNr7bvm0DJWdvH54vbrO39PBUo42Ou0c/f7xMX0zdQxljbc/qyX3W5VSgpKCjQ5s2blZSU9N0VhIQoKSlJmZmZ5fbJzMx0aS9JKSkpzvb79+9XVlaWS5vo6GglJiZWeJ3Tpk1TdHS08xIfH+/OMCRJp04VV6FatJBCQ4t/5uQUb78UPL29QOnnqUC5vYvJGSjHkH41i7lGP2/NNU8x12q+n6cCZXyB9LjmqUDaN54IlPEF0lyjX83281Qgja+275tAyVnbx+eL26zt/TwVKONjrtHP3+8TF9M3UMZY23P6ql9cXOXtLuRWoeSrr75SYWGhYmNjXbbHxsYqKyur3D5ZWVmVti/56c51Tpw4UTk5Oc7LoUOH3BmGJKlx4+KP6hw5IhUWFv+Mji7efil4enuB0s9TgXJ7F5MzUI4h/WoWc41+3pprnmKu1Xw/TwXK+ALpcc1TgbRvPBEo4wukuUa/mu3nqUAaX23fN4GSs7aPzxe3Wdv7eSpQxsdco5+/3ycupm+gjLG25/RVv6NHK293IY++zN3XwsPDFRUV5XJxV8uWxeczCwmRPv+8+OfQocXbLwVPby9Q+nkqUG7vYnIGyjGkX81irtHPW3PNU8y1mu/nqUAZXyA9rnkqkPaNJwJlfIE01+hXs/08FUjjq+37JlBy1vbx+eI2a3s/TwXK+Jhr9PP3+8TF9A2UMdb2nL7sV10OM7PqNi4oKFBkZKT+/ve/a8iQIc7tw4cPV3Z2tlatWlWmT0JCgtLS0jRu3DjntilTpmjlypXavn279u3bp3bt2mnr1q3q1q2bs82NN96obt266dlnn60yV25urqKjo5WTk+N20eTw4eKP4jRufOkmXk3cXqD081Sg3N7F5AyUY0i/msVco5+35pqnmGs1389TgTK+QHpc81Qg7RtPBMr4Ammu0a9m+3kqkMZX2/eNpwLpGHoikOYa/WpWoIyPuUY/f79PXEzfQBljbc/p7X47d+bqqquqVzdwq1AiSYmJierdu7dmz54tSSoqKlJCQoLGjBmjCRMmlGk/dOhQnTt3Tm+++aZzW9++fdWlSxfNmzdPZqa4uDj95je/0a9//WtJxYWPmJgYvfDCC7rjjjuqzHQxhRIAAAAAAAAAAFC7uFM3qOPulaelpWn48OHq2bOnevfurVmzZikvL08jRoyQJA0bNkwtWrTQtGnTJEljx47VjTfeqKeeekoDBgzQ8uXLtWnTJj3//POSJIfDoXHjxukPf/iDOnTooDZt2mjSpEmKi4tz+dQKAAAAAAAAAABATXO7UDJ06FCdPHlSkydPVlZWlrp166b09HTnl7EfPHhQIRec/Ktv377629/+pscee0y//e1v1aFDB61cuVLXXHONs82jjz6qvLw8PfTQQ8rOztb111+v9PR0RUREVCtTyYdicnNz3R0OAAAAAAAAAACoZUrqBdU5qZbbp97yR4cPH1Z8fLyvYwAAAAAAAAAAAD9y6NAhtaziy01qRaGkqKhIR48e1eWXXy6Hw+FW39zcXMXHx+vQoUN+/f0m5KxZgZJTCpys5KxZgZJTCpys5Kx5gZKVnDUrUHJKgZOVnDUrUHJKgZOVnDUvULKSs2YFSk4pcLKSs2YFSk4pcLKSs+YFSlZy1ixPc5qZzpw5o7i4OJezYJXH7VNv+aOQkJAqK0JViYqK8uvJUIKcNStQckqBk5WcNStQckqBk5WcNS9QspKzZgVKTilwspKzZgVKTilwspKz5gVKVnLWrEDJKQVOVnLWrEDJKQVOVnLWvEDJSs6a5UnO6OjoarWrvIwCAAAAAAAAAABQi1EoAQAAAAAAAAAAQSvoCyXh4eGaMmWKwsPDfR2lUuSsWYGSUwqcrOSsWYGSUwqcrOSseYGSlZw1K1BySoGTlZw1K1BySoGTlZw1L1CykrNmBUpOKXCykrNmBUpOKXCykrPmBUpWctYsb+SsFV/mDgAAAAAAAAAA4Img/0QJAAAAAAAAAAAIXhRKAAAAAAAAAABA0KJQAgAAAAAAAAAAghaFEgAAAAAAAAAAELQolAAAAAAAAAAAgKAVtIWS999/XwMHDlRcXJwcDodWrlzp60hVmj59uhwOh8aNG+frKGW0bt1aDoejzOXhhx/2aa6qjrOZafLkyWrevLnq1aunpKQk7d692+9yPv744+rUqZPq16+vhg0bKikpSRs2bPB6zupklaSdO3dq0KBBio6OVv369dWrVy8dPHjQr3IeP35c9913n+Li4hQZGanU1FSfHPtp06apV69euvzyyxUTE6MhQ4Zo165dzt+fPn1av/jFL9SxY0fVq1dPCQkJ+uUvf6mcnBy/yilJN910U5nHgFGjRvldzqysLN17771q1qyZ6tevr+7du+sf//iHV3NK0nPPPacuXbooKipKUVFR6tOnj9566y3n759//nnddNNNioqKksPhUHZ2ttczVidnCTNT//79fbamVpbzwIED5a5RDodDr776qtezXqi8tf3bb7/Vww8/rMaNG+uyyy7TT37yEx0/ftx3IVV+zp/97Gdq166d6tWrp6ZNm2rw4MH6/PPPfRfy/1X091JmZqZ++MMfqn79+oqKitINN9ygb775xjchVX7OvXv36pZbblHTpk0VFRWl22+/3SfH/vHHHy9zX+nUqZMk/1mXqsop+ce6VJ2c/rIuSdKRI0d0zz33qHHjxqpXr546d+6sTZs2OX//2muvKTk5WY0bN5bD4dC2bdv8MueFRo0aJYfDoVmzZnk35P+rKmtF69MTTzzhtYxVPZfzl3Wpqpz+tC5V5/mxP6xLVeX0l3WpsLBQkyZNUps2bVSvXj21a9dOU6dOlZlJks6fP6/x48erc+fOql+/vuLi4jRs2DAdPXrUr3JK0n333Vdmf6empno1Z3Wznj17VmPGjFHLli1Vr149XXXVVZo3b57Xs545c0bjxo1Tq1atVK9ePfXt21cff/yx8/f+sjZVlfNCvlybqsrpq3WpJl6/O336tO6++25FRUWpQYMGeuCBB3T27Fm/yzlo0CAlJCQoIiJCzZs317333lvjj1c19Xro6tWrlZiYqHr16qlhw4YaMmRIjeasqaxbtmzRzTffrAYNGqhx48Z66KGHPDr2QVsoycvLU9euXTV37lxfR6mWjz/+WPPnz1eXLl18HaVcH3/8sY4dO+a8rF27VpJ02223+TRXVcd55syZ+vOf/6x58+Zpw4YNql+/vlJSUvTtt9/6Vc4rrrhCc+bM0SeffKIPPvhArVu3VnJysk6ePOnVnFLVWffu3avrr79enTp10rp167Rjxw5NmjRJERERfpPTzDRkyBDt27dPq1at0tatW9WqVSslJSUpLy/PqznXr1+vhx9+WB999JHWrl2r8+fPKzk52Znj6NGjOnr0qJ588kl9+umneuGFF5Senq4HHnjAr3KWGDlypMtjwcyZM/0u57Bhw7Rr1y698cYb+uSTT3Trrbfq9ttv19atW72atWXLlpo+fbo2b96sTZs26Yc//KEGDx6szz77TJJ07tw5paam6re//a1Xc7mbs8SsWbPkcDh8lLLynPHx8S7z8tixY/qf//kfXXbZZerfv7/PMle0tv/qV7/Sm2++qVdffVXr16/X0aNHdeutt/ooZcU5e/ToocWLF2vnzp1as2aNzEzJyckqLCz0UdKKs2ZmZio1NVXJycnauHGjPv74Y40ZM0YhIb75U7i8nHl5eUpOTpbD4dC7776rDz/8UAUFBRo4cKCKioq8nvHqq692uc988MEHkvxnXaoqZwlfr0vVyekv69LXX3+t6667TnXr1tVbb72l//znP3rqqafUsGFDZ5u8vDxdf/31mjFjhlezuZuzxOuvv66PPvpIcXFxPkhavayl16dFixbJ4XDoJz/5iddyVvVczl/Wpapy+tO6VFVWf1mXKsvpT+vSjBkz9Nxzz2nOnDnauXOnZsyYoZkzZ2r27NmSiv9u3rJliyZNmqQtW7botdde065duzRo0CC/ylkiNTXVZb8vW7bMqzmrmzUtLU3p6el66aWXtHPnTo0bN05jxozRG2+84dWsDz74oNauXasXX3xRn3zyiZKTk5WUlKQjR45I8o+1qTo5S/h6baoqp6/WpZp4/e7uu+/WZ599prVr1+qf//yn3n//fT300EN+l/MHP/iBXnnlFe3atUv/+Mc/tHfvXv30pz/1u5z/+Mc/dO+992rEiBHavn27PvzwQ9111101mrMmsh49elRJSUlq3769NmzYoPT0dH322We677773A9jMEn2+uuv+zpGhc6cOWMdOnSwtWvX2o033mhjx471daQqjR071tq1a2dFRUW+juJU+jgXFRVZs2bN7IknnnBuy87OtvDwcFu2bJkPEharznzMyckxSfbOO+94J1QFyss6dOhQu+eee3wTqAKlc+7atcsk2aeffurcVlhYaE2bNrUFCxb4IOF3Tpw4YZJs/fr1FbZ55ZVXLCwszM6fP+/FZK7Ky+mPj0/l5axfv74tXbrUpV2jRo18fuzNzBo2bGh//etfXba99957Jsm+/vpr34QqR+mcW7dutRYtWtixY8f8ak0tb3+W6Natm91///1eTvSditb27Oxsq1u3rr366qvOtjt37jRJlpmZ6Tc5y7N9+3aTZHv27PFewAtUljUxMdEee+wxn+QqraKca9assZCQEMvJyXG2zc7ONofDYWvXrvVqxilTpljXrl2r3d5X61JVOf1lXaoqp7+sS+PHj7frr7++Wm33799vkmzr1q2XNlQ5qpvz8OHD1qJFC/v000+tVatW9swzz1z6cKW4s09LDB482H74wx9eokTVc+FzOX9bly5U1XNOX69LFyqd1Z/WpQtdmNOf1qUBAwaU+bvt1ltvtbvvvrvCPhs3bjRJ9uWXX17qeE7VyTl8+HAbPHiw1zJVpDpZr776avv973/v0qZ79+72u9/9zisZzczOnTtnoaGh9s9//rPKHL5cm6qb09drkzv7s4Qv1iVPXr/7z3/+Y5Ls448/drZ56623zOFw2JEjR/wmZ3lWrVplDofDCgoK/Cbn+fPnrUWLFhU+n75UPMk6f/58i4mJscLCQmebHTt2mCTbvXu3W7cftJ8oCSQPP/ywBgwYoKSkJF9HqZaCggK99NJLuv/++3367uKq7N+/X1lZWS77NTo6WomJicrMzPRhssoVFBTo+eefV3R0tLp27errOC6Kioq0evVqXXHFFUpJSVFMTIwSExP97tR2+fn5kuTyKZeQkBCFh4eXeSeqt5WcuqRRo0aVtomKilKdOnW8FavcDFLZnC+//LKaNGmia665RhMnTtS5c+d8Ec+pvJx9+/bVihUrdPr0aRUVFWn58uX69ttvddNNN/koZfFH35cvX668vDz16dPHZzmqUl7Oc+fO6a677tLcuXPVrFkzHycsVtX+3Lx5s7Zt2+azd8BLFa/tmzdv1vnz5122d+rUSQkJCT5Zm6r7N0heXp4WL16sNm3aKD4+3kvpXFWU9cSJE9qwYYNiYmLUt29fxcbG6sYbb/TZ431FOfPz8+VwOBQeHu7cFhERoZCQEJ9k3b17t+Li4tS2bVvdfffdlZ5C05frUlU5/WVdqiynv6xLb7zxhnr27KnbbrtNMTExuvbaa7VgwQKvZqiO6uQsKirSvffeq0ceeURXX321j5K6v0+PHz+u1atX+3R9Kv1czt/WpYpyluYP61KJ0ln9bV2qKKc/rUt9+/ZVRkaGvvjiC0nS9u3b9cEHH1T6yeCcnBw5HA41aNDASymrn3PdunWKiYlRx44dNXr0aJ06dcprGd3J2rdvX73xxhs6cuSIzEzvvfeevvjiCyUnJ3st53//+18VFhaWOUNFvXr1fH6fuVB1cvrD2uTu/vSHdUmq3ut3mZmZatCggXr27Olsk5SUpJCQEK+dut6T1xlPnz6tl19+WX379lXdunX9JueWLVt05MgRhYSE6Nprr1Xz5s3Vv39/ffrpp17J6E7W/Px8hYWFuXwqs169epLk/uPERZV5agn50btfS1u2bJldc8019s0335iZ/7wzrjIrVqyw0NDQS1ax9VTp4/zhhx+aJDt69KhLu9tuu81uv/12L6f7TkXz8c0337T69eubw+GwuLg427hxo/fDlVI6a8m7ySMjI+3pp5+2rVu32rRp08zhcNi6dev8JmdBQYElJCTYbbfdZqdPn7b8/HybPn26SbLk5GSf5SwsLLQBAwbYddddV2GbkydPWkJCgv32t7/1YjJXFeWcP3++paen244dO+yll16yFi1a2C233OKjlBXn/Prrry05OdkkWZ06dSwqKsrWrFnjk4w7duyw+vXrW2hoqEVHR9vq1avLtPGHT5RUlvOhhx6yBx54wPl/X66p1dmfZmajR4+2K6+80svpvlPZ2v7yyy9bWFhYmT69evWyRx991Jsxq/U3yNy5c61+/fomyTp27Oizd+1WljUzM9MkWaNGjWzRokW2ZcsWGzdunIWFhdkXX3zhNzlPnDhhUVFRNnbsWMvLy7OzZ8/amDFjTJI99NBDXs35r3/9y1555RXbvn27paenW58+fSwhIcFyc3PLtPXlulRVTn9Zl6rK6S/rUnh4uIWHh9vEiRNty5YtNn/+fIuIiLAXXnihTFtfvmu3Ojn/9Kc/2c033+x8976vPlHizj41M5sxY4Y1bNjQ+RjhC6Wfy/nTunShip5z+su6dKHSWf1pXaospz+tS4WFhTZ+/HhzOBxWp04dczgc9qc//anC9t988411797d7rrrLi+mrF7OZcuW2apVq2zHjh32+uuv25VXXmm9evWy//73v36X9dtvv7Vhw4Y516ewsDBbsmSJV3OamfXp08duvPFGO3LkiP33v/+1F1980UJCQuyKK65waefLtak6Of1lbaru/jTz3brkyet3f/zjH8sdQ9OmTe0vf/mL3+Qs8eijj1pkZKRJsu9973v21VdfXZKMnuZctmyZSbKEhAT7+9//bps2bbI777zTGjdubKdOnfKrrJ9++qnVqVPHZs6cafn5+Xb69Gn7yU9+YpIqXSvKvf2LHkEt4K+FkoMHD1pMTIxt377duS0QCiXJycn24x//2Ncxygj0QsnZs2dt9+7dlpmZaffff7+1bt3ajh8/7v2AFyid9ciRIybJ7rzzTpd2AwcOtDvuuMPL6b5T3j7dtGmTde3a1SRZaGiopaSkWP/+/S01NdU3Ic1s1KhR1qpVKzt06FC5v8/JybHevXtbamrqJftIZnVUlbNERkaGT093UFHOMWPGWO/eve2dd96xbdu22eOPP27R0dG2Y8cOr2fMz8+33bt326ZNm2zChAnWpEkT++yzz1za+EOhpKKcq1atsvbt29uZM2ecbX25plZnf547d86io6PtySef9EnGqtZ2f3lBqrp/g2RnZ9sXX3xh69evt4EDB1r37t29/kSqqqwl6/3EiRNd+nXu3NkmTJjgNznNik+/1bZtW3M4HBYaGmr33HOPde/e3UaNGuW1nOX5+uuvLSoqqsxH7/1lXSpRUc4Svl6XSpTO6S/rUt26da1Pnz4u237xi1/Y9773vTJtffliVFU5N23aZLGxsS4voPvqxSh39qmZWceOHW3MmDHeiFah0s/l/GVdKq2i55z+sC6VVjqrv6xLpZW3T/1lXVq2bJm1bNnSli1bZjt27LClS5dao0aNyi06FhQU2MCBA+3aa691OW2Yv+UssXfvXp+cVrs6WZ944gm74oor7I033rDt27fb7Nmz7bLLLvP6qdf27NljN9xwg/O5e69evezuu++2Tp06ubTzdaGkspz+tDZVd3+a+W5dCoZCycmTJ23Xrl329ttv23XXXWc/+tGPLtnXF3iS8+WXXzZJNn/+fOfvv/32W2vSpInNmzfvkuT0NGtJ3tjYWAsNDbWwsDD7zW9+Y7GxsTZ9+nT3bv+i0tcS/looef31150PXCUXSc4/Urz9joPqOHDggIWEhNjKlSt9HaWM0se55A+S0ovoDTfcYL/85S+9G+4C1Z2P7du3d7syWtNKZ83Pz7c6derY1KlTXdo9+uij1rdvXy+n+05l+zQ7O9tOnDhhZma9e/e2n//8515M9p2HH37YWrZsafv27Sv397m5udanTx/r16+fT5/sVZXzQmfPnjVJlp6e7oVkrirKuWfPnjLfT2Nm1q9fP/vZz37mzYjl6tevX5l36PlDoaS0kpxjx451rkkXrlMhISF24403+jpmuftz6dKlVrduXef93tuqWtvfeeedco93QkKCPf30036Ts7y/QfLz8y0yMtL+9re/eS1ndbKW3O9ffPFFl3633367V99p6s4+PXnypHMOxMbG2syZM72WsyI9e/Z0eQHPX9al0krnvJAv16XSSnL607qUkJDg8glBM7O//OUvFhcXV6atL1+MqirnM888U+Ha1KpVK7/KeqH333/fJNm2bdu8Fa+M8p7LlRQYfb0uXai6zzl9tS5dqLys+/bt84t16UJV7VNfr0stW7a0OXPmuGybOnWqdezY0WVbQUGBDRkyxLp06XJJ351dkermLO1Sv/BYnqqynjt3zurWrVvmuyweeOABS0lJ8VrOC509e9b5Yuntt99uP/rRj1x+7+tCSYnycvrT2lRZzgv5cl3y5PW7hQsXWoMGDVx+f/78eQsNDbXXXnvNb3KW59ChQybJ/v3vf/tNznfffdck2f/+7/+6tOndu/cl/ST5xe7TrKwsO3PmjJ09e9ZCQkLslVdecev2+Y4SP9avXz998skn2rZtm/PSs2dP3X333dq2bZtCQ0N9HbGMxYsXKyYmRgMGDPB1lCq1adNGzZo1U0ZGhnNbbm6uNmzY4NffEVCiqKjI+V0b/iIsLEy9evXSrl27XLZ/8cUXatWqlY9SVS46OlpNmzbV7t27tWnTJg0ePNirt29mGjNmjF5//XW9++67atOmTZk2ubm5Sk5OVlhYmN54440y5xP1l5ylbdu2TZLUvHnzS5zuO1XlLDk3/YXnrpSk0NBQFRUVeS1nRfzxfl2ekpwTJkzQjh07XNYpSXrmmWe0ePFi34ZU+ftz4cKFGjRokJo2beqTTFWt7T179lTdunVd1qZdu3bp4MGDXl2bPPkbxIrfgOP1OVxV1rZt2youLs7na5M7+7RJkyZq0KCB3n33XZ04cUKDBg3yWs7ynD17Vnv37nU+nvvDulSe0jlL88W6VJ4Lc/rTunTdddf5/H5SHVXlvPfee8usTXFxcXrkkUe0Zs0av8p6oYULF6pHjx4+/Q7C8p7L9ejRwy/WpQtV9zmnr9alC5WXtXXr1n6xLl2oqn3q63Xp3LlzVT5Onj9/Xrfffrt2796td955R40bN/ZqRql6OUs7fPiwTp065fW1qaqs58+f1/nz5/1ifSpRv359NW/eXF9//bXWrFnj9efu1VVeTn9amyrLeSF/WJdKVOf1uz59+ig7O1ubN292tnn33XdVVFSkxMREv8lZnpL7lLfWq+rk7NGjh8LDw13WqvPnz+vAgQNeXavc3aexsbG67LLLtGLFCkVEROjmm2927wbdKqvUImfOnLGtW7fa1q1bTZLz+xS+/PJLX0erlD+fequwsNASEhJs/Pjxvo7iVNVxnj59ujVo0MB5jtDBgwdbmzZtvP7OyMpynj171iZOnGiZmZl24MAB27Rpk40YMcLCw8PLvPvQ11nNzF577TWrW7euPf/887Z7926bPXu2hYaGlqlC+zrnK6+8Yu+9957t3bvXVq5caa1atbJbb73VqxnNir8rITo62tatW2fHjh1zXs6dO2dmxac1SUxMtM6dO9uePXtc2njzU2VV5dyzZ4/9/ve/t02bNtn+/ftt1apV1rZtW7vhhhu8lrE6OQsKCqx9+/b2/e9/3zZs2GB79uyxJ5980hwOR4XfZ3GpTJgwwdavX2/79++3HTt22IQJE8zhcNjbb79tZsXf+bN161ZbsGCBSbL333/ftm7deknPB+pJztLko09pVifn7t27zeFw2FtvveX1fJUpvbaPGjXKEhIS7N1337VNmzZZnz59ypy6xRcuzLl3717705/+ZJs2bbIvv/zSPvzwQxs4cKA1atTI56eFNCu7T5955hmLioqyV1991Xbv3m2PPfaYRURE+PwUTKVzLlq0yDIzM23Pnj324osvWqNGjSwtLc3ruX7961/bunXrbP/+/fbhhx9aUlKSNWnSxE6cOOE361JVOf1lXaoqpz+tSxs3brQ6derYH//4R9u9e7e9/PLLFhkZaS+99JKzzalTp2zr1q22evVqk2TLly+3rVu32rFjx/wqZ2m+Or1JdbPm5ORYZGSkPffcc17PWKKy53L+tC5VlNMf16XK9qk/rUuV5fSXdWn48OHWokUL++c//2n79++31157zZo0aeI8/VtBQYENGjTIWrZsadu2bXNZm/Lz8/0m55kzZ+w3v/mNZWZm2v79++2dd96x7t27W4cOHezbb7/1Ws7qZDUr/jvl6quvtvfee8/27dtnixcvtoiIiEt2GqOKpKen21tvvWX79u2zt99+27p27WqJiYnO0376w9pUnZyl+Wptqk5OX6xLNfH6XWpqql177bW2YcMG++CDD6xDhw5lTg3v65wfffSRzZ4927Zu3WoHDhywjIwM69u3r7Vr165GHwdqYn+OHTvWWrRoYWvWrLHPP//cHnjgAYuJibHTp0/XWM6ayjp79mzbvHmz7dq1y+bMmWP16tWzZ5991u0sQVsoKTmVSenL8OHDfR2tUv5cKFmzZo1Jsl27dvk6ilNVx7moqMgmTZpksbGxFh4ebv369fNJ/spyfvPNN3bLLbdYXFychYWFWfPmzW3QoEE++zL36tx3Fi5caO3bt7eIiAjr2rWrT07FVlXOZ5991lq2bGl169a1hIQEe+yxx7z6R3SJ8jJKssWLF1c6Dkm2f/9+v8l58OBBu+GGG6xRo0YWHh5u7du3t0ceecTr5wWuKqeZ2RdffGG33nqrxcTEWGRkpHXp0sWWLl3q1ZxmZvfff7+1atXKwsLCrGnTptavXz+XF/WnTJlS5Vj8IWdpviqUVCfnxIkTLT4+3goLC72erzKl1/ZvvvnGfv7zn1vDhg0tMjLSbrnlFq8/0SvPhTmPHDli/fv3t5iYGKtbt661bNnS7rrrLvv88899G/L/lff30rRp06xly5YWGRlpffr08XoBvzylc44fP95iY2Otbt261qFDB3vqqacu2bmKKzN06FBr3ry5hYWFWYsWLWzo0KHOF+/8ZV2qKqe/rEtV5TTzn3XJzOzNN9+0a665xsLDw61Tp072/PPPu/x+8eLF5R77KVOm+FXO0nz1YpRZ9bLOnz/f6tWrZ9nZ2T5IWKyy53L+tC5VlNMf16Wqnh/7y7pUWU5/WZdyc3Nt7NixlpCQYBEREda2bVv73e9+53z+VnLKpfIu7733nt/kPHfunCUnJ1vTpk2tbt261qpVKxs5cqRlZWV5LWN1s5oVv3Hrvvvus7i4OIuIiLCOHTv6ZA6sWLHC2rZta2FhYdasWTN7+OGHXR4v/WVtqipnab5am6qT0xfrUk28fnfq1Cm788477bLLLrOoqCgbMWKEy/dp+kPOHTt22A9+8APn36itW7e2UaNG2eHDh/0qp1lxEfrXv/61xcTE2OWXX25JSUmX5A3bNZH13nvvtUaNGllYWNhF/S3tMDMTAAAAAAAAAABAEOI7SgAAAAAAAAAAQNCiUAIAAAAAAAAAAIIWhRIAAAAAAAAAABC0KJQAAAAAAAAAAICgRaEEAAAAAAAAAAAELQolAAAAAAAAAAAgaFEoAQAAAAAAAAAAQYtCCQAAAAAAAAAACFoUSgAAAAAAAAAAQNCiUAIAAAAAAAAAAIIWhRIAAAAAAAAAABC0/g+V4Fs1sweYtAAAAABJRU5ErkJggg==",
      "text/plain": [
       "<Figure size 2000x200 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# plot is polymorif against position in a nice wide scatterplot\n",
    "\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.figure(figsize=(20, 2))\n",
    "df_focus_region = ref_seq_w_afrequency_df[(ref_seq_w_afrequency_df['relative_pos'] >= 1764) & (ref_seq_w_afrequency_df['relative_pos'] <= 1874)]\n",
    "#have x axis limits between 1 and 1874-1764+1\n",
    "plt.xlim(0, 1874-1764)\n",
    "#have ticks every 3 bases\n",
    "plt.xticks(np.arange(1, 1874-1764+1, 3))\n",
    "plt.scatter(df_focus_region['relative_pos']-1764+1, df_focus_region['is_polymorfic_pos'], s=10, color='blue', alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snp_reconstruct' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[39], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m plot_weights((\u001b[43msnp_reconstruct\u001b[49m[i]\u001b[38;5;241m*\u001b[39mcompute_per_position_ic(snp_reconstruct[i])\u001b[38;5;241m.\u001b[39mswapaxes(\u001b[38;5;241m0\u001b[39m,\u001b[38;5;241m1\u001b[39m))\u001b[38;5;241m.\u001b[39mnumpy()[\u001b[38;5;241m1764\u001b[39m:\u001b[38;5;241m1874\u001b[39m], subticks_frequency\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'snp_reconstruct' is not defined"
     ]
    }
   ],
   "source": [
    "plot_weights((snp_reconstruct[i]*compute_per_position_ic(snp_reconstruct[i]).swapaxes(0,1)).numpy()[1764:1874], subticks_frequency=3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>relative_pos</th>\n",
       "      <th>chrom</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>ReadPosRankSum</th>\n",
       "      <th>VarDP</th>\n",
       "      <th>AS_VQSLOD</th>\n",
       "      <th>AC_grpmax</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "      <th>is_polymorfic_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>C</td>\n",
       "      <td>11087549</td>\n",
       "      <td>0</td>\n",
       "      <td>19</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>150800.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>59.8265</td>\n",
       "      <td>17.1919</td>\n",
       "      <td>0.857</td>\n",
       "      <td>99.0</td>\n",
       "      <td>-0.6348</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>T</td>\n",
       "      <td>11087550</td>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>151718.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>17.1739</td>\n",
       "      <td>0.424</td>\n",
       "      <td>46.0</td>\n",
       "      <td>4.8335</td>\n",
       "      <td>1.0</td>\n",
       "      <td>67936.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000015</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>T</td>\n",
       "      <td>11087551</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>C</td>\n",
       "      <td>11087552</td>\n",
       "      <td>3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>T</td>\n",
       "      <td>11087553</td>\n",
       "      <td>4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2074</th>\n",
       "      <td>G</td>\n",
       "      <td>11089547</td>\n",
       "      <td>1998</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2075</th>\n",
       "      <td>C</td>\n",
       "      <td>11089548</td>\n",
       "      <td>1999</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2076</th>\n",
       "      <td>A</td>\n",
       "      <td>11089549</td>\n",
       "      <td>2000</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2077</th>\n",
       "      <td>T</td>\n",
       "      <td>11089550</td>\n",
       "      <td>2001</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2078</th>\n",
       "      <td>G</td>\n",
       "      <td>11089551</td>\n",
       "      <td>2002</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2079 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq       pos  relative_pos chrom ref alt filter   AC        AN  \\\n",
       "0      C  11087549             0    19   C   T   PASS  1.0  150800.0   \n",
       "1      T  11087550             1    19   T   C   PASS  1.0  151718.0   \n",
       "2      T  11087551             2     0   0   0      0  0.0       0.0   \n",
       "3      C  11087552             3     0   0   0      0  0.0       0.0   \n",
       "4      T  11087553             4     0   0   0      0  0.0       0.0   \n",
       "...   ..       ...           ...   ...  ..  ..    ...  ...       ...   \n",
       "2074   G  11089547          1998     0   0   0      0  0.0       0.0   \n",
       "2075   C  11089548          1999     0   0   0      0  0.0       0.0   \n",
       "2076   A  11089549          2000     0   0   0      0  0.0       0.0   \n",
       "2077   T  11089550          2001     0   0   0      0  0.0       0.0   \n",
       "2078   G  11089551          2002     0   0   0      0  0.0       0.0   \n",
       "\n",
       "            AF       MQ       QD  ReadPosRankSum  VarDP  AS_VQSLOD  AC_grpmax  \\\n",
       "0     0.000007  59.8265  17.1919           0.857   99.0    -0.6348        0.0   \n",
       "1     0.000007     60.0  17.1739           0.424   46.0     4.8335        1.0   \n",
       "2     0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "3     0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "4     0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "...        ...      ...      ...             ...    ...        ...        ...   \n",
       "2074  0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "2075  0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "2076  0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "2077  0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "2078  0.000000        0   0.0000           0.000    0.0     0.0000        0.0   \n",
       "\n",
       "      AN_grpmax  AF_grpmax  AF_eas    AF_nfe  AF_fin  AF_afr  AF_asj  \\\n",
       "0           0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "1       67936.0   0.000015     0.0  0.000015     0.0     0.0     0.0   \n",
       "2           0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "3           0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "4           0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "...         ...        ...     ...       ...     ...     ...     ...   \n",
       "2074        0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "2075        0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "2076        0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "2077        0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "2078        0.0   0.000000     0.0  0.000000     0.0     0.0     0.0   \n",
       "\n",
       "     is_polymorfic_pos  \n",
       "0                 True  \n",
       "1                 True  \n",
       "2                    0  \n",
       "3                    0  \n",
       "4                    0  \n",
       "...                ...  \n",
       "2074                 0  \n",
       "2075                 0  \n",
       "2076                 0  \n",
       "2077                 0  \n",
       "2078                 0  \n",
       "\n",
       "[2079 rows x 24 columns]"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seq_w_afrequency_df = ref_seq_df.merge(df_region, how='left')\n",
    "ref_seq_w_afrequency_df.fillna(0, inplace=True)\n",
    "ref_seq_w_afrequency_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>relative_pos</th>\n",
       "      <th>AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>0.000007</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>4</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1998</th>\n",
       "      <td>1998</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1999</th>\n",
       "      <td>1999</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2000</th>\n",
       "      <td>2000</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2001</th>\n",
       "      <td>2001</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2002</th>\n",
       "      <td>2002</td>\n",
       "      <td>0.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2003 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      relative_pos        AF\n",
       "0                0  0.000007\n",
       "1                1  0.000007\n",
       "2                2  0.000000\n",
       "3                3  0.000000\n",
       "4                4  0.000000\n",
       "...            ...       ...\n",
       "1998          1998  0.000000\n",
       "1999          1999  0.000000\n",
       "2000          2000  0.000000\n",
       "2001          2001  0.000000\n",
       "2002          2002  0.000000\n",
       "\n",
       "[2003 rows x 2 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seq_max_alternative_af_df = ref_seq_w_afrequency_df.groupby('relative_pos')[['AF']].max().reset_index()\n",
    "ref_seq_max_alternative_af_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<matplotlib.collections.PathCollection at 0x7f0e1c34ad10>"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABk0AAADFCAYAAAD5X7VGAAAAOnRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjEwLjMsIGh0dHBzOi8vbWF0cGxvdGxpYi5vcmcvZiW1igAAAAlwSFlzAAAPYQAAD2EBqD+naQAAMvlJREFUeJzt3Xl8U3W+//F3WmgLaAIUWihtQVxAXIpCqaA/N0qB8YKII1xXXO44aHH04gYzruPD3RmZkYzbVbwug9VRFvUKIqhcFYGyisoqKouAgqS0aIH28/ujt7V7Ewg5J+nr+Xj0kUdPTpt3TpLzSc4n5/v1mJkJAAAAAAAAAACgmYtzOgAAAAAAAAAAAIAb0DQBAAAAAAAAAAAQTRMAAAAAAAAAAABJNE0AAAAAAAAAAAAk0TQBAAAAAAAAAACQRNMEAAAAAAAAAABAEk0TAAAAAAAAAAAASVILpwOEW3l5ubZu3aojjzxSHo/H6TgAAAAAAAAAAMBBZqY9e/YoLS1NcXGNn0sSc02TrVu3KiMjw+kYAAAAAAAAAADARTZt2qT09PRG14mZponf75ff79eBAwckVdx5r9frcCoAAAAAAAAAAOCkoqIiZWRk6Mgjj2xyXY+ZWQQyRUxRUZF8Pp8CgQBNEwAAAAAAAAAAmrlQ+gZMBA8AAAAAAAAAAKAYGp4LACJl82Zp504pOVlqYghEAAAAAAAAAFGEpgkAhGDOHKmgQAoEJJ9PGj1aGjTI6VQAAAAAAAAAwoHhuQAgSJs3VzRMysulnj0rLgsKKpYDAAAAAAAAiH40TQAgSDt3Vpxh0qWLFB9fcRkIVCwHAAAAAAAAEP1ipmni9/vVq1cvZWdnOx0FQIxKTq4YkmvLFqmsrOLS56tYDgAAAAAAACD6eczMnA4RTkVFRfL5fAoEAvJ6vU7HARBjmNMEAAAAAAAAiC6h9A2YCB4AQjBokHT88RVDciUnS+npTicCAAAAAAAAEC40TQAgROnpNEsAAAAAAACAWBQzc5oAAAAAAAAAAAAcCpomAAAAAAAAAAAAomkCAAAAAAAAAAAgiaYJAAAAAAAAAACAJJomAAAAAAAAAAAAkmKoaeL3+9WrVy9lZ2c7HQUAAAAAAAAAAEQhj5mZ0yHCqaioSD6fT4FAQF6v1+k4AAAAAAAAAADAQaH0DWLmTBMAAAAAAAAAAIBDQdMEAAAAAAAAAABANE0AAAAAAAAAAAAk0TQBAAAAAAAAAACQRNMEAAAAAAAAAABAEk0TAAAAAAAAAAAASTRNAAAAAAAAAAAAJNE0AQAAAAAAAAAAkETTBAAAAAAAAAAAQJILmya7d+9W37591bt3b5144ol69tlnnY4EAAAAAAAAAACagRZOB6jtyCOP1Pz589W6dWuVlJToxBNP1MiRI5WcnOx0NAAAAAAAAAAAEMNcd6ZJfHy8WrduLUkqLS2VmcnMHE4FAAAAAAAAAABiXchNk/nz52vYsGFKS0uTx+PR9OnT66zj9/vVrVs3JSUlKScnR4sWLQrpNnbv3q2srCylp6fr1ltvVYcOHUKNCQAAAAAAAAAAEJKQmyYlJSXKysqS3++v9/qCggKNHz9ed999t5YuXaqsrCwNHjxYO3bsqFqncr6S2j9bt26VJLVt21YrVqzQxo0b9c9//lPbt29vME9paamKiopq/AAAAAAAAAAAAITKY4cw9pXH49G0adM0YsSIqmU5OTnKzs7W5MmTJUnl5eXKyMjQDTfcoAkTJoR8G9dff73OPfdc/fa3v633+nvuuUf33ntvneWBQEBerzfk2wMAAAAAAAAAALGjqKhIPp8vqL5BWOc02bdvn5YsWaLc3NxfbyAuTrm5uVqwYEFQ/2P79u3as2ePpIrGx/z589WjR48G1584caICgUDVz6ZNmw7tTgAAAAAAAAAAgGapRTj/2Y8//qiysjKlpqbWWJ6amqrVq1cH9T++/fZbXXvttVUTwN9www066aSTGlw/MTFRiYmJ8vv98vv9KisrO6T7AAAAAAAAAAAAmqewNk3CoV+/flq+fHnIf5efn6/8/Pyq02wAAAAAAAAAAABCEdbhuTp06KD4+Pg6E7dv375dnTp1CudNAQAAAAAAAAAAhFVYmyYJCQnq06eP5s6dW7WsvLxcc+fOVf/+/cN5U3X4/X716tVL2dnZh/V2AAAAAAAAAABAbAp5eK7i4mKtX7++6veNGzdq+fLlat++vTIzMzV+/HiNGTNGffv2Vb9+/TRp0iSVlJToqquuCmvw2hieCwAAAAAAAAAAHIqQmyaFhYU655xzqn4fP368JGnMmDF64YUXNHr0aP3www+66667tG3bNvXu3VuzZs2qMzk8AAAAAAAAAACAm3jMzJwOEQ5+v19+v19lZWVau3atAoGAvF6v07EAAAAAAAAAAICDKkeoCqZvEDNNk0qh3HkAAAAAAAAAABDbQukbhHUieAAAAAAAAAAAgGhF0wQAAAAAAAAAAEAx1DTx+/3q1auXsrOznY4CAAAAAAAAAACiEHOaAAAAAAAAAACAmMWcJgAAAAAAAAAAACGiaQIAAAAAAAAAAKAYapowpwkAAAAAAAAAADgUzGkCAAAAAAAAAABiFnOaAAAAAAAAAAAAhIimCQAAAAAAAAAAgGiaAAAAAAAAAAAASIqhpgkTwQMAAAAAAAAAgEPBRPAAAAAAAAAAACBmMRE8AAAAAAAAAABAiGiaAAAAAAAAAAAAiKYJAAAAAAAAAACAJJomAAAAAAAAAAAAkmKoaeL3+9WrVy9lZ2c7HQUAAAAAAAAAAEQhj5mZ0yHCqaioSD6fT4FAQF6v1+k4AAAAAAAAAADAQaH0DWLmTBMAAAAAAAAAAIBDQdMEAAAAAAAAAABANE0AAAAAAAAAAAAk0TQBAAAAAAAAAACQRNMEAAAAAAAAAABAEk0TAAAAAAAAAAAASTRNAAAAAAAAAAAAJNE0AQAAAAAAAAAAkBRDTRO/369evXopOzvb6SgAAAAAAAAAACAKeczMnA4RTkVFRfL5fAoEAvJ6vU7HAQAAAAAAAAAADgqlbxAzZ5oAAAAAAAAAAAAcCpomAAAAAAAAAAAAomkCAAAAAAAAAAAgiaYJAAAAAAAAAACAJJomAAAAAAAAAAAAkmiaAAAAAAAAAAAASKJpAgAAAAAAAAAAIImmCQAAAAAAAAAAgCQXN0327t2rrl276pZbbnE6CgAAAAAAAAAAaAZc2zS5//77ddpppzkdAwAAAAAAAAAANBOubJqsW7dOq1ev1tChQ52OAgAAAAAAAAAAmomQmybz58/XsGHDlJaWJo/Ho+nTp9dZx+/3q1u3bkpKSlJOTo4WLVoU0m3ccsstevDBB0ONBgAAAAAAAAAAcNBahPoHJSUlysrK0tVXX62RI0fWub6goEDjx4/XU089pZycHE2aNEmDBw/WmjVrlJKSIknq3bu3Dhw4UOdv33vvPS1evFjHHXecjjvuOH366adN5iktLVVpaWnV70VFRaHeJQAAAAAAAAAAAHnMzA76jz0eTZs2TSNGjKhalpOTo+zsbE2ePFmSVF5eroyMDN1www2aMGFCk/9z4sSJevnllxUfH6/i4mLt379fN998s+66665617/nnnt077331lkeCATk9XoP7o4BAAAAAAAAAICYUFRUJJ/PF1TfIKxNk3379ql169b617/+VaORMmbMGO3evVszZswI6f+/8MILWrVqlR577LEG16nvTJOMjAyaJgAAAAAAAAAAIKSmScjDczXmxx9/VFlZmVJTU2ssT01N1erVq8N5U1USExOVmJh4WP43AAAAAAAAAABoPsLaNAm3K6+8Muh1/X6//H6/ysrKDl8gAAAAAAAAAAAQs+LC+c86dOig+Ph4bd++vcby7du3q1OnTuG8qTry8/P15ZdfavHixYf1dgAAAAAAAAAAQGwKa9MkISFBffr00dy5c6uWlZeXa+7cuerfv384b6oOv9+vXr16KTs7+7DeDgAAAAAAAAAAiE0hD89VXFys9evXV/2+ceNGLV++XO3bt1dmZqbGjx+vMWPGqG/fvurXr58mTZqkkpISXXXVVWENXlt+fr7y8/OrJnQBAAAAAAAAAAAIRchNk8LCQp1zzjlVv48fP16SNGbMGL3wwgsaPXq0fvjhB911113atm2bevfurVmzZtWZHB4AAAAAAAAAAMBNPGZmTocIh+oTwa9du1aBQEBer9fpWAAAAAAAAAAAwEGVI1QF0zeImaZJpVDuPAAAAAAAAAAAiG2h9A3COhE8AAAAAAAAAABAtKJpAgAAAAAAAAAAoBhqmvj9fvXq1UvZ2dlORwEAAAAAAAAAAFGIOU0AAAAAAAAAAEDMYk4TAAAAAAAAAACAENE0AQAAAAAAAAAAUAw1TZjTBAAAAAAAAAAAHArmNAEAAAAAAAAAADGLOU0AAAAAAAAAAABCRNMEAAAAAAAAAABANE0AAAAAAAAAAAAkxVDThIngAQAAAAAAAADAoWAieAAAAAAAAAAAELOYCB4AAAAAAAAAACBENE0AAAAAAAAAAABE0wQAAAAAAAAAAEASTRMAAAAAAAAAAABJNE0AAAAAAAAAAAAkxVDTxO/3q1evXsrOznY6CgAAAAAAAAAAiEIeMzOnQ4RTUVGRfD6fAoGAvF6v03EAAAAAAAAAAICDQukbxMyZJgAAAAAAAAAAAIeCpgkAAAAAAAAAAIBomgAAAAAAAAAAAEiiaQIAAAAAAAAAACBJauF0ACBcNm+Wdu6UkpOl9HSn0wAAAAAAgFjFMQigJl4TiCU0TRAT5syRCgqkQEDy+aTRo6VBg5xOBQAAAAAAYg3HIICaeE0g1jA8F6Le5s0VO+bycqlnz4rLgoKK5QAAAAAAAOHCMQigJl4TiEUx0zTx+/3q1auXsrOznY6CCNu5s6KT3aWLFB9fcRkIVCwHAAAAAAAIF45BADXxmkAsipmmSX5+vr788kstXrzY6SiIsOTkilP/tmyRysoqLn2+iuUAAAAAAADhwjEIoCZeE4hFMdM0QfOVnl4xVmJcnLR6dcXl6NFMOgUAAAAAAMKLYxBATbwmEIs8ZmZOhwinoqIi+Xw+BQIBeb1ep+MggjZvrjj1LzmZHTMAAAAAADh8OAYB1MRrAm4XSt+gRYQyAYddejo7ZQAAAAAAcPhxDAKoidcEYgnDcwEAAAAAAAAAAIgzTQAAAAAAAA4Kw9EAABB7aJoAAAAAAACEaM4cqaBACgQkn69i4uNBg5xOBQAADhXDcwEAAAAAAIRg8+aKhkl5udSzZ8VlQUHFcgAAEN1omgAAAAAAAIRg586KM0y6dJHi4ysuA4GK5QAAILq5cniubt26yev1Ki4uTu3atdMHH3zgdCQAAAAAAABJFXOY+HzSli0VDZMtWyp+T052OhkAADhUrmyaSNKnn36qI444wukYAAAAAAAANaSnV8xhUlAgrV7965wmTAYPAED0c23TBAAAAAAAwK0GDZKOP75iSK7kZBomAADEipDnNJk/f76GDRumtLQ0eTweTZ8+vc46fr9f3bp1U1JSknJycrRo0aKQbsPj8eiss85Sdna2XnnllVAjAgAAAAAAHHbp6VJWFg0TAABiSchnmpSUlCgrK0tXX321Ro4cWef6goICjR8/Xk899ZRycnI0adIkDR48WGvWrFFKSookqXfv3jpw4ECdv33vvfeUlpamjz/+WF26dNH333+v3NxcnXTSSTr55JMP4u4BAAAAAAAAAAAEx2NmdtB/7PFo2rRpGjFiRNWynJwcZWdna/LkyZKk8vJyZWRk6IYbbtCECRNCvo1bb71VJ5xwgq688sp6ry8tLVVpaWnV70VFRcrIyFAgEJDX6w359gAAAAAAAAAAQOwoKiqSz+cLqm8Q8vBcjdm3b5+WLFmi3NzcX28gLk65ublasGBBUP+jpKREe/bskSQVFxdr3rx5OuGEExpc/8EHH5TP56v6ycjIOLQ7AQAAAAAAAAAAmqWwNk1+/PFHlZWVKTU1tcby1NRUbdu2Laj/sX37dp1xxhnKysrSaaedpiuuuELZ2dkNrj9x4kQFAoGqn02bNh3SfQAAAAAAAAAAAM1TyHOaHG7du3fXihUrgl4/MTFRiYmJ8vv98vv9KisrO4zpAAAAAAAAAABArArrmSYdOnRQfHy8tm/fXmP59u3b1alTp3DeVB35+fn68ssvtXjx4sN6OwAAAAAAAAAAIDaFtWmSkJCgPn36aO7cuVXLysvLNXfuXPXv3z+cNwUAAAAAAAAAABBWIQ/PVVxcrPXr11f9vnHjRi1fvlzt27dXZmamxo8frzFjxqhv377q16+fJk2apJKSEl111VVhDV4bw3MBAAAAAAAAAIBD4TEzC+UPPvzwQ51zzjl1lo8ZM0YvvPCCJGny5Ml69NFHtW3bNvXu3Vt///vflZOTE5bATSkqKpLP51MgEJDX6z3st7d5s7Rzp5ScLKWnH/abiypsGwA4NAe7H2X/6x48Fu7QHB6H5nAf0Tzx3K5fNG2XaMoKAEBDoqmeRVPWSAqlbxBy08TtItk0mTNHKiiQAgHJ55NGj5YGDTqsNxk12DYAcGgOdj/K/tc9eCzcoTk8Ds3hPqJ54rldv2jaLtGUFQCAhkRTPYumrJEWSt8grHOaOMnv96tXr17Kzs6OyO1t3lzxBCwvl3r2rLgsKKhY3tyxbQDg0BzsfpT9r3vwWLhDc3gcmsN9RPPEc7t+0bRdoikrAAANiaZ6Fk1Z3S5mmib5+fn68ssvtXjx4ojc3s6dFR27Ll2k+PiKy0CgYnlzx7YBgENzsPtR9r/uwWPhDs3hcWgO9xHNE8/t+kXTdommrAAANCSa6lk0ZXW7mGmaRFpycsUpTlu2SGVlFZc+X8Xy5o5tAwCH5mD3o+x/3YPHwh2aw+PQHO4jmiee2/WLpu0STVkBAGhINNWzaMrqdjHTNIn08Fzp6RVjwsXFSatXV1yOHs3kOhLbBgAO1cHuR9n/ugePhTs0h8ehOdxHNE88t+sXTdslmrICANCQaKpn0ZTV7ZgI/hBt3lxxilNyMk/A2tg2AHBoDnY/yv7XPXgs3KE5PA7N4T6ieeK5Xb9o2i7RlBUAgIZEUz2LpqyRFErfgKYJAAAAAAAAAACIWaH0DWJmeC4AAAAAAAAAAIBD0cLpAOHi9/vl9/t14MABSRWdIwAAAAAAAAAA0LxV9guCGXgr5obn2rx5szIyMpyOAQAAAAAAAAAAXGTTpk1Kb2Kyl5hrmpSXl2vr1q068sgj5fF4gv67oqIiZWRkaNOmTa6fCyVaspIzvKIlpxQ9WckZXtGSU4qerOQMv2jJSs7wipacUvRkJWd4RUtOKXqykjP8oiUrOcMrWnJK0ZOVnOEVLTml6MlKzvCLlqyxntPMtGfPHqWlpSkurvFZS2JmeK5KcXFxTXaKGuP1el39pKguWrKSM7yiJacUPVnJGV7RklOKnqzkDL9oyUrO8IqWnFL0ZCVneEVLTil6spIz/KIlKznDK1pyStGTlZzhFS05pejJSs7wi5assZzT5/MFtR4TwQMAAAAAAAAAAIimCQAAAAAAAAAAgCSaJlUSExN19913KzEx0ekoTYqWrOQMr2jJKUVPVnKGV7TklKInKznDL1qykjO8oiWnFD1ZyRle0ZJTip6s5Ay/aMlKzvCKlpxS9GQlZ3hFS04perKSM/yiJSs5fxVzE8EDAAAAAAAAAAAcDM40AQAAAAAAAAAAEE0TAAAAAAAAAAAASTRNAAAAAAAAAAAAJNE0AQAAAAAAAAAAkETTBAAAAAAAAAAAQBJNE82fP1/Dhg1TWlqaPB6Ppk+f7nSkoDz00EPyeDy66aabnI5SR7du3eTxeOr85OfnO5qrqcfazHTXXXepc+fOatWqlXJzc7Vu3TrX5bznnnvUs2dPtWnTRu3atVNubq4WLlzoupyS9NVXX2n48OHy+Xxq06aNsrOz9d1337ku6/bt23XllVcqLS1NrVu31pAhQyL+2D/44IPKzs7WkUceqZSUFI0YMUJr1qypun7Xrl264YYb1KNHD7Vq1UqZmZn6wx/+oEAgENGcwWSVpLPPPrvOPmDs2LGuy7lt2zZdfvnl6tSpk9q0aaNTTz1Vb7zxRkRzPvnkkzr55JPl9Xrl9XrVv39/vfvuu1XXP/PMMzr77LPl9Xrl8Xi0e/fuiOYLJWslM9PQoUMdq6uN5fzmm2/qrVEej0evv/56xLNWV19t/+WXX5Sfn6/k5GQdccQRuvDCC7V9+3bnQqr+nL///e919NFHq1WrVurYsaPOP/98rV692rmQavi90oIFC3TuueeqTZs28nq9OvPMM/Xzzz87E/L/1Jd1w4YNuuCCC9SxY0d5vV6NGjUq4o/9PffcU+d10rNnT0nuqktNZZXcUZeCyemGulRpy5Ytuuyyy5ScnKxWrVrppJNOUmFhYdX1b775pvLy8pScnCyPx6Ply5e7Mmd1Y8eOlcfj0aRJkyIbUk3nbKg2PfrooxHN2dRnObfUpaZyuqUuBfPZ2C11qamsbqhLklRWVqY777xTRx11lFq1aqWjjz5a9913n8xMkrR//37dfvvtOumkk9SmTRulpaXpiiuu0NatW12VU5KuvPLKOtt7yJAhrstZXFyscePGKT09Xa1atVKvXr301FNPRTSnJO3Zs0c33XSTunbtqlatWmnAgAFavHhx1fVuqUvBZK3OydrUVE6nalM4jt/t2rVLl156qbxer9q2batrrrlGxcXFrss5fPhwZWZmKikpSZ07d9bll18e9v1VuI6HvvPOO8rJyVGrVq3Url07jRgxwnU5ly5dqkGDBqlt27ZKTk7Wtddee9CPe7NvmpSUlCgrK0t+v9/pKEFbvHixnn76aZ188slOR6nX4sWL9f3331f9zJkzR5J00UUXOZqrqcf6kUce0d///nc99dRTWrhwodq0aaPBgwfrl19+cVXO4447TpMnT9bnn3+ujz/+WN26dVNeXp5++OEHV+XcsGGDzjjjDPXs2VMffvihVq5cqTvvvFNJSUkRzSk1ntXMNGLECH399deaMWOGli1bpq5duyo3N1clJSURy/jRRx8pPz9fn332mebMmaP9+/crLy+vKsPWrVu1detWPfbYY1q1apVeeOEFzZo1S9dcc03EMgabtdLvfve7GvuCRx55xHU5r7jiCq1Zs0YzZ87U559/rpEjR2rUqFFatmxZxHKmp6froYce0pIlS1RYWKhzzz1X559/vr744gtJ0t69ezVkyBD98Y9/jFimhjSVtdKkSZPk8XgcStl4zoyMjBrPy++//1733nuvjjjiCA0dOtSxzA3V9v/8z//UW2+9pddff10fffSRtm7dqpEjRzqUsuGcffr00ZQpU/TVV19p9uzZMjPl5eWprKzMVTkXLFigIUOGKC8vT4sWLdLixYs1btw4xcU595a4vqwlJSXKy8uTx+PRvHnz9Mknn2jfvn0aNmyYysvLI5rvhBNOqPF6+fjjjyW5qy41lbWS03UpmJxuqEuS9NNPP+n0009Xy5Yt9e677+rLL7/UX/7yF7Vr165qnZKSEp1xxhl6+OGHI5ot1JyVpk2bps8++0xpaWmuzFm7Nj3//PPyeDy68MILI5q1qc9ybqlLTeV0S11qKqeb6lJjWd1Ulx5++GE9+eSTmjx5sr766is9/PDDeuSRR/TEE09IqnjvvHTpUt15551aunSp3nzzTa1Zs0bDhw93Vc5KQ4YMqbHdp06d6rqc48eP16xZs/Tyyy/rq6++0k033aRx48Zp5syZEc36H//xH5ozZ45eeuklff7558rLy1Nubq62bNkiyR11qVJTWSs5WZuCyelUbQrH8btLL71UX3zxhebMmaO3335b8+fP17XXXuu6nOecc45ee+01rVmzRm+88YY2bNig3/72t67L+cYbb+jyyy/XVVddpRUrVuiTTz7RJZdc4qqcW7duVW5uro455hgtXLhQs2bN0hdffKErr7zy4AIZqkiyadOmOR2jUXv27LFjjz3W5syZY2eddZbdeOONTkdq0o033mhHH320lZeXOx2lSu3Hury83Dp16mSPPvpo1bLdu3dbYmKiTZ061YGEFYJ5TgYCAZNk77//fmRC1aO+nKNHj7bLLrvMmUCNqJ11zZo1JslWrVpVtaysrMw6duxozz77rAMJK+zYscMk2UcffdTgOq+99polJCTY/v37I5isrvqyunH/VF/ONm3a2Isvvlhjvfbt2zv62JuZtWvXzv7rv/6rxrIPPvjAJNlPP/3kTKgG1M66bNky69Kli33//feuqqv1bdNKvXv3tquvvjrCiX7VUG3fvXu3tWzZ0l5//fWqdb/66iuTZAsWLHBNzvqsWLHCJNn69esjF/D/NJYzJyfH7rjjjohnakhDWWfPnm1xcXEWCASq1t29e7d5PB6bM2dOxPLdfffdlpWVFfT6TtalprK6pS41ldMtden222+3M844I6h1N27caJJs2bJlhzdUPYLNuXnzZuvSpYutWrXKunbtao8//vjhD1dNKNuz0vnnn2/nnnvuYUoUvOqf5dxWl6pr6jOnk3Wputo53VaXqque1S11yczsvPPOq/O+beTIkXbppZc2+DeLFi0ySfbtt98e7nhVgsk5ZswYO//88yOWqT7B5DzhhBPsz3/+c411Tj31VPvTn/4UkYxmZnv37rX4+Hh7++23m8zhZF0yCz6r07UplG1ayYnadDDH77788kuTZIsXL65a59133zWPx2NbtmxxTc76zJgxwzwej+3bt881Offv329dunRp8PO0W3I+/fTTlpKSYmVlZVXrrFy50iTZunXrQs7Q7M80iTb5+fk677zzlJub63SUoOzbt08vv/yyrr76ake/ddyUjRs3atu2bTW2q8/nU05OjhYsWOBgssbt27dPzzzzjHw+n7KyspyOU6W8vFzvvPOOjjvuOA0ePFgpKSnKyclx5fB3paWlklTjDJi4uDglJibW+YZqJFUOb9K+fftG1/F6vWrRokWkYjWYQ6qb9ZVXXlGHDh104oknauLEidq7d68T8arUl3PAgAEqKCjQrl27VF5erldffVW//PKLzj77bEcylpWV6dVXX1VJSYn69+/vSIZg1Zd17969uuSSS+T3+9WpUyeHE1ZoapsuWbJEy5cvd/Tb8Q3V9iVLlmj//v01lvfs2VOZmZmO1KZg34OUlJRoypQpOuqoo5SRkRGhdL9qKOeOHTu0cOFCpaSkaMCAAUpNTdVZZ53l6L6+oaylpaXyeDxKTEysWpaUlKS4uLiI5123bp3S0tLUvXt3XXrppY0Os+l0XWoqq1vqUmM53VKXZs6cqb59++qiiy5SSkqKTjnlFD377LMRzRCMYHKWl5fr8ssv16233qoTTjjBtTmr2759u9555x1Ha5NU97Oc2+pSQzlrc7ouVaqd0411qaGsbqpLAwYM0Ny5c7V27VpJ0ooVK/Txxx83esZwIBCQx+NR27ZtI5Qy+JwffvihUlJS1KNHD1133XXauXNnxDIGm3PAgAGaOXOmtmzZIjPTBx98oLVr1yovLy9iOQ8cOKCysrI6I1e0atXKFa+Z6oLJ6obaFOo2dUttCub43YIFC9S2bVv17du3ap3c3FzFxcVFbHj7gznOuGvXLr3yyisaMGCAWrZs6ZqcS5cu1ZYtWxQXF6dTTjlFnTt31tChQ7Vq1aqIZAw2Z2lpqRISEmqcrdmqVStJOrj9xEG3fGKQXPSN2PpMnTrVTjzxRPv555/NzD3fmGtMQUGBxcfHH7ZO7sGq/Vh/8sknJsm2bt1aY72LLrrIRo0aFeF0v2roOfnWW29ZmzZtzOPxWFpami1atCjy4aqpnbPyG+atW7e2v/71r7Zs2TJ78MEHzePx2IcffuhcUKubdd++fZaZmWkXXXSR7dq1y0pLS+2hhx4ySZaXl+dIxrKyMjvvvPPs9NNPb3CdH374wTIzM+2Pf/xjBJPV1VDWp59+2mbNmmUrV660l19+2bp06WIXXHCBQykbzvnTTz9ZXl6eSbIWLVqY1+u12bNnRzzfypUrrU2bNhYfH28+n8/eeeedOuu45UyTxrJee+21ds0111T97mRdDWabmpldd911dvzxx0c43a8aq+2vvPKKJSQk1Pmb7Oxsu+222yIZM6j3IH6/39q0aWOSrEePHo58m7exnAsWLDBJ1r59e3v++edt6dKldtNNN1lCQoKtXbvWVVl37NhhXq/XbrzxRispKbHi4mIbN26cSbJrr702Yhn/53/+x1577TVbsWKFzZo1y/r372+ZmZlWVFRUZ12n61JTWd1Sl5rK6Za6lJiYaImJiTZx4kRbunSpPf3005aUlGQvvPBCnXWd/EZvMDkfeOABGzRoUNU3+534Nm8o29PM7OGHH7Z27dpV7R+cUvuznJvqUnUNfeZ0Q12qrnZOt9WlxrK6pS6ZVbyvv/32283j8ViLFi3M4/HYAw880OD6P//8s5166ql2ySWXRDBlcDmnTp1qM2bMsJUrV9q0adPs+OOPt+zsbDtw4ICrcv7yyy92xRVXVNWmhIQE++///u+IZazUv39/O+uss2zLli124MABe+mllywuLs6OO+64Gus5faaJWdNZ3VCbgslZnVO16WCO391///313oeOHTvaP/7xD9fkrHTbbbdZ69atTZKddtpp9uOPPx6WjAebc+rUqSbJMjMz7V//+pcVFhbaxRdfbMnJybZz507X5Fy1apW1aNHCHnnkESstLbVdu3bZhRdeaJIarRMNZjikexBj3Nw0+e677ywlJcVWrFhRtSwamiZ5eXn2b//2b07HqCPamybFxcW2bt06W7BggV199dXWrVs32759e+QD/p/aObds2WKS7OKLL66x3rBhw+zf//3fI5yupvq2aWFhoWVlZZkki4+Pt8GDB9vQoUNtyJAhjmQcO3asde3a1TZt2lTv9YFAwPr162dDhgw5bKdsBquprJXmzp3r6JAIDeUcN26c9evXz95//31bvny53XPPPebz+WzlypURzVdaWmrr1q2zwsJCmzBhgnXo0MG++OKLGuu4pWnSUNYZM2bYMcccY3v27Kla18m6Gsw23bt3r/l8PnvsscccydhUbXfLwalg34Ps3r3b1q5dax999JENGzbMTj311Ih+qGoqZ2WtnzhxYo2/O+mkk2zChAkRy2kW3DadPXu2de/e3Twej8XHx9tll11mp556qo0dOzaiWav76aefzOv11jk13011qVJDWSs5XZcq1c7plrrUsmVL69+/f41lN9xwg5122ml11nXy4FRTOQsLCy01NbXGwXQnDkyFsj3NzHr06GHjxo2LRLRG1f4s55a6VFtDnzmdrku11c7pprpUW33b1C11aerUqZaenm5Tp061lStX2osvvmjt27evtwm5b98+GzZsmJ1yyik1hhZzW85KGzZsiPjQ28HkfPTRR+24446zmTNn2ooVK+yJJ56wI444IuJDs61fv97OPPPMqs/t2dnZdumll1rPnj1rrOeGpkljWd1Sm5rKWZtTtak5NE1++OEHW7Nmjb333nt2+umn229+85vDNsXBweR85ZVXTJI9/fTTVdf/8ssv1qFDB3vqqadck7Mya2pqqsXHx1tCQoLdcsstlpqaag899FDoGQ46fQxyc9Nk2rRpVTuxyh9JVW9YIvlNhGB98803FhcXZ9OnT3c6Sh21H+vKNye1i+qZZ55pf/jDHyIbrppgn5PHHHPMQXVNw6V2ztLSUmvRooXdd999Nda77bbbbMCAARFOV1Nj23T37t22Y8cOMzPr16+fXX/99RFMViE/P9/S09Pt66+/rvf6oqIi69+/vw0cONDxbx82lbW64uJik2SzZs2KQLKaGsq5fv36OvPZmJkNHDjQfv/730cyYh0DBw6s8609tzRNaqvMeuONN1bVpOp1Ki4uzs466yynY9a7TV988UVr2bJl1es+0pqq7e+//369j3lmZqb99a9/dU3O+t6DlJaWWuvWre2f//yna3JWvuZfeumlGn83atSoiH/7NJRt+sMPP1Q9B1JTU+2RRx6JaNba+vbtW+NgnpvqUm21s1bnZF2qrTKnm+pSZmZmjTMHzcz+8Y9/WFpaWp11nTw41VTOxx9/vMHa1LVrV9fkrG7+/PkmyZYvXx6pePWq77NcZbPR6bpUXbCfOZ2oS9XVl/Prr792TV2qrqlt6nRdSk9Pt8mTJ9dYdt9991mPHj1qLNu3b5+NGDHCTj755MP6re2GBJuztsN5ELI+TeXcu3evtWzZss68F9dcc40NHjw4YjmrKy4urjpwOmrUKPvNb35T43o3NE0q1ZfVLbWpqZzVOVmbDub43XPPPWdt27atcf3+/fstPj7e3nzzTdfkrM+mTZtMkn366aeuyTlv3jyTZP/7v/9bY51+/fodtrPMD3V7btu2zfbs2WPFxcUWFxdnr732WsgZmNMkSgwcOFCff/65li9fXvXTt29fXXrppVq+fLni4+OdjljHlClTlJKSovPOO8/pKE066qij1KlTJ82dO7dqWVFRkRYuXOj6eQWkivEwK+fmcIOEhARlZ2drzZo1NZavXbtWXbt2dShV03w+nzp27Kh169apsLBQ559/fsRu28w0btw4TZs2TfPmzdNRRx1VZ52ioiLl5eUpISFBM2fOrDP2aKQEk7W25cuXS5I6d+58mNP9qqmclWPZVx/vUpLi4+NVXl4esZz1cdtrujGVWSdMmKCVK1fWqFOS9Pjjj2vKlCnOhlT92/S5557T8OHD1bFjR0cyNVXb+/btq5YtW9aoTWvWrNF3330X0dp0MO9BrOKLORF9HjeVs3v37kpLS3NFbQplm3bo0EFt27bVvHnztGPHDg0fPjyiWasrLi7Whg0bqvblbqlL9amdtTYn6lJ9qud0U106/fTTXfFaaUpTOS+//PI6tSktLU233nqrZs+e7Zqc1T333HPq06eP4/MV1vdZrk+fPq6oS9UF+5nTibpUXX05u3Xr5pq6VF1T29TpurR3794m95P79+/XqFGjtG7dOr3//vtKTk6OaEYpuJy1bd68WTt37oxobWoq5/79+7V//35X1KZKbdq0UefOnfXTTz9p9uzZEf3cHqr6srqlNjWVszq31CYpuON3/fv31+7du7VkyZKqdebNm6fy8nLl5OS4Jmd9Kl9XkapXweTs06ePEhMTa9Sr/fv365tvvolYvQp1e6ampuqII45QQUGBkpKSNGjQoNBvNOQ2S4zZs2ePLVu2zJYtW2aSquZf+Pbbb52O1iQ3D89VVlZmmZmZdvvttzsdpUpTj/VDDz1kbdu2rRpT9Pzzz7ejjjoq4t+abCxncXGxTZw40RYsWGDffPONFRYW2lVXXWWJiYl1vpXoZE4zszfffNNatmxpzzzzjK1bt86eeOIJi4+Pr9OZdkPW1157zT744APbsGGDTZ8+3bp27WojR46MaMbrrrvOfD6fffjhh/b9999X/ezdu9fMKoY+ycnJsZNOOsnWr19fY51In2nWVNb169fbn//8ZyssLLSNGzfajBkzrHv37nbmmWe6Kue+ffvsmGOOsf/3//6fLVy40NavX2+PPfaYeTyeBue/OBwmTJhgH330kW3cuNFWrlxpEyZMMI/HY++9956ZVcwRtGzZMnv22WdNks2fP9+WLVt22MYOPZSstcmhMziDyblu3TrzeDz27rvvRjxfY2rX9rFjx1pmZqbNmzfPCgsLrX///nWGeHFC9ZwbNmywBx54wAoLC+3bb7+1Tz75xIYNG2bt27d3dOjI2jnNKr517vV67fXXX7d169bZHXfcYUlJSY4P0WRWN+vzzz9vCxYssPXr19tLL71k7du3t/Hjx0c0080332wffvihbdy40T755BPLzc21Dh062I4dO1xVl5rK6pa61FROt9QlM7NFixZZixYt7P7777d169bZK6+8Yq1bt7aXX365ap2dO3fasmXL7J133jFJ9uqrr9qyZcvs+++/d1XO2pwYAiXYnIFAwFq3bm1PPvlkRPPV1thnOTfVpYZyuq0uNbY93VaXGsvqhrpkZjZmzBjr0qWLvf3227Zx40Z78803rUOHDlVDxO3bt8+GDx9u6enptnz58hr1qbS01DU59+zZY7fccostWLDANm7caO+//76deuqpduyxx9ovv/zimpxmFe9RTjjhBPvggw/s66+/tilTplhSUtJhG+aoIbNmzbJ3333Xvv76a3vvvfcsKyvLcnJyqoYGdUNdCjZrbU4NzxVMTidqUziO3w0ZMsROOeUUW7hwoX388cd27LHH1hlC3umcn332mT3xxBO2bNky++abb2zu3Lk2YMAAO/roo8O6HwjH9rzxxhutS5cuNnv2bFu9erVdc801lpKSYrt27XJVzieeeMKWLFlia9asscmTJ1urVq3sb3/720HlafZNk8rhTmr/jBkzxuloTXJz02T27NkmydasWeN0lCpNPdbl5eV25513WmpqqiUmJtrAgQMdyd9Yzp9//tkuuOACS0tLs4SEBOvcubMNHz7ckYngg3ntPPfcc3bMMcdYUlKSZWVlOTZUW1NZ//a3v1l6erq1bNnSMjMz7Y477ojoG2ozqzefJJsyZUqj90GSbdy40VVZv/vuOzvzzDOtffv2lpiYaMccc4zdeuutER9HuKmcZmZr1661kSNHWkpKirVu3dpOPvlke/HFFyOa8+qrr7auXbtaQkKCdezY0QYOHFjj4P7dd9/d5P1wS9banGqaBJNz4sSJlpGRYWVlZRHP15jatf3nn3+266+/3tq1a2etW7e2Cy64wJEPfrVVz7llyxYbOnSopaSkWMuWLS09Pd0uueQSW716tbMhrf73Sg8++KClp6db69atrX///o408+tTO+vtt99uqamp1rJlSzv22GPtL3/5y2Eb27gho0ePts6dO1tCQoJ16dLFRo8eXXUgz011qamsbqlLTeU0c0ddqvTWW2/ZiSeeaImJidazZ0975plnalw/ZcqUeh//u+++21U5a3PqwFQwOZ9++mlr1aqV7d69O+L5qmvss5yb6lJDOd1Wl5r6bOymutRYVjfUJbOKYSFvvPFGy8zMtKSkJOvevbv96U9/qvr8Vjk0U30/H3zwgWty7t271/Ly8qxjx47WsmVL69q1q/3ud7+zbdu2RSxjMDnNKr7EdeWVV1paWpolJSVZjx49HHn8CwoKrHv37paQkGCdOnWy/Pz8GvtLt9SlYLLW5lRtCianE7UpHMfvdu7caRdffLEdccQR5vV67aqrrqox/6Ybcq5cudLOOeecqveo3bp1s7Fjx9rmzZtdldOsoiF98803W0pKih155JGWm5sb9i9vhyPn5Zdfbu3bt7eEhIRDfh/tMTMTAAAAAAAAAABAM8ecJgAAAAAAAAAAAKJpAgAAAAAAAAAAIImmCQAAAAAAAAAAgCSaJgAAAAAAAAAAAJJomgAAAAAAAAAAAEiiaQIAAAAAAAAAACCJpgkAAAAAAAAAAIAkmiYAAAAAAAAAAACSaJoAAAAAAAAAAABIomkCAAAAAAAAAAAgiaYJAAAAAAAAAACAJOn/A6vIs8Z8saA2AAAAAElFTkSuQmCC",
      "text/plain": [
       "<Figure size 2000x200 with 1 Axes>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "\n",
    "plt.figure(figsize=(20, 2))\n",
    "df_focus_region = ref_seq_max_alternative_af_df[(ref_seq_max_alternative_af_df['relative_pos'] >= 1764) & (ref_seq_max_alternative_af_df['relative_pos'] <= 1874)]\n",
    "#have x axis limits between 1 and 1874-1764+1\n",
    "plt.xlim(0, 1874-1764)\n",
    "#have ticks every 3 bases\n",
    "plt.xticks(np.arange(1, 1874-1764+1, 3))\n",
    "#log scale y\n",
    "plt.yscale('log')\n",
    "\n",
    "plt.scatter(df_focus_region['relative_pos']-1764+1, df_focus_region['AF'], s=10, color='blue', alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snp_reconstruct' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[43], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m plot_weights((\u001b[43msnp_reconstruct\u001b[49m[i]\u001b[38;5;241m*\u001b[39mcompute_per_position_ic(snp_reconstruct[i])\u001b[38;5;241m.\u001b[39mswapaxes(\u001b[38;5;241m0\u001b[39m,\u001b[38;5;241m1\u001b[39m))\u001b[38;5;241m.\u001b[39mnumpy()[\u001b[38;5;241m1764\u001b[39m:\u001b[38;5;241m1874\u001b[39m], subticks_frequency\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'snp_reconstruct' is not defined"
     ]
    }
   ],
   "source": [
    "plot_weights((snp_reconstruct[i]*compute_per_position_ic(snp_reconstruct[i]).swapaxes(0,1)).numpy()[1764:1874], subticks_frequency=3)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>pos</th>\n",
       "      <th>relative_pos</th>\n",
       "      <th>chrom</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>filter</th>\n",
       "      <th>AC</th>\n",
       "      <th>AN</th>\n",
       "      <th>AF</th>\n",
       "      <th>MQ</th>\n",
       "      <th>QD</th>\n",
       "      <th>ReadPosRankSum</th>\n",
       "      <th>VarDP</th>\n",
       "      <th>AS_VQSLOD</th>\n",
       "      <th>AC_grpmax</th>\n",
       "      <th>AN_grpmax</th>\n",
       "      <th>AF_grpmax</th>\n",
       "      <th>AF_eas</th>\n",
       "      <th>AF_nfe</th>\n",
       "      <th>AF_fin</th>\n",
       "      <th>AF_afr</th>\n",
       "      <th>AF_asj</th>\n",
       "      <th>is_polymorfic_pos</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1837</th>\n",
       "      <td>G</td>\n",
       "      <td>11089313</td>\n",
       "      <td>1764</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1838</th>\n",
       "      <td>C</td>\n",
       "      <td>11089314</td>\n",
       "      <td>1765</td>\n",
       "      <td>19</td>\n",
       "      <td>CTT</td>\n",
       "      <td>C</td>\n",
       "      <td>PASS</td>\n",
       "      <td>1.0</td>\n",
       "      <td>152146.0</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>60.0</td>\n",
       "      <td>21.6842</td>\n",
       "      <td>-1.28</td>\n",
       "      <td>19.0</td>\n",
       "      <td>0.3414</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4824.0</td>\n",
       "      <td>0.000207</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>True</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1839</th>\n",
       "      <td>T</td>\n",
       "      <td>11089315</td>\n",
       "      <td>1766</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1840</th>\n",
       "      <td>T</td>\n",
       "      <td>11089316</td>\n",
       "      <td>1767</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1841</th>\n",
       "      <td>C</td>\n",
       "      <td>11089317</td>\n",
       "      <td>1768</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1942</th>\n",
       "      <td>G</td>\n",
       "      <td>11089418</td>\n",
       "      <td>1869</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1943</th>\n",
       "      <td>C</td>\n",
       "      <td>11089419</td>\n",
       "      <td>1870</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1944</th>\n",
       "      <td>T</td>\n",
       "      <td>11089420</td>\n",
       "      <td>1871</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1945</th>\n",
       "      <td>A</td>\n",
       "      <td>11089421</td>\n",
       "      <td>1872</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1946</th>\n",
       "      <td>G</td>\n",
       "      <td>11089422</td>\n",
       "      <td>1873</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>110 rows × 24 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     seq       pos  relative_pos chrom  ref alt filter   AC        AN  \\\n",
       "1837   G  11089313          1764     0    0   0      0  0.0       0.0   \n",
       "1838   C  11089314          1765    19  CTT   C   PASS  1.0  152146.0   \n",
       "1839   T  11089315          1766     0    0   0      0  0.0       0.0   \n",
       "1840   T  11089316          1767     0    0   0      0  0.0       0.0   \n",
       "1841   C  11089317          1768     0    0   0      0  0.0       0.0   \n",
       "...   ..       ...           ...   ...  ...  ..    ...  ...       ...   \n",
       "1942   G  11089418          1869     0    0   0      0  0.0       0.0   \n",
       "1943   C  11089419          1870     0    0   0      0  0.0       0.0   \n",
       "1944   T  11089420          1871     0    0   0      0  0.0       0.0   \n",
       "1945   A  11089421          1872     0    0   0      0  0.0       0.0   \n",
       "1946   G  11089422          1873     0    0   0      0  0.0       0.0   \n",
       "\n",
       "            AF    MQ       QD  ReadPosRankSum  VarDP  AS_VQSLOD  AC_grpmax  \\\n",
       "1837  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "1838  0.000007  60.0  21.6842           -1.28   19.0     0.3414        1.0   \n",
       "1839  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "1840  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "1841  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "...        ...   ...      ...             ...    ...        ...        ...   \n",
       "1942  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "1943  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "1944  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "1945  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "1946  0.000000     0   0.0000            0.00    0.0     0.0000        0.0   \n",
       "\n",
       "      AN_grpmax  AF_grpmax  AF_eas  AF_nfe  AF_fin  AF_afr  AF_asj  \\\n",
       "1837        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "1838     4824.0   0.000207     0.0     0.0     0.0     0.0     0.0   \n",
       "1839        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "1840        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "1841        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "...         ...        ...     ...     ...     ...     ...     ...   \n",
       "1942        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "1943        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "1944        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "1945        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "1946        0.0   0.000000     0.0     0.0     0.0     0.0     0.0   \n",
       "\n",
       "     is_polymorfic_pos  \n",
       "1837                 0  \n",
       "1838              True  \n",
       "1839                 0  \n",
       "1840                 0  \n",
       "1841                 0  \n",
       "...                ...  \n",
       "1942                 0  \n",
       "1943                 0  \n",
       "1944                 0  \n",
       "1945                 0  \n",
       "1946                 0  \n",
       "\n",
       "[110 rows x 24 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref_seq_w_afrequency_df[(ref_seq_w_afrequency_df['relative_pos'] >= 1764) & (ref_seq_w_afrequency_df['relative_pos'] < 1874)]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## GPN-MSA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This model was trained on sequences spanning 128bp with the corresponding alignment (MSA) to 90 vertebrate species."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "ename": "ModuleNotFoundError",
     "evalue": "No module named 'gpn.data'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mModuleNotFoundError\u001b[0m                       Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[45], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mos\u001b[39;00m\n\u001b[0;32m----> 2\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mgpn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mdata\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mimport\u001b[39;00m GenomeMSA, Tokenizer\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mgpn\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mmodel\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mmatplotlib\u001b[39;00m\u001b[38;5;21;01m.\u001b[39;00m\u001b[38;5;21;01mpyplot\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;28;01mas\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21;01mplt\u001b[39;00m\n",
      "\u001b[0;31mModuleNotFoundError\u001b[0m: No module named 'gpn.data'"
     ]
    }
   ],
   "source": [
    "import os\n",
    "from gpn.data import GenomeMSA, Tokenizer\n",
    "import gpn.model\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import seaborn as sns\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "import torch\n",
    "from transformers import AutoModel, AutoModelForMaskedLM"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "model_path = \"songlab/gpn-msa-sapiens\"\n",
    "#'zip:///::https://huggingface.co/datasets/songlab/multiz100way/resolve/main/89.zarr.zip'\n",
    "msa_path = \"/s/project/benchmark-lm/ssd-cache/89.zarr\""
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prepare model and dataloader"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'GenomeMSA' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[47], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m genome_msa \u001b[38;5;241m=\u001b[39m \u001b[43mGenomeMSA\u001b[49m(msa_path)  \u001b[38;5;66;03m# can take a minute or two\u001b[39;00m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'GenomeMSA' is not defined"
     ]
    }
   ],
   "source": [
    "genome_msa = GenomeMSA(msa_path)  # can take a minute or two"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "The checkpoint you are trying to load has model type `GPNRoFormer` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.\n\nYou can update Transformers with the command `pip install --upgrade transformers`. If this does not work, and the checkpoint is very new, then there may not be a release version that supports this model yet. In this case, you can get the most up-to-date code by installing Transformers from source with the command `pip install git+https://github.com/huggingface/transformers.git`",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py:1218\u001b[0m, in \u001b[0;36mAutoConfig.from_pretrained\u001b[0;34m(cls, pretrained_model_name_or_path, **kwargs)\u001b[0m\n\u001b[1;32m   1217\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m-> 1218\u001b[0m     config_class \u001b[38;5;241m=\u001b[39m \u001b[43mCONFIG_MAPPING\u001b[49m\u001b[43m[\u001b[49m\u001b[43mconfig_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[38;5;124;43mmodel_type\u001b[39;49m\u001b[38;5;124;43m\"\u001b[39;49m\u001b[43m]\u001b[49m\u001b[43m]\u001b[49m\n\u001b[1;32m   1219\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m:\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py:914\u001b[0m, in \u001b[0;36m_LazyConfigMapping.__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m    913\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m key \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mapping:\n\u001b[0;32m--> 914\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key)\n\u001b[1;32m    915\u001b[0m value \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_mapping[key]\n",
      "\u001b[0;31mKeyError\u001b[0m: 'GPNRoFormer'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[48], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m model \u001b[38;5;241m=\u001b[39m \u001b[43mAutoModelForMaskedLM\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_pretrained\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmodel_path\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      2\u001b[0m model\u001b[38;5;241m.\u001b[39meval();\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/transformers/models/auto/auto_factory.py:547\u001b[0m, in \u001b[0;36m_BaseAutoModelClass.from_pretrained\u001b[0;34m(cls, pretrained_model_name_or_path, *model_args, **kwargs)\u001b[0m\n\u001b[1;32m    544\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwargs\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquantization_config\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[1;32m    545\u001b[0m     _ \u001b[38;5;241m=\u001b[39m kwargs\u001b[38;5;241m.\u001b[39mpop(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mquantization_config\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m--> 547\u001b[0m config, kwargs \u001b[38;5;241m=\u001b[39m \u001b[43mAutoConfig\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mfrom_pretrained\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    548\u001b[0m \u001b[43m    \u001b[49m\u001b[43mpretrained_model_name_or_path\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    549\u001b[0m \u001b[43m    \u001b[49m\u001b[43mreturn_unused_kwargs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mTrue\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    550\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcode_revision\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcode_revision\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    551\u001b[0m \u001b[43m    \u001b[49m\u001b[43m_commit_hash\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcommit_hash\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    552\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mhub_kwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    553\u001b[0m \u001b[43m    \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    554\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    556\u001b[0m \u001b[38;5;66;03m# if torch_dtype=auto was passed here, ensure to pass it on\u001b[39;00m\n\u001b[1;32m    557\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m kwargs_orig\u001b[38;5;241m.\u001b[39mget(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mtorch_dtype\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m==\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mauto\u001b[39m\u001b[38;5;124m\"\u001b[39m:\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/transformers/models/auto/configuration_auto.py:1220\u001b[0m, in \u001b[0;36mAutoConfig.from_pretrained\u001b[0;34m(cls, pretrained_model_name_or_path, **kwargs)\u001b[0m\n\u001b[1;32m   1218\u001b[0m         config_class \u001b[38;5;241m=\u001b[39m CONFIG_MAPPING[config_dict[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmodel_type\u001b[39m\u001b[38;5;124m\"\u001b[39m]]\n\u001b[1;32m   1219\u001b[0m     \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m:\n\u001b[0;32m-> 1220\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   1221\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe checkpoint you are trying to load has model type `\u001b[39m\u001b[38;5;132;01m{\u001b[39;00mconfig_dict[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmodel_type\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m` \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1222\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbut Transformers does not recognize this architecture. This could be because of an \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1223\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124missue with the checkpoint, or because your version of Transformers is out of date.\u001b[39m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;130;01m\\n\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1224\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mYou can update Transformers with the command `pip install --upgrade transformers`. If this \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1225\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdoes not work, and the checkpoint is very new, then there may not be a release version \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1226\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mthat supports this model yet. In this case, you can get the most up-to-date code by installing \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1227\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mTransformers from source with the command \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1228\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m`pip install git+https://github.com/huggingface/transformers.git`\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   1229\u001b[0m         )\n\u001b[1;32m   1230\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m config_class\u001b[38;5;241m.\u001b[39mfrom_dict(config_dict, \u001b[38;5;241m*\u001b[39m\u001b[38;5;241m*\u001b[39munused_kwargs)\n\u001b[1;32m   1231\u001b[0m \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m   1232\u001b[0m     \u001b[38;5;66;03m# Fallback: use pattern matching on the string.\u001b[39;00m\n\u001b[1;32m   1233\u001b[0m     \u001b[38;5;66;03m# We go from longer names to shorter names to catch roberta before bert (for instance)\u001b[39;00m\n",
      "\u001b[0;31mValueError\u001b[0m: The checkpoint you are trying to load has model type `GPNRoFormer` but Transformers does not recognize this architecture. This could be because of an issue with the checkpoint, or because your version of Transformers is out of date.\n\nYou can update Transformers with the command `pip install --upgrade transformers`. If this does not work, and the checkpoint is very new, then there may not be a release version that supports this model yet. In this case, you can get the most up-to-date code by installing Transformers from source with the command `pip install git+https://github.com/huggingface/transformers.git`"
     ]
    }
   ],
   "source": [
    "model = AutoModelForMaskedLM.from_pretrained(model_path)\n",
    "model.eval();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "RotaryBertForMaskedLM(\n",
       "  (bert): RotaryBertModel(\n",
       "    (embeddings): BertEmbeddings(\n",
       "      (word_embeddings): Embedding(4608, 768, padding_idx=0)\n",
       "      (position_embeddings): Embedding(8192, 768)\n",
       "      (token_type_embeddings): Embedding(2, 768)\n",
       "      (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "      (dropout): Dropout(p=0.1, inplace=False)\n",
       "    )\n",
       "    (encoder): RotaryBertEncoder(\n",
       "      (layer): ModuleList(\n",
       "        (0-11): 12 x RotaryBertLayer(\n",
       "          (attention): RotaryBertAttention(\n",
       "            (self): RotaryBertSdpaSelfAttention(\n",
       "              (query): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (key): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (value): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (dropout): Dropout(p=0.1, inplace=False)\n",
       "            )\n",
       "            (output): BertSelfOutput(\n",
       "              (dense): Linear(in_features=768, out_features=768, bias=True)\n",
       "              (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "              (dropout): Dropout(p=0.1, inplace=False)\n",
       "            )\n",
       "          )\n",
       "          (intermediate): BertIntermediate(\n",
       "            (dense): Linear(in_features=768, out_features=3072, bias=True)\n",
       "            (intermediate_act_fn): NewGELUActivation()\n",
       "          )\n",
       "          (output): BertOutput(\n",
       "            (dense): Linear(in_features=3072, out_features=768, bias=True)\n",
       "            (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "            (dropout): Dropout(p=0.1, inplace=False)\n",
       "          )\n",
       "        )\n",
       "      )\n",
       "    )\n",
       "    (pooler): BertPooler(\n",
       "      (dense): Linear(in_features=768, out_features=768, bias=True)\n",
       "      (activation): Tanh()\n",
       "    )\n",
       "  )\n",
       "  (cls): BertOnlyMLMHead(\n",
       "    (predictions): BertLMPredictionHead(\n",
       "      (transform): BertPredictionHeadTransform(\n",
       "        (dense): Linear(in_features=768, out_features=768, bias=True)\n",
       "        (transform_act_fn): NewGELUActivation()\n",
       "        (LayerNorm): LayerNorm((768,), eps=1e-12, elementwise_affine=True)\n",
       "      )\n",
       "      (decoder): Linear(in_features=768, out_features=4608, bias=True)\n",
       "    )\n",
       "  )\n",
       ")"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'Tokenizer' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[50], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m tokenizer_gpn \u001b[38;5;241m=\u001b[39m \u001b[43mTokenizer\u001b[49m()\n\u001b[1;32m      2\u001b[0m mask_idx \u001b[38;5;241m=\u001b[39m tokenizer_gpn\u001b[38;5;241m.\u001b[39mmask_token_id()\n\u001b[1;32m      3\u001b[0m vocab \u001b[38;5;241m=\u001b[39m tokenizer_gpn\u001b[38;5;241m.\u001b[39mvocab\n",
      "\u001b[0;31mNameError\u001b[0m: name 'Tokenizer' is not defined"
     ]
    }
   ],
   "source": [
    "tokenizer_gpn = Tokenizer()\n",
    "mask_idx = tokenizer_gpn.mask_token_id()\n",
    "vocab = tokenizer_gpn.vocab\n",
    "vocab_dict = {v: k for k, v in enumerate(vocab)}\n",
    "acgt_idxs = [vocab_dict.get(nc) for nc in ['A', 'C', 'G', 'T']]\n",
    "\n",
    "def tokenize_region(region, mask_aux_features=False):\n",
    "\n",
    "    if 'Strand' not in region:\n",
    "        region['Strand'] = '+'\n",
    "        #print(\"Strand not provided, defaulting to +\")\n",
    "    # tokenized msa\n",
    "    tokenized_msa = genome_msa.get_msa(region['Chromosome'], region['Start'], region['End'], strand=region['Strand'], tokenize=True)\n",
    "    tokenized_msa = torch.tensor(np.expand_dims(tokenized_msa, 0).astype(np.int64))\n",
    "    input_ids, aux_features = tokenized_msa[:, :, 0], tokenized_msa[:, :, 1:]\n",
    "\n",
    "    if (region['var_pos'] is not None) & (region['mask'] is False):\n",
    "        assert (region['ref'] is not None) | (region['alt'] is not None), \"ref and alt must be provided if var_pos is provided\"\n",
    "        assert input_ids[0, region['var_pos']] == vocab_dict.get(region['ref'].upper())\n",
    "        input_ids[0, region['var_pos']] = vocab_dict.get(region['alt'].upper())\n",
    "\n",
    "    if region['mask']:\n",
    "        assert region['var_pos'] is not None, \"var_pos must be provided if mask is True\"\n",
    "        input_ids[0, region['var_pos']] = mask_idx\n",
    "        if mask_aux_features:\n",
    "            aux_features[0, region['var_pos'], :] = mask_idx\n",
    "\n",
    "    return input_ids, aux_features"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Data loader example"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'genome_msa' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[51], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m alignment \u001b[38;5;241m=\u001b[39m \u001b[43mgenome_msa\u001b[49m\u001b[38;5;241m.\u001b[39mget_msa(\u001b[38;5;28mstr\u001b[39m(ldlr_chr), ldlr_upstream_region_interval[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m1764\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m, ldlr_upstream_region_interval[\u001b[38;5;241m0\u001b[39m]\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m1764\u001b[39m\u001b[38;5;241m+\u001b[39m\u001b[38;5;241m128\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m, strand\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m'\u001b[39m, tokenize\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m      2\u001b[0m \u001b[38;5;66;03m# transpose\u001b[39;00m\n\u001b[1;32m      3\u001b[0m alignment \u001b[38;5;241m=\u001b[39m alignment\u001b[38;5;241m.\u001b[39mT\u001b[38;5;241m.\u001b[39mastype(\u001b[38;5;28mstr\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'genome_msa' is not defined"
     ]
    }
   ],
   "source": [
    "alignment = genome_msa.get_msa(str(ldlr_chr), ldlr_upstream_region_interval[0]+1764-1, ldlr_upstream_region_interval[0]+1764+128-1, strand='+', tokenize=False)\n",
    "# transpose\n",
    "alignment = alignment.T.astype(str)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.set_printoptions(threshold=np.inf, linewidth=100000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'alignment' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[53], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43malignment\u001b[49m\u001b[38;5;241m.\u001b[39mshape\n",
      "\u001b[0;31mNameError\u001b[0m: name 'alignment' is not defined"
     ]
    }
   ],
   "source": [
    "alignment.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'GCTTCACGGGTTAAAAAGCCGATGTCACATCGGCCGTTCGAAACTCCTCCTCTTGCAGTGAGGTGAAGACATTTGAAAATCACCCCACTGCAAACTCCTCCCCCTGCTAGAAACCTCACATTGAAATG'"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ldlr_five_prime_seq[1764:1764+128]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'alignment' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[55], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43malignment\u001b[49m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'alignment' is not defined"
     ]
    }
   ],
   "source": [
    "alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'tokenize_region' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[56], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m input_ids, aux_features \u001b[38;5;241m=\u001b[39m \u001b[43mtokenize_region\u001b[49m({\n\u001b[1;32m      2\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mChromosome\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28mstr\u001b[39m(ldlr_chr),\n\u001b[1;32m      3\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mStart\u001b[39m\u001b[38;5;124m'\u001b[39m: ldlr_upstream_region_interval[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1764\u001b[39m \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m,\n\u001b[1;32m      4\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEnd\u001b[39m\u001b[38;5;124m'\u001b[39m: ldlr_upstream_region_interval[\u001b[38;5;241m0\u001b[39m] \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m1764\u001b[39m \u001b[38;5;241m+\u001b[39m \u001b[38;5;241m128\u001b[39m \u001b[38;5;241m-\u001b[39m \u001b[38;5;241m1\u001b[39m,\n\u001b[1;32m      5\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mStrand\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124m+\u001b[39m\u001b[38;5;124m'\u001b[39m,\n\u001b[1;32m      6\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mvar_pos\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m      7\u001b[0m         \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mmask\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;28;01mFalse\u001b[39;00m\n\u001b[1;32m      8\u001b[0m     })\n",
      "\u001b[0;31mNameError\u001b[0m: name 'tokenize_region' is not defined"
     ]
    }
   ],
   "source": [
    "input_ids, aux_features = tokenize_region({\n",
    "        'Chromosome': str(ldlr_chr),\n",
    "        'Start': ldlr_upstream_region_interval[0] + 1764 - 1,\n",
    "        'End': ldlr_upstream_region_interval[0] + 1764 + 128 - 1,\n",
    "        'Strand': '+',\n",
    "        'var_pos': None,\n",
    "        'mask': False\n",
    "    })"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'input_ids' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[57], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43minput_ids\u001b[49m\u001b[38;5;241m.\u001b[39mshape\n",
      "\u001b[0;31mNameError\u001b[0m: name 'input_ids' is not defined"
     ]
    }
   ],
   "source": [
    "input_ids.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'aux_features' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[58], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43maux_features\u001b[49m\u001b[38;5;241m.\u001b[39mshape\n",
      "\u001b[0;31mNameError\u001b[0m: name 'aux_features' is not defined"
     ]
    }
   ],
   "source": [
    "aux_features.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To get the probabilities for each nucleotide and position we have to mask each position and run the model for each masked position."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'input_ids' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[59], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m all_input_ids \u001b[38;5;241m=\u001b[39m \u001b[43minput_ids\u001b[49m\u001b[38;5;241m.\u001b[39mrepeat(input_ids\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m], \u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m      2\u001b[0m all_input_ids\n",
      "\u001b[0;31mNameError\u001b[0m: name 'input_ids' is not defined"
     ]
    }
   ],
   "source": [
    "all_input_ids = input_ids.repeat(input_ids.shape[1], 1)\n",
    "all_input_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'all_input_ids' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[60], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mall_input_ids\u001b[49m[np\u001b[38;5;241m.\u001b[39marange(input_ids\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m]), np\u001b[38;5;241m.\u001b[39marange(input_ids\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m])] \u001b[38;5;241m=\u001b[39m tokenizer\u001b[38;5;241m.\u001b[39mmask_token_id\n\u001b[1;32m      2\u001b[0m all_input_ids\u001b[38;5;241m.\u001b[39mshape\n",
      "\u001b[0;31mNameError\u001b[0m: name 'all_input_ids' is not defined"
     ]
    }
   ],
   "source": [
    "all_input_ids[np.arange(input_ids.shape[1]), np.arange(input_ids.shape[1])] = tokenizer.mask_token_id\n",
    "all_input_ids.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'aux_features' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[61], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m all_aux_features \u001b[38;5;241m=\u001b[39m \u001b[43maux_features\u001b[49m\u001b[38;5;241m.\u001b[39mrepeat(aux_features\u001b[38;5;241m.\u001b[39mshape[\u001b[38;5;241m1\u001b[39m], \u001b[38;5;241m1\u001b[39m, \u001b[38;5;241m1\u001b[39m)\n\u001b[1;32m      2\u001b[0m all_aux_features\u001b[38;5;241m.\u001b[39mshape\n",
      "\u001b[0;31mNameError\u001b[0m: name 'aux_features' is not defined"
     ]
    }
   ],
   "source": [
    "all_aux_features = aux_features.repeat(aux_features.shape[1], 1, 1)\n",
    "all_aux_features.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "ename": "RuntimeError",
     "evalue": "No CUDA GPUs are available",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mRuntimeError\u001b[0m                              Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[62], line 6\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mrescale_probs_to_sum_1\u001b[39m(probs):\n\u001b[1;32m      4\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m probs \u001b[38;5;241m/\u001b[39m probs\u001b[38;5;241m.\u001b[39msum(axis\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m-\u001b[39m\u001b[38;5;241m1\u001b[39m, keepdims\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m----> 6\u001b[0m \u001b[43mmodel\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcuda\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m      8\u001b[0m probs_list \u001b[38;5;241m=\u001b[39m []\n\u001b[1;32m      9\u001b[0m batch_size \u001b[38;5;241m=\u001b[39m \u001b[38;5;241m8\u001b[39m\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/transformers/modeling_utils.py:4039\u001b[0m, in \u001b[0;36mPreTrainedModel.cuda\u001b[0;34m(self, *args, **kwargs)\u001b[0m\n\u001b[1;32m   4034\u001b[0m     \u001b[38;5;28;01melif\u001b[39;00m version\u001b[38;5;241m.\u001b[39mparse(importlib\u001b[38;5;241m.\u001b[39mmetadata\u001b[38;5;241m.\u001b[39mversion(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbitsandbytes\u001b[39m\u001b[38;5;124m\"\u001b[39m)) \u001b[38;5;241m<\u001b[39m version\u001b[38;5;241m.\u001b[39mparse(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124m0.43.2\u001b[39m\u001b[38;5;124m\"\u001b[39m):\n\u001b[1;32m   4035\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\n\u001b[1;32m   4036\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCalling `cuda()` is not supported for `4-bit` quantized models with the installed version of bitsandbytes. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   4037\u001b[0m             \u001b[38;5;124mf\u001b[39m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mThe current device is `\u001b[39m\u001b[38;5;132;01m{\u001b[39;00m\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mdevice\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m`. If you intended to move the model, please install bitsandbytes >= 0.43.2.\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m   4038\u001b[0m         )\n\u001b[0;32m-> 4039\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43msuper\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcuda\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43margs\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/torch/nn/modules/module.py:1050\u001b[0m, in \u001b[0;36mModule.cuda\u001b[0;34m(self, device)\u001b[0m\n\u001b[1;32m   1033\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcuda\u001b[39m(\u001b[38;5;28mself\u001b[39m: T, device: Optional[Union[\u001b[38;5;28mint\u001b[39m, device]] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m T:\n\u001b[1;32m   1034\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Move all model parameters and buffers to the GPU.\u001b[39;00m\n\u001b[1;32m   1035\u001b[0m \n\u001b[1;32m   1036\u001b[0m \u001b[38;5;124;03m    This also makes associated parameters and buffers different objects. So\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1048\u001b[0m \u001b[38;5;124;03m        Module: self\u001b[39;00m\n\u001b[1;32m   1049\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1050\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43;01mlambda\u001b[39;49;00m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[43m:\u001b[49m\u001b[43m \u001b[49m\u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcuda\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/torch/nn/modules/module.py:900\u001b[0m, in \u001b[0;36mModule._apply\u001b[0;34m(self, fn, recurse)\u001b[0m\n\u001b[1;32m    898\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[1;32m    899\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[0;32m--> 900\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[1;32m    903\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[1;32m    904\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[1;32m    905\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    910\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[1;32m    911\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/torch/nn/modules/module.py:900\u001b[0m, in \u001b[0;36mModule._apply\u001b[0;34m(self, fn, recurse)\u001b[0m\n\u001b[1;32m    898\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[1;32m    899\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[0;32m--> 900\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[1;32m    903\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[1;32m    904\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[1;32m    905\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    910\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[1;32m    911\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/torch/nn/modules/module.py:900\u001b[0m, in \u001b[0;36mModule._apply\u001b[0;34m(self, fn, recurse)\u001b[0m\n\u001b[1;32m    898\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m recurse:\n\u001b[1;32m    899\u001b[0m     \u001b[38;5;28;01mfor\u001b[39;00m module \u001b[38;5;129;01min\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mchildren():\n\u001b[0;32m--> 900\u001b[0m         \u001b[43mmodule\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_apply\u001b[49m\u001b[43m(\u001b[49m\u001b[43mfn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    902\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcompute_should_use_set_data\u001b[39m(tensor, tensor_applied):\n\u001b[1;32m    903\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m torch\u001b[38;5;241m.\u001b[39m_has_compatible_shallow_copy_type(tensor, tensor_applied):\n\u001b[1;32m    904\u001b[0m         \u001b[38;5;66;03m# If the new tensor has compatible tensor type as the existing tensor,\u001b[39;00m\n\u001b[1;32m    905\u001b[0m         \u001b[38;5;66;03m# the current behavior is to change the tensor in-place using `.data =`,\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    910\u001b[0m         \u001b[38;5;66;03m# global flag to let the user control whether they want the future\u001b[39;00m\n\u001b[1;32m    911\u001b[0m         \u001b[38;5;66;03m# behavior of overwriting the existing tensor or not.\u001b[39;00m\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/torch/nn/modules/module.py:927\u001b[0m, in \u001b[0;36mModule._apply\u001b[0;34m(self, fn, recurse)\u001b[0m\n\u001b[1;32m    923\u001b[0m \u001b[38;5;66;03m# Tensors stored in modules are graph leaves, and we don't want to\u001b[39;00m\n\u001b[1;32m    924\u001b[0m \u001b[38;5;66;03m# track autograd history of `param_applied`, so we have to use\u001b[39;00m\n\u001b[1;32m    925\u001b[0m \u001b[38;5;66;03m# `with torch.no_grad():`\u001b[39;00m\n\u001b[1;32m    926\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m torch\u001b[38;5;241m.\u001b[39mno_grad():\n\u001b[0;32m--> 927\u001b[0m     param_applied \u001b[38;5;241m=\u001b[39m \u001b[43mfn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mparam\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    928\u001b[0m p_should_use_set_data \u001b[38;5;241m=\u001b[39m compute_should_use_set_data(param, param_applied)\n\u001b[1;32m    930\u001b[0m \u001b[38;5;66;03m# subclasses may have multiple child tensors so we need to use swap_tensors\u001b[39;00m\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/torch/nn/modules/module.py:1050\u001b[0m, in \u001b[0;36mModule.cuda.<locals>.<lambda>\u001b[0;34m(t)\u001b[0m\n\u001b[1;32m   1033\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m\u001b[38;5;250m \u001b[39m\u001b[38;5;21mcuda\u001b[39m(\u001b[38;5;28mself\u001b[39m: T, device: Optional[Union[\u001b[38;5;28mint\u001b[39m, device]] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m T:\n\u001b[1;32m   1034\u001b[0m \u001b[38;5;250m    \u001b[39m\u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Move all model parameters and buffers to the GPU.\u001b[39;00m\n\u001b[1;32m   1035\u001b[0m \n\u001b[1;32m   1036\u001b[0m \u001b[38;5;124;03m    This also makes associated parameters and buffers different objects. So\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1048\u001b[0m \u001b[38;5;124;03m        Module: self\u001b[39;00m\n\u001b[1;32m   1049\u001b[0m \u001b[38;5;124;03m    \"\"\"\u001b[39;00m\n\u001b[0;32m-> 1050\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_apply(\u001b[38;5;28;01mlambda\u001b[39;00m t: \u001b[43mt\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcuda\u001b[49m\u001b[43m(\u001b[49m\u001b[43mdevice\u001b[49m\u001b[43m)\u001b[49m)\n",
      "File \u001b[0;32m/opt/modules/i12g/anaconda/envs/ml4rg_p7_2/lib/python3.10/site-packages/torch/cuda/__init__.py:319\u001b[0m, in \u001b[0;36m_lazy_init\u001b[0;34m()\u001b[0m\n\u001b[1;32m    317\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCUDA_MODULE_LOADING\u001b[39m\u001b[38;5;124m\"\u001b[39m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;129;01min\u001b[39;00m os\u001b[38;5;241m.\u001b[39menviron:\n\u001b[1;32m    318\u001b[0m     os\u001b[38;5;241m.\u001b[39menviron[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mCUDA_MODULE_LOADING\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mLAZY\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m--> 319\u001b[0m \u001b[43mtorch\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_C\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_cuda_init\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    320\u001b[0m \u001b[38;5;66;03m# Some of the queued calls may reentrantly call _lazy_init();\u001b[39;00m\n\u001b[1;32m    321\u001b[0m \u001b[38;5;66;03m# we need to just return without initializing in that case.\u001b[39;00m\n\u001b[1;32m    322\u001b[0m \u001b[38;5;66;03m# However, we must not let any *other* threads in!\u001b[39;00m\n\u001b[1;32m    323\u001b[0m _tls\u001b[38;5;241m.\u001b[39mis_initializing \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mTrue\u001b[39;00m\n",
      "\u001b[0;31mRuntimeError\u001b[0m: No CUDA GPUs are available"
     ]
    }
   ],
   "source": [
    "from tqdm import tqdm\n",
    "\n",
    "def rescale_probs_to_sum_1(probs):\n",
    "    return probs / probs.sum(axis=-1, keepdims=True)\n",
    "\n",
    "model.cuda()\n",
    "\n",
    "probs_list = []\n",
    "batch_size = 8\n",
    "for i in tqdm(range(0, aux_features.shape[1], batch_size)):\n",
    "    \n",
    "    batch_aux_features = all_aux_features[i:i+batch_size].to('cuda')\n",
    "    batch_input_ids = all_input_ids[i:i+batch_size].to('cuda')\n",
    "\n",
    "    with torch.no_grad():\n",
    "        all_logits = model(input_ids=batch_input_ids, aux_features=batch_aux_features).logits\n",
    "        probs = torch.nn.functional.softmax(all_logits, dim=-1)\n",
    "    probs = probs[:, :, acgt_idxs]\n",
    "    probs = rescale_probs_to_sum_1(probs)\n",
    "    probs_list.append(probs.detach().cpu())\n",
    "\n",
    "probs = torch.cat(probs_list, axis=0)\n",
    "probs = probs[torch.arange(probs.shape[0]), torch.arange(probs.shape[0])].unsqueeze(0)\n",
    "probs.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'probs' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[63], line 2\u001b[0m\n\u001b[1;32m      1\u001b[0m i\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m0\u001b[39m\n\u001b[0;32m----> 2\u001b[0m probs \u001b[38;5;241m=\u001b[39m torch\u001b[38;5;241m.\u001b[39mTensor(\u001b[43mprobs\u001b[49m)\n\u001b[1;32m      3\u001b[0m plot_weights((probs[i]\u001b[38;5;241m*\u001b[39mcompute_per_position_ic(probs[i])\u001b[38;5;241m.\u001b[39mswapaxes(\u001b[38;5;241m0\u001b[39m,\u001b[38;5;241m1\u001b[39m))\u001b[38;5;241m.\u001b[39mnumpy()[:\u001b[38;5;241m110\u001b[39m], subticks_frequency\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'probs' is not defined"
     ]
    }
   ],
   "source": [
    "i=0\n",
    "probs = torch.Tensor(probs)\n",
    "plot_weights((probs[i]*compute_per_position_ic(probs[i]).swapaxes(0,1)).numpy()[:110], subticks_frequency=3)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This was the reconstruction of the same region by the SpeciesLM model. GPN-MSA more confidently predicts the original sequence, but thus it mean it will lead to better fitness scores?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'snp_reconstruct' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[64], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m plot_weights((\u001b[43msnp_reconstruct\u001b[49m[i]\u001b[38;5;241m*\u001b[39mcompute_per_position_ic(snp_reconstruct[i])\u001b[38;5;241m.\u001b[39mswapaxes(\u001b[38;5;241m0\u001b[39m,\u001b[38;5;241m1\u001b[39m))\u001b[38;5;241m.\u001b[39mnumpy()[\u001b[38;5;241m1764\u001b[39m:\u001b[38;5;241m1874\u001b[39m], subticks_frequency\u001b[38;5;241m=\u001b[39m\u001b[38;5;241m3\u001b[39m)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'snp_reconstruct' is not defined"
     ]
    }
   ],
   "source": [
    "plot_weights((snp_reconstruct[i]*compute_per_position_ic(snp_reconstruct[i]).swapaxes(0,1)).numpy()[1764:1874], subticks_frequency=3)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Mutation rate model"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'df' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[65], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m mr\u001b[38;5;241m=\u001b[39m \u001b[43mdf\u001b[49m\u001b[38;5;241m.\u001b[39mto_csv(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m/s/project/ml4rg_students/2025/project07/relative_mutation_rate.tsv\u001b[39m\u001b[38;5;124m'\u001b[39m, sep\u001b[38;5;241m=\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;130;01m\\t\u001b[39;00m\u001b[38;5;124m'\u001b[39m, index\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m)\n\u001b[1;32m      2\u001b[0m \u001b[38;5;28mprint\u001b[39m(mr\u001b[38;5;241m.\u001b[39mhead)\n",
      "\u001b[0;31mNameError\u001b[0m: name 'df' is not defined"
     ]
    }
   ],
   "source": [
    "mr= df.to_csv('/s/project/ml4rg_students/2025/project07/relative_mutation_rate.tsv', sep='\\t', index=False)\n",
    "print(mr.head)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sequence_context</th>\n",
       "      <th>ref</th>\n",
       "      <th>alt</th>\n",
       "      <th>estimated_p_mutated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AAAAAAA</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.002696</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AAAAAAC</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.003904</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AAAAAAG</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.001344</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AAAAAAT</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.002309</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AAAAACA</td>\n",
       "      <td>A</td>\n",
       "      <td>C</td>\n",
       "      <td>0.002943</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49145</th>\n",
       "      <td>ACAGAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.002508</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49146</th>\n",
       "      <td>TAAGAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.002509</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49147</th>\n",
       "      <td>GAAGAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.004403</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49148</th>\n",
       "      <td>CAAGAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.003803</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49149</th>\n",
       "      <td>AAAGAAA</td>\n",
       "      <td>G</td>\n",
       "      <td>T</td>\n",
       "      <td>0.007277</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>49150 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      sequence_context ref alt  estimated_p_mutated\n",
       "0              AAAAAAA   A   C             0.002696\n",
       "1              AAAAAAC   A   C             0.003904\n",
       "2              AAAAAAG   A   C             0.001344\n",
       "3              AAAAAAT   A   C             0.002309\n",
       "4              AAAAACA   A   C             0.002943\n",
       "...                ...  ..  ..                  ...\n",
       "49145          ACAGAAA   G   T             0.002508\n",
       "49146          TAAGAAA   G   T             0.002509\n",
       "49147          GAAGAAA   G   T             0.004403\n",
       "49148          CAAGAAA   G   T             0.003803\n",
       "49149          AAAGAAA   G   T             0.007277\n",
       "\n",
       "[49150 rows x 4 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mutation_rate_model_df = pd.read_csv('/s/project/ml4rg_students/2025/project07/relative_mutation_rate.tsv', sep='\\t')\n",
    "mutation_rate_model_df"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Load genome annotation file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pyranges as pr\n",
    "\n",
    "gtf_df = pr.read_gtf('/s/project/ml4rg_students/2025/project07/data/gencode.v48.annotation.gtf').df\n",
    "gtf_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "gtf_df.Feature.unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The coding regions of the genome are annotated as CDS in the feature column."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Future steps:\n",
    "- You have now seen how to run each DNA LM: SpeciesLM and GPN-MSA, and how to fetch the allele frequencies and counts for a specific region. \n",
    "- Your next goal will be to fine-tune each model on the allele frequency per position in the genome.\n",
    "- Most positions do not have variants, there the allele frequency for variant nucleotides is 0 and the reference nucleotide is 1\n",
    "- Start by creating a train/validation/test split for positions in the genome. The focus is on non-coding regions of the genome. \n",
    "  - The speciesLM model was trained on regions spanning 2kb upstream (5') of the start codon and including it. You could start by fine-tuning your models only in these regions. They are only a small fraction of the genome and are part of what this model has seen during its trainng so this could serve as a good starting point.\n",
    "- Allele frequencies are shaped by mutation biases, unrelated to fitness effects, you can decide whether to correct for this when training the model or afterwards\n",
    "  - using the mutation probabilities estimated by the neutral mutation rate model as another feature or in the loss function. This way you provide what to expect under no selection pressure and the model can focus on the selection/fitness effects\n",
    "  - correcting the predicted allele frequencies by the estimated mutation rate\n",
    "\n",
    "Next week I will send you another notebook containing benchmarks for the models you will develop. In the meantime you can focus on the steps above."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ml4rg_p7_2",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
